{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "f67d4519-07b0-4f27-bcde-f41d89e1be6b",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total cres tested = 882 \n",
      "Total number CREs with cgenes = 145 \n",
      "Total number of target-cgene pairs = 180 \n",
      "Total number of cgenes = 161 \n",
      "Total number CREs with egenes = 515 \n"
     ]
    },
    {
     "data": {
      "text/html": [
       "563"
      ],
      "text/latex": [
       "563"
      ],
      "text/markdown": [
       "563"
      ],
      "text/plain": [
       "[1] 563"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total number of target-egene pairs = 1300 \n",
      "Total number of egenes = 931 \n",
      "CREs without target genes = 319 \n",
      "Total number of overlapping CRES = 97 \n",
      "Number of target cgenes overlapping CRES = 122 \n",
      "Number of target egenes overlapping CRES = 266 \n",
      "Number of overlapping genes = 69 \n"
     ]
    }
   ],
   "source": [
    "suppressPackageStartupMessages({\n",
    "  library(tidyverse)\n",
    "  library(cowplot)\n",
    "  library(gridExtra)\n",
    "  library(data.table)\n",
    "})\n",
    "\n",
    "cres_w_grnas = fread(\"cres_with_grnas.txt\")\n",
    "cres_w_grnas_egene = fread(\"cres_with_grna_eqtls.txt\")\n",
    "\n",
    "total_cres = unique(cres_w_grnas$grna_target)\n",
    "cat(\"Total cres tested =\",length(total_cres),\"\\n\")\n",
    "\n",
    "grna_cres_w_cgenes = unique(cres_w_grnas$grna_target[cres_w_grnas$significant == 1])\n",
    "cat(\"Total number CREs with cgenes =\",length(grna_cres_w_cgenes),\"\\n\")\n",
    "\n",
    "target_cgenes = unique(cres_w_grnas$target_gene[cres_w_grnas$significant == 1])\n",
    "cat(\"Total number of target-cgene pairs =\",length(target_cgenes),\"\\n\")\n",
    "\n",
    "cgenes = unique(cres_w_grnas$ensembl_id[cres_w_grnas$significant == 1])\n",
    "cat(\"Total number of cgenes =\",length(cgenes),\"\\n\")\n",
    "\n",
    "grna_cres_w_egenes = unique(cres_w_grnas_egene$grna_target)\n",
    "cat(\"Total number CREs with egenes =\",length(grna_cres_w_egenes),\"\\n\")\n",
    "\n",
    "length(unique(c(grna_cres_w_cgenes,grna_cres_w_egenes)))\n",
    "\n",
    "target_egenes = unique(cres_w_grnas_egene$target_gene)\n",
    "cat(\"Total number of target-egene pairs =\",length(target_egenes),\"\\n\")\n",
    "\n",
    "egenes = unique(cres_w_grnas_egene$ensembl_id)\n",
    "cat(\"Total number of egenes =\",length(egenes),\"\\n\")\n",
    "\n",
    "cres_no_target = unique(cres_w_grnas$grna_target[!cres_w_grnas$grna_target %in% c(grna_cres_w_cgenes,grna_cres_w_egenes)])\n",
    "cat(\"CREs without target genes =\",length(cres_no_target),\"\\n\")\n",
    "\n",
    "overlapping_cres = grna_cres_w_cgenes[grna_cres_w_cgenes %in% grna_cres_w_egenes]\n",
    "cat(\"Total number of overlapping CRES =\",length(overlapping_cres),\"\\n\")\n",
    "\n",
    "overlapping_cres_cgene = unique(cres_w_grnas$target_gene[cres_w_grnas$grna_target %in% overlapping_cres & cres_w_grnas$significant == 1])\n",
    "cat(\"Number of target cgenes overlapping CRES =\",length(overlapping_cres_cgene),\"\\n\")\n",
    "\n",
    "overlapping_cres_egene = unique(cres_w_grnas_egene$target_gene[cres_w_grnas_egene$grna_target %in% overlapping_cres])\n",
    "cat(\"Number of target egenes overlapping CRES =\",length(overlapping_cres_egene),\"\\n\")\n",
    "\n",
    "overlapping_target_genes = overlapping_cres_cgene[overlapping_cres_cgene %in% overlapping_cres_egene]\n",
    "cat(\"Number of overlapping genes =\",length(overlapping_target_genes),\"\\n\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "id": "b11bd608-1034-48aa-b750-bc4162ba86b1",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "319"
      ],
      "text/latex": [
       "319"
      ],
      "text/markdown": [
       "319"
      ],
      "text/plain": [
       "[1] 319"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "319"
      ],
      "text/latex": [
       "319"
      ],
      "text/markdown": [
       "319"
      ],
      "text/plain": [
       "[1] 319"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A tibble: 17 × 4</caption>\n",
       "<thead>\n",
       "\t<tr><th scope=col>bins</th><th scope=col>n</th><th scope=col>prop</th><th scope=col>Gene_type</th></tr>\n",
       "\t<tr><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;chr&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><td>-800</td><td>  2</td><td>0.006269592</td><td>GWAS (no target)</td></tr>\n",
       "\t<tr><td>-700</td><td>  1</td><td>0.003134796</td><td>GWAS (no target)</td></tr>\n",
       "\t<tr><td>-600</td><td>  1</td><td>0.003134796</td><td>GWAS (no target)</td></tr>\n",
       "\t<tr><td>-500</td><td>  2</td><td>0.006269592</td><td>GWAS (no target)</td></tr>\n",
       "\t<tr><td>-400</td><td>  2</td><td>0.006269592</td><td>GWAS (no target)</td></tr>\n",
       "\t<tr><td>-300</td><td> 12</td><td>0.037617555</td><td>GWAS (no target)</td></tr>\n",
       "\t<tr><td>-200</td><td> 31</td><td>0.097178683</td><td>GWAS (no target)</td></tr>\n",
       "\t<tr><td>-100</td><td>126</td><td>0.394984326</td><td>GWAS (no target)</td></tr>\n",
       "\t<tr><td> 100</td><td> 97</td><td>0.304075235</td><td>GWAS (no target)</td></tr>\n",
       "\t<tr><td> 200</td><td> 19</td><td>0.059561129</td><td>GWAS (no target)</td></tr>\n",
       "\t<tr><td> 300</td><td> 13</td><td>0.040752351</td><td>GWAS (no target)</td></tr>\n",
       "\t<tr><td> 400</td><td>  4</td><td>0.012539185</td><td>GWAS (no target)</td></tr>\n",
       "\t<tr><td> 500</td><td>  1</td><td>0.003134796</td><td>GWAS (no target)</td></tr>\n",
       "\t<tr><td> 600</td><td>  1</td><td>0.003134796</td><td>GWAS (no target)</td></tr>\n",
       "\t<tr><td> 700</td><td>  4</td><td>0.012539185</td><td>GWAS (no target)</td></tr>\n",
       "\t<tr><td> 800</td><td>  1</td><td>0.003134796</td><td>GWAS (no target)</td></tr>\n",
       "\t<tr><td> 900</td><td>  2</td><td>0.006269592</td><td>GWAS (no target)</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A tibble: 17 × 4\n",
       "\\begin{tabular}{llll}\n",
       " bins & n & prop & Gene\\_type\\\\\n",
       " <dbl> & <int> & <dbl> & <chr>\\\\\n",
       "\\hline\n",
       "\t -800 &   2 & 0.006269592 & GWAS (no target)\\\\\n",
       "\t -700 &   1 & 0.003134796 & GWAS (no target)\\\\\n",
       "\t -600 &   1 & 0.003134796 & GWAS (no target)\\\\\n",
       "\t -500 &   2 & 0.006269592 & GWAS (no target)\\\\\n",
       "\t -400 &   2 & 0.006269592 & GWAS (no target)\\\\\n",
       "\t -300 &  12 & 0.037617555 & GWAS (no target)\\\\\n",
       "\t -200 &  31 & 0.097178683 & GWAS (no target)\\\\\n",
       "\t -100 & 126 & 0.394984326 & GWAS (no target)\\\\\n",
       "\t  100 &  97 & 0.304075235 & GWAS (no target)\\\\\n",
       "\t  200 &  19 & 0.059561129 & GWAS (no target)\\\\\n",
       "\t  300 &  13 & 0.040752351 & GWAS (no target)\\\\\n",
       "\t  400 &   4 & 0.012539185 & GWAS (no target)\\\\\n",
       "\t  500 &   1 & 0.003134796 & GWAS (no target)\\\\\n",
       "\t  600 &   1 & 0.003134796 & GWAS (no target)\\\\\n",
       "\t  700 &   4 & 0.012539185 & GWAS (no target)\\\\\n",
       "\t  800 &   1 & 0.003134796 & GWAS (no target)\\\\\n",
       "\t  900 &   2 & 0.006269592 & GWAS (no target)\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A tibble: 17 × 4\n",
       "\n",
       "| bins &lt;dbl&gt; | n &lt;int&gt; | prop &lt;dbl&gt; | Gene_type &lt;chr&gt; |\n",
       "|---|---|---|---|\n",
       "| -800 |   2 | 0.006269592 | GWAS (no target) |\n",
       "| -700 |   1 | 0.003134796 | GWAS (no target) |\n",
       "| -600 |   1 | 0.003134796 | GWAS (no target) |\n",
       "| -500 |   2 | 0.006269592 | GWAS (no target) |\n",
       "| -400 |   2 | 0.006269592 | GWAS (no target) |\n",
       "| -300 |  12 | 0.037617555 | GWAS (no target) |\n",
       "| -200 |  31 | 0.097178683 | GWAS (no target) |\n",
       "| -100 | 126 | 0.394984326 | GWAS (no target) |\n",
       "|  100 |  97 | 0.304075235 | GWAS (no target) |\n",
       "|  200 |  19 | 0.059561129 | GWAS (no target) |\n",
       "|  300 |  13 | 0.040752351 | GWAS (no target) |\n",
       "|  400 |   4 | 0.012539185 | GWAS (no target) |\n",
       "|  500 |   1 | 0.003134796 | GWAS (no target) |\n",
       "|  600 |   1 | 0.003134796 | GWAS (no target) |\n",
       "|  700 |   4 | 0.012539185 | GWAS (no target) |\n",
       "|  800 |   1 | 0.003134796 | GWAS (no target) |\n",
       "|  900 |   2 | 0.006269592 | GWAS (no target) |\n",
       "\n"
      ],
      "text/plain": [
       "   bins n   prop        Gene_type       \n",
       "1  -800   2 0.006269592 GWAS (no target)\n",
       "2  -700   1 0.003134796 GWAS (no target)\n",
       "3  -600   1 0.003134796 GWAS (no target)\n",
       "4  -500   2 0.006269592 GWAS (no target)\n",
       "5  -400   2 0.006269592 GWAS (no target)\n",
       "6  -300  12 0.037617555 GWAS (no target)\n",
       "7  -200  31 0.097178683 GWAS (no target)\n",
       "8  -100 126 0.394984326 GWAS (no target)\n",
       "9   100  97 0.304075235 GWAS (no target)\n",
       "10  200  19 0.059561129 GWAS (no target)\n",
       "11  300  13 0.040752351 GWAS (no target)\n",
       "12  400   4 0.012539185 GWAS (no target)\n",
       "13  500   1 0.003134796 GWAS (no target)\n",
       "14  600   1 0.003134796 GWAS (no target)\n",
       "15  700   4 0.012539185 GWAS (no target)\n",
       "16  800   1 0.003134796 GWAS (no target)\n",
       "17  900   2 0.006269592 GWAS (no target)"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# Load gene distances and find closest gene to non-sig CREs\n",
    "gene_dis = cres_w_grnas[grna_target %in% cres_no_target,] %>%\n",
    "  arrange(abs(tss_distance)) %>%\n",
    "  distinct(grna_target, .keep_all = T)\n",
    "\n",
    "gwas = gene_dis %>%\n",
    "  filter(!is.na(tss_distance)) %>%\n",
    "  mutate(bins = case_when(\n",
    "    tss_distance > -1000000 & tss_distance < -900000 ~ -1000,\n",
    "    tss_distance >= -900000 & tss_distance < -800000 ~ -900,\n",
    "    tss_distance >= -800000 & tss_distance < -700000 ~ -800,\n",
    "    tss_distance >= -700000 & tss_distance < -600000 ~ -700,\n",
    "    tss_distance >= -600000 & tss_distance < -500000 ~ -600,\n",
    "    tss_distance >= -500000 & tss_distance < -400000 ~ -500,\n",
    "    tss_distance >= -400000 & tss_distance < -300000 ~ -400,\n",
    "    tss_distance >= -300000 & tss_distance < -200000 ~ -300,\n",
    "    tss_distance >= -200000 & tss_distance < -100000 ~ -200,\n",
    "    tss_distance >= -100000 & tss_distance < 0 ~ -100,\n",
    "    tss_distance > 0 & tss_distance < 100000 ~ 100,\n",
    "    tss_distance >= 100000 & tss_distance < 200000 ~ 200,\n",
    "    tss_distance >= 200000 & tss_distance < 300000 ~ 300,\n",
    "    tss_distance >= 300000 & tss_distance < 400000 ~ 400,\n",
    "    tss_distance >= 400000 & tss_distance < 500000 ~ 500,\n",
    "    tss_distance >= 500000 & tss_distance < 600000 ~ 600,\n",
    "    tss_distance >= 600000 & tss_distance < 700000 ~ 700,\n",
    "    tss_distance >= 700000 & tss_distance < 800000 ~ 800,\n",
    "    tss_distance >= 800000 & tss_distance < 900000 ~ 900,\n",
    "    tss_distance >= 900000 & tss_distance <= 1000000 ~ 1000,\n",
    "    TRUE ~ NA_real_  # Handle edge cases\n",
    "  )) %>%\n",
    "  group_by(bins) %>%\n",
    "  summarise(n = n(),\n",
    "            prop = n / nrow(gene_dis),\n",
    "            Gene_type = \"GWAS (no target)\")\n",
    "\n",
    "length(cres_no_target)\n",
    "\n",
    "length(gene_dis$grna_target)\n",
    "\n",
    "gwas"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "id": "eaaf8f28-3bf7-4581-9eca-9119e65b7c25",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "0.68384074941452"
      ],
      "text/latex": [
       "0.68384074941452"
      ],
      "text/markdown": [
       "0.68384074941452"
      ],
      "text/plain": [
       "[1] 0.6838407"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "0.811111111111111"
      ],
      "text/latex": [
       "0.811111111111111"
      ],
      "text/markdown": [
       "0.811111111111111"
      ],
      "text/plain": [
       "[1] 0.8111111"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<strong>png:</strong> 2"
      ],
      "text/latex": [
       "\\textbf{png:} 2"
      ],
      "text/markdown": [
       "**png:** 2"
      ],
      "text/plain": [
       "png \n",
       "  2 "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "1281"
      ],
      "text/latex": [
       "1281"
      ],
      "text/markdown": [
       "1281"
      ],
      "text/plain": [
       "[1] 1281"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "180"
      ],
      "text/latex": [
       "180"
      ],
      "text/markdown": [
       "180"
      ],
      "text/plain": [
       "[1] 180"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# set color to economist color pallette\n",
    "colors <- c(\n",
    "  \"cGenes\" = \"#d35e60\",\n",
    "    \"eGenes\" = \"#76c0c1\",\n",
    "  \"GWAS (no target)\" = \"#014d64\"\n",
    ")\n",
    "\n",
    "# Bar plot of TSS\n",
    "n_targets = cres_w_grnas_egene %>%\n",
    "  filter(!is.na(tss_distance)) %>%\n",
    "  distinct(target_gene, .keep_all = TRUE)\n",
    "\n",
    "nrow(n_targets[n_targets$tss_distance < 100000 & n_targets$tss_distance > -100000,])/nrow(n_targets)\n",
    "\n",
    "df1 = cres_w_grnas_egene %>%\n",
    "  filter(!is.na(tss_distance)) %>%\n",
    "  distinct(target_gene, .keep_all = TRUE) %>%\n",
    "  mutate(bins = case_when(\n",
    "    tss_distance > -1000000 & tss_distance < -900000 ~ -1000,\n",
    "    tss_distance >= -900000 & tss_distance < -800000 ~ -900,\n",
    "    tss_distance >= -800000 & tss_distance < -700000 ~ -800,\n",
    "    tss_distance >= -700000 & tss_distance < -600000 ~ -700,\n",
    "    tss_distance >= -600000 & tss_distance < -500000 ~ -600,\n",
    "    tss_distance >= -500000 & tss_distance < -400000 ~ -500,\n",
    "    tss_distance >= -400000 & tss_distance < -300000 ~ -400,\n",
    "    tss_distance >= -300000 & tss_distance < -200000 ~ -300,\n",
    "    tss_distance >= -200000 & tss_distance < -100000 ~ -200,\n",
    "    tss_distance >= -100000 & tss_distance < 0 ~ -100,\n",
    "    tss_distance > 0 & tss_distance < 100000 ~ 100,\n",
    "    tss_distance >= 100000 & tss_distance < 200000 ~ 200,\n",
    "    tss_distance >= 200000 & tss_distance < 300000 ~ 300,\n",
    "    tss_distance >= 300000 & tss_distance < 400000 ~ 400,\n",
    "    tss_distance >= 400000 & tss_distance < 500000 ~ 500,\n",
    "    tss_distance >= 500000 & tss_distance < 600000 ~ 600,\n",
    "    tss_distance >= 600000 & tss_distance < 700000 ~ 700,\n",
    "    tss_distance >= 700000 & tss_distance < 800000 ~ 800,\n",
    "    tss_distance >= 800000 & tss_distance < 900000 ~ 900,\n",
    "    tss_distance >= 900000 & tss_distance <= 1000000 ~ 1000,\n",
    "    TRUE ~ NA_real_  # Handle edge cases\n",
    "  )) %>%\n",
    "  group_by(bins) %>%\n",
    "  summarise(n = n(),\n",
    "            prop = n / nrow(n_targets),\n",
    "            Gene_type = \"eGenes\")\n",
    "\n",
    "n_targets2 = cres_w_grnas %>%\n",
    "  distinct(target_gene, .keep_all = TRUE) %>%\n",
    "  filter(significant ==1)\n",
    "\n",
    "nrow(n_targets2[n_targets2$tss_distance < 100000 & n_targets2$tss_distance > -100000,])/nrow(n_targets2)\n",
    "  \n",
    "df2 <- cres_w_grnas %>% \n",
    "  distinct(target_gene, .keep_all = TRUE) %>% # Remove duplicate target-genes for different eqtl/gwas\n",
    "  filter(significant == 1) %>%\n",
    "  mutate(bins = case_when(\n",
    "    tss_distance > -1000000 & tss_distance < -900000 ~ -1000,\n",
    "    tss_distance >= -900000 & tss_distance < -800000 ~ -900,\n",
    "    tss_distance >= -800000 & tss_distance < -700000 ~ -800,\n",
    "    tss_distance >= -700000 & tss_distance < -600000 ~ -700,\n",
    "    tss_distance >= -600000 & tss_distance < -500000 ~ -600,\n",
    "    tss_distance >= -500000 & tss_distance < -400000 ~ -500,\n",
    "    tss_distance >= -400000 & tss_distance < -300000 ~ -400,\n",
    "    tss_distance >= -300000 & tss_distance < -200000 ~ -300,\n",
    "    tss_distance >= -200000 & tss_distance < -100000 ~ -200,\n",
    "    tss_distance >= -100000 & tss_distance < 0 ~ -100,\n",
    "    tss_distance > 0 & tss_distance < 100000 ~ 100,\n",
    "    tss_distance >= 100000 & tss_distance < 200000 ~ 200,\n",
    "    tss_distance >= 200000 & tss_distance < 300000 ~ 300,\n",
    "    tss_distance >= 300000 & tss_distance < 400000 ~ 400,\n",
    "    tss_distance >= 400000 & tss_distance < 500000 ~ 500,\n",
    "    tss_distance >= 500000 & tss_distance < 600000 ~ 600,\n",
    "    tss_distance >= 600000 & tss_distance < 700000 ~ 700,\n",
    "    tss_distance >= 700000 & tss_distance < 800000 ~ 800,\n",
    "    tss_distance >= 800000 & tss_distance < 900000 ~ 900,\n",
    "    tss_distance >= 900000 & tss_distance <= 1000000 ~ 1000,\n",
    "    TRUE ~ NA_real_  # Handle edge cases\n",
    "  )) %>%\n",
    "  group_by(bins) %>%\n",
    "  summarise(\n",
    "    n = n(),\n",
    "    prop = n / nrow(n_targets2),\n",
    "    Gene_type = \"cGenes\"\n",
    "  )\n",
    "\n",
    "# Add a zero for -1000kb\n",
    "#extra = data.frame(bins\t= c(-1000,-800,-700,500,600),\n",
    "#                   n = c(0,0,0,0,0),\n",
    "#                   prop = c(0,0,0,0,0),\n",
    "#                   Gene_type = \"cGene\")\n",
    "\n",
    "#png(\"plots/figure_plots/tss_distance_barplots.png\", width = 8, height = 8,units = \"in\", res = 300)\n",
    "svg(\"plots/figure_plots/tss_distance_barplots.svg\", width = 8, height = 8)\n",
    "tss_distance_plot = bind_rows(df1, df2, gwas) %>% drop_na() %>%\n",
    "  ggplot(aes(x = factor(bins), y = prop, fill = Gene_type)) +\n",
    "  geom_col(color = \"black\", width = 0.8,position = position_dodge(width = 0.7)) +\n",
    "  labs(x = \"TSS Distance from TSS (kb)\", y = \"Proportion of CRE-gene pairs\", fill = \"Gene Type\") +\n",
    "  scale_fill_manual(\n",
    "    values = colors) +\n",
    "  theme_cowplot() +\n",
    "  theme(\n",
    "    axis.title.x = element_text(size = 16),\n",
    "    axis.title.y = element_text(size = 16, margin = margin(r = 10)),\n",
    "    axis.text.x = element_text(size = 16, angle = 45, hjust = 1),\n",
    "    axis.text.y = element_text(size = 16),\n",
    "    legend.position = \"top\"\n",
    "  )\n",
    "tss_distance_plot\n",
    "dev.off()\n",
    "\n",
    "nrow(n_targets)\n",
    "nrow(n_targets2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "id": "3fc7db7c-ddde-4ffc-95a3-5dac3b047c82",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "proportion of cres within 10kb, with no CRE targets 0.206278 \n",
      "proportion of cres within 10kb, with eQTL targets 0.1873536 \n",
      "proportion of cres within 10kb, with CRISRPi targets 0.4111111 \n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<strong>png:</strong> 2"
      ],
      "text/latex": [
       "\\textbf{png:} 2"
      ],
      "text/markdown": [
       "**png:** 2"
      ],
      "text/plain": [
       "png \n",
       "  2 "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "gwas = gene_dis %>%\n",
    "  filter(!is.na(tss_distance),tss_distance < 100000 & tss_distance > -100000) %>%\n",
    "  mutate(bins = case_when(\n",
    "    tss_distance > -100000 & tss_distance < -90000 ~ -100,\n",
    "    tss_distance >= -90000 & tss_distance < -80000 ~ -90,\n",
    "    tss_distance >= -80000 & tss_distance < -70000 ~ -80,\n",
    "    tss_distance >= -70000 & tss_distance < -60000 ~ -70,\n",
    "    tss_distance >= -60000 & tss_distance < -50000 ~ -60,\n",
    "    tss_distance >= -50000 & tss_distance < -40000 ~ -50,\n",
    "    tss_distance >= -40000 & tss_distance < -30000 ~ -40,\n",
    "    tss_distance >= -30000 & tss_distance < -20000 ~ -30,\n",
    "    tss_distance >= -20000 & tss_distance < -10000 ~ -20,\n",
    "    tss_distance >= -10000 & tss_distance < 0 ~ -10,\n",
    "    tss_distance > 0 & tss_distance < 10000 ~ 10,\n",
    "    tss_distance >= 10000 & tss_distance < 20000 ~ 20,\n",
    "    tss_distance >= 20000 & tss_distance < 30000 ~ 30,\n",
    "    tss_distance >= 30000 & tss_distance < 40000 ~ 40,\n",
    "    tss_distance >= 40000 & tss_distance < 50000 ~ 50,\n",
    "    tss_distance >= 50000 & tss_distance < 60000 ~ 60,\n",
    "    tss_distance >= 60000 & tss_distance < 70000 ~ 70,\n",
    "    tss_distance >= 70000 & tss_distance < 80000 ~ 80,\n",
    "    tss_distance >= 80000 & tss_distance < 90000 ~ 90,\n",
    "    tss_distance >= 90000 & tss_distance <= 100000 ~ 100,\n",
    "    TRUE ~ NA_real_  # Handle edge cases\n",
    "  )) %>%\n",
    "  group_by(bins) %>%\n",
    "  summarise(n = n(),\n",
    "            prop = n / nrow(gene_dis[tss_distance < 100000 & tss_distance > -100000,]),\n",
    "            Gene_type = \"GWAS genes\")\n",
    "\n",
    "cat(\"proportion of cres within 10kb, with no CRE targets\",gwas$prop[gwas$bins == -10] + gwas$prop[gwas$bins == 10],\"\\n\")\n",
    "\n",
    "# line graph of TSS\n",
    "n_targets = cres_w_grnas_egene %>%\n",
    "  filter(!is.na(tss_distance), tss_distance < 100000 & tss_distance > -100000) %>%\n",
    "  distinct(target_gene, .keep_all = TRUE)\n",
    "\n",
    "df1 = cres_w_grnas_egene %>%\n",
    "  filter(!is.na(tss_distance), tss_distance < 100000 & tss_distance > -100000) %>%\n",
    "  distinct(target_gene, .keep_all = TRUE) %>%\n",
    "  mutate(bins = case_when(\n",
    "    tss_distance > -100000 & tss_distance < -90000 ~ -100,\n",
    "    tss_distance >= -90000 & tss_distance < -80000 ~ -90,\n",
    "    tss_distance >= -80000 & tss_distance < -70000 ~ -80,\n",
    "    tss_distance >= -70000 & tss_distance < -60000 ~ -70,\n",
    "    tss_distance >= -60000 & tss_distance < -50000 ~ -60,\n",
    "    tss_distance >= -50000 & tss_distance < -40000 ~ -50,\n",
    "    tss_distance >= -40000 & tss_distance < -30000 ~ -40,\n",
    "    tss_distance >= -30000 & tss_distance < -20000 ~ -30,\n",
    "    tss_distance >= -20000 & tss_distance < -10000 ~ -20,\n",
    "    tss_distance >= -10000 & tss_distance < 0 ~ -10,\n",
    "    tss_distance > 0 & tss_distance < 10000 ~ 10,\n",
    "    tss_distance >= 10000 & tss_distance < 20000 ~ 20,\n",
    "    tss_distance >= 20000 & tss_distance < 30000 ~ 30,\n",
    "    tss_distance >= 30000 & tss_distance < 40000 ~ 40,\n",
    "    tss_distance >= 40000 & tss_distance < 50000 ~ 50,\n",
    "    tss_distance >= 50000 & tss_distance < 60000 ~ 60,\n",
    "    tss_distance >= 60000 & tss_distance < 70000 ~ 70,\n",
    "    tss_distance >= 70000 & tss_distance < 80000 ~ 80,\n",
    "    tss_distance >= 80000 & tss_distance < 90000 ~ 90,\n",
    "    tss_distance >= 90000 & tss_distance <= 100000 ~ 100,\n",
    "    TRUE ~ NA_real_  # Handle edge cases\n",
    "  )) %>%\n",
    "  filter(!is.na(bins)) %>%  # Remove any NA bins\n",
    "  group_by(bins) %>%\n",
    "  summarise(n = n(),\n",
    "            prop = n / 1281,\n",
    "            Gene_type = \"eGenes\")\n",
    "\n",
    "cat(\"proportion of cres within 10kb, with eQTL targets\",df1$prop[df1$bins == -10] + df1$prop[df1$bins == 10],\"\\n\")\n",
    "\n",
    "n_targets2 = cres_w_grnas %>%\n",
    "  distinct(target_gene, .keep_all = TRUE) %>%\n",
    "  filter(significant ==1, tss_distance < 100000 & tss_distance > -100000)\n",
    "  \n",
    "df2 <- cres_w_grnas %>% \n",
    "  distinct(target_gene, .keep_all = TRUE) %>% # Remove duplicate target-genes for different eqtl/gwas\n",
    "  filter(significant == 1, tss_distance < 100000 & tss_distance > -100000) %>%\n",
    "  mutate(bins = case_when(\n",
    "    tss_distance > -100000 & tss_distance < -90000 ~ -100,\n",
    "    tss_distance >= -90000 & tss_distance < -80000 ~ -90,\n",
    "    tss_distance >= -80000 & tss_distance < -70000 ~ -80,\n",
    "    tss_distance >= -70000 & tss_distance < -60000 ~ -70,\n",
    "    tss_distance >= -60000 & tss_distance < -50000 ~ -60,\n",
    "    tss_distance >= -50000 & tss_distance < -40000 ~ -50,\n",
    "    tss_distance >= -40000 & tss_distance < -30000 ~ -40,\n",
    "    tss_distance >= -30000 & tss_distance < -20000 ~ -30,\n",
    "    tss_distance >= -20000 & tss_distance < -10000 ~ -20,\n",
    "    tss_distance >= -10000 & tss_distance < 0 ~ -10,\n",
    "    tss_distance > 0 & tss_distance < 10000 ~ 10,\n",
    "    tss_distance >= 10000 & tss_distance < 20000 ~ 20,\n",
    "    tss_distance >= 20000 & tss_distance < 30000 ~ 30,\n",
    "    tss_distance >= 30000 & tss_distance < 40000 ~ 40,\n",
    "    tss_distance >= 40000 & tss_distance < 50000 ~ 50,\n",
    "    tss_distance >= 50000 & tss_distance < 60000 ~ 60,\n",
    "    tss_distance >= 60000 & tss_distance < 70000 ~ 70,\n",
    "    tss_distance >= 70000 & tss_distance < 80000 ~ 80,\n",
    "    tss_distance >= 80000 & tss_distance < 90000 ~ 90,\n",
    "    tss_distance >= 90000 & tss_distance <= 100000 ~ 100,\n",
    "    TRUE ~ NA_real_  # Handle edge cases\n",
    "  )) %>%\n",
    "  filter(!is.na(bins)) %>%  # Remove any NA bins\n",
    "  group_by(bins) %>%\n",
    "  summarise(\n",
    "    n = n(),\n",
    "    prop = n / 180,\n",
    "    Gene_type = \"cGenes\"\n",
    "  )\n",
    "\n",
    "cat(\"proportion of cres within 10kb, with CRISRPi targets\",df2$prop[df2$bins == -10] + df2$prop[df2$bins == 10],\"\\n\")\n",
    "\n",
    "# Add a zero for -1000kb\n",
    "extra = data.frame(bins\t= c(-90),\n",
    "                   n = c(0),\n",
    "                   prop = c(0),\n",
    "                   Gene_type = \"cGenes\")\n",
    "\n",
    "# Plot with ggplot2\n",
    "#png(\"plots/figure_plots/tss_distance_100kb.png\", width = 8, height = 8,units = \"in\", res = 300)\n",
    "svg(\"plots/figure_plots/tss_distance_100kb.svg\", width = 8, height = 8)\n",
    "\n",
    "colors <- c(\n",
    "  \"cGenes\" = \"#d35e60\",\n",
    "    \"eGenes\" = \"#76c0c1\",\n",
    "  \"GWAS genes\" = \"#014d64\"\n",
    ")\n",
    "\n",
    "tss_distance_plot2 = bind_rows(df2,df1,gwas) %>%\n",
    "  ggplot(aes(x = bins, y = prop, color = Gene_type)) +\n",
    "  geom_line(size = 1) +          # Add a line for each group\n",
    "  geom_point(size = 5) +         # Add dots at each data point\n",
    "  geom_vline(xintercept = 0, linetype = \"dotted\", color = \"black\", size = 0.8) + # Add a vertical dotted line at x = 0\n",
    "  scale_x_continuous(breaks = c(-100,-50,-10,10, 50, 100)) +\n",
    "  scale_color_manual(values = colors) + \n",
    "  labs(\n",
    "    title = \"\",\n",
    "    x = \"TSS Distance from TSS (kb)\",\n",
    "    y = \"Proportion of CRE-gene pairs\",\n",
    "    color = \"Gene Type\"\n",
    "  ) + ylim(0,0.35) +\n",
    "  theme_cowplot() +    \n",
    "  theme(\n",
    "    axis.title.x = element_text(size = 16),\n",
    "    axis.title.y = element_text(size = 16, margin = margin(r = 10)),\n",
    "    axis.text.x = element_text(size = 16, angle = 45, hjust = 1),\n",
    "    axis.text.y = element_text(size = 16),\n",
    "    legend.position = \"none\"\n",
    "  )\n",
    "\n",
    "tss_distance_plot2\n",
    "\n",
    "dev.off()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "id": "b6aaad1e-2e64-4fe6-b162-6f59fad0bf2d",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] 26\n",
      "[1] 16\n",
      "\n",
      "\tWilcoxon rank sum test with continuity correction\n",
      "\n",
      "data:  cgene_metric and egene_metric\n",
      "W = 78164, p-value = 0.000208\n",
      "alternative hypothesis: true location shift is not equal to 0\n",
      "\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message:\n",
      "“\u001b[1m\u001b[22mRemoved 134 rows containing missing values (`geom_violin()`).”\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<strong>png:</strong> 2"
      ],
      "text/latex": [
       "\\textbf{png:} 2"
      ],
      "text/markdown": [
       "**png:** 2"
      ],
      "text/plain": [
       "png \n",
       "  2 "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "gene_metrics = fread(\"data/gene_metrics_vgh_202407.tsv\", fill=T)\n",
    "colnames(gene_metrics) = c(\"ensemble_id\", colnames(gene_metrics)[1:15])\n",
    "\n",
    "### Plot number of enhancers ###\n",
    "# Extract cgene_metric\n",
    "cgene_metric = gene_metrics[ensemble_id %in% cgenes, \"num_enh\"]\n",
    "cgene_metric = na.omit(cgene_metric)\n",
    "cgene_metric <- unlist(cgene_metric)\n",
    "cgene_metric <- as.numeric(cgene_metric)\n",
    "  \n",
    "# Extract egene_metric\n",
    "egene_metric = gene_metrics[ensemble_id %in% egenes, \"num_enh\"]\n",
    "egene_metric = na.omit(egene_metric)\n",
    "egene_metric <- unlist(egene_metric)\n",
    "egene_metric <- as.numeric(egene_metric)\n",
    "\n",
    "print(median(cgene_metric))\n",
    "print(median(egene_metric))\n",
    "\n",
    "test = wilcox.test(cgene_metric, egene_metric)\n",
    "print(test)\n",
    "\n",
    "  # Combine the data into a data frame for plotting\n",
    "combined_data <- data.frame(\n",
    "    Gene_type = c(rep(\"cGenes\", length(cgene_metric)), rep(\"eGenes\", length(egene_metric))),\n",
    "    Value = c(cgene_metric, egene_metric)\n",
    "  )\n",
    "\n",
    "  # Calculate the y-axis limits (just above the last outlier)\n",
    "stats_cgene <- boxplot.stats(cgene_metric)\n",
    "stats_egene <- boxplot.stats(egene_metric)\n",
    "  \n",
    "  # Max of the outliers for cgene and egene\n",
    "  # Check if there are outliers; if not, use the max values from the metrics\n",
    "if (length(stats_cgene$out) > 0 | length(stats_egene$out) > 0) {\n",
    "    max_outlier <- max(c(stats_cgene$out, stats_egene$out), na.rm = TRUE)\n",
    "} else {\n",
    "    max_outlier <- max(c(cgene_metric, egene_metric), na.rm = TRUE)\n",
    "}\n",
    "  \n",
    "# Set the lower limit as the minimum value and upper limit slightly above the max outlier\n",
    "min_value <- min(combined_data$Value, na.rm = TRUE)\n",
    "max_value <- max_outlier * 1.1  # 10% above the highest outlier\n",
    "\n",
    "#png(\"plots/figure_plots/enhancer_plot.png\", width = 4, height = 6,units = \"in\", res = 300)\n",
    "svg(\"plots/figure_plots/enhancer_plot.svg\", width = 4, height = 6)\n",
    "\n",
    "enhancer_plot <- combined_data %>%\n",
    "  filter(Value<130) %>% #improve plot look\n",
    "  ggplot(aes(x = Gene_type, y = Value, fill = Gene_type)) +\n",
    "  geom_violin(trim = FALSE, fill = \"light grey\", color = \"black\", alpha = 0.5) +  # Violin plot\n",
    "  geom_boxplot(width = 0.1, color = \"black\", outlier.shape = NA) +  # Boxplot inside violin plot\n",
    "  scale_fill_manual(values = colors) +\n",
    "  theme_cowplot() +\n",
    "  annotate(\"text\", x = 1.5, y = 175, label = paste0(\"P = \", round(test$p.value, 4)), size = 5, hjust = 0.5, family = \"sans\") +\n",
    "  labs(title = \"\", x = \"\", y = \"Number of enhancers\") +\n",
    "  ylim(0,175) +\n",
    "  theme(\n",
    "    axis.title.x = element_text(size = 16),\n",
    "    axis.title.y = element_text(size = 16, margin = margin(r = 10)),\n",
    "    axis.text.x = element_text(size = 16),\n",
    "    axis.text.y = element_text(size = 16),\n",
    "    legend.position = \"none\"\n",
    "  )\n",
    "enhancer_plot\n",
    "\n",
    "dev.off()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "id": "f7ce2e71-af5b-4f30-b2db-ea41954a22bc",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "     [,1] [,2]\n",
      "[1,]   44  155\n",
      "[2,]  109  574\n",
      "\n",
      "\tFisher's Exact Test for Count Data\n",
      "\n",
      "data:  matrix(c(a, b, c, d), nrow = 2, ncol = 2)\n",
      "p-value = 0.05501\n",
      "alternative hypothesis: true odds ratio is not equal to 1\n",
      "95 percent confidence interval:\n",
      " 0.9837918 2.2436478\n",
      "sample estimates:\n",
      "odds ratio \n",
      "  1.494149 \n",
      "\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A data.frame: 2 × 2</caption>\n",
       "<thead>\n",
       "\t<tr><th scope=col>Gene_type</th><th scope=col>Proportion</th></tr>\n",
       "\t<tr><th scope=col>&lt;fct&gt;</th><th scope=col>&lt;dbl&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><td>cGenes</td><td>0.2875817</td></tr>\n",
       "\t<tr><td>eGenes</td><td>0.2126200</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.frame: 2 × 2\n",
       "\\begin{tabular}{ll}\n",
       " Gene\\_type & Proportion\\\\\n",
       " <fct> & <dbl>\\\\\n",
       "\\hline\n",
       "\t cGenes & 0.2875817\\\\\n",
       "\t eGenes & 0.2126200\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.frame: 2 × 2\n",
       "\n",
       "| Gene_type &lt;fct&gt; | Proportion &lt;dbl&gt; |\n",
       "|---|---|\n",
       "| cGenes | 0.2875817 |\n",
       "| eGenes | 0.2126200 |\n",
       "\n"
      ],
      "text/plain": [
       "  Gene_type Proportion\n",
       "1 cGenes    0.2875817 \n",
       "2 eGenes    0.2126200 "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<strong>png:</strong> 2"
      ],
      "text/latex": [
       "\\textbf{png:} 2"
      ],
      "text/markdown": [
       "**png:** 2"
      ],
      "text/plain": [
       "png \n",
       "  2 "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "### Plot pLI ###\n",
    "# Extract cgene_metric\n",
    "cgene_metric = gene_metrics[ensemble_id %in% cgenes, \"pLI\"]\n",
    "cgene_metric = na.omit(cgene_metric)\n",
    "cgene_metric <- unlist(cgene_metric)\n",
    "cgene_metric <- as.numeric(cgene_metric)\n",
    "  \n",
    "# Extract egene_metric\n",
    "egene_metric = gene_metrics[ensemble_id %in% egenes, \"pLI\"]\n",
    "egene_metric = na.omit(egene_metric)\n",
    "egene_metric <- unlist(egene_metric)\n",
    "egene_metric <- as.numeric(egene_metric)\n",
    "\n",
    "cgene_metric = ifelse(cgene_metric > 0.9, 1, 0)\n",
    "egene_metric = ifelse(egene_metric > 0.9, 1, 0)\n",
    "\n",
    "#Fisher exact test\n",
    "a = sum(cgene_metric == 1)\n",
    "b = sum(cgene_metric == 0)\n",
    "c = sum(egene_metric == 1)\n",
    "d = sum(egene_metric == 0)\n",
    "print(matrix(c(a,b,c,d),nrow=2,ncol=2))\n",
    "test = fisher.test(matrix(c(a,b,c,d),nrow=2,ncol=2))\n",
    "\n",
    "print(test)\n",
    "\n",
    "# Proportion of genes with pLI > 0.9\n",
    "data <- data.frame(\n",
    "  Gene_type = factor(c(\"cGenes\",\"eGenes\"), levels = c(\"cGenes\",\"eGenes\")),\n",
    "  Proportion = c(sum(cgene_metric == 1)/length(cgene_metric),sum(egene_metric == 1)/length(egene_metric)))\n",
    "data\n",
    "#png(\"plots/figure_plots/pli_plot.png\", width = 4, height = 6,units = \"in\", res = 300)\n",
    "svg(\"plots/figure_plots/pli_plot.svg\", width = 4, height = 6)\n",
    "\n",
    "pli_plot = ggplot(data, aes(x = Gene_type, y = Proportion, fill = Gene_type)) +\n",
    "  geom_bar(stat = \"identity\", color = \"black\", width = 0.5, position = position_dodge(width = 0.7)) +\n",
    "  labs(title = \"\", x = \"\", y = \"Proportion of genes \\n with pLi > 0.9\", color = \"Gene Type\") +\n",
    "  annotate(\"text\", x = 1.5, y = 0.35, label = paste0(\"P = \", formatC(test$p.value, format = \"e\", digits = 1)), size = 5, hjust = 0.5, family = \"sans\") +\n",
    "  scale_fill_manual(values = colors) +\n",
    "  theme_cowplot() +\n",
    "  theme(\n",
    "    axis.title.x = element_text(size = 16),\n",
    "    axis.title.y = element_text(size = 16, margin = margin(r = 10)),\n",
    "    axis.text.x = element_text(size = 16),\n",
    "    axis.text.y = element_text(size = 16),\n",
    "    legend.position = \"none\"\n",
    "  )\n",
    "\n",
    "pli_plot\n",
    "\n",
    "dev.off()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "id": "8119d5a6-c290-4736-ba0c-ff3c9c644541",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] 0.4385\n",
      "[1] 0.366\n",
      "\n",
      "\tWilcoxon rank sum test with continuity correction\n",
      "\n",
      "data:  cgene_metric and egene_metric\n",
      "W = 56708, p-value = 0.005675\n",
      "alternative hypothesis: true location shift is not equal to 0\n",
      "\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message:\n",
      "“\u001b[1m\u001b[22mRemoved 264 rows containing missing values (`geom_violin()`).”\n",
      "Warning message:\n",
      "“\u001b[1m\u001b[22mRemoved 1 rows containing missing values (`geom_text()`).”\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<strong>png:</strong> 2"
      ],
      "text/latex": [
       "\\textbf{png:} 2"
      ],
      "text/markdown": [
       "**png:** 2"
      ],
      "text/plain": [
       "png \n",
       "  2 "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "### Plot Episcore ###\n",
    "# Extract cgene_metric\n",
    "cgene_metric = gene_metrics[ensemble_id %in% cgenes, \"Episcore\"]\n",
    "cgene_metric = na.omit(cgene_metric)\n",
    "cgene_metric <- unlist(cgene_metric)\n",
    "cgene_metric <- as.numeric(cgene_metric)\n",
    "  \n",
    "# Extract egene_metric\n",
    "egene_metric = gene_metrics[ensemble_id %in% egenes, \"Episcore\"]\n",
    "egene_metric = na.omit(egene_metric)\n",
    "egene_metric <- unlist(egene_metric)\n",
    "egene_metric <- as.numeric(egene_metric)\n",
    "\n",
    "print(median(cgene_metric))\n",
    "print(median(egene_metric))\n",
    "\n",
    "test = wilcox.test(cgene_metric, egene_metric)\n",
    "print(test)\n",
    "\n",
    "  # Combine the data into a data frame for plotting\n",
    "combined_data <- data.frame(\n",
    "    Gene_type = c(rep(\"cGenes\", length(cgene_metric)), rep(\"eGenes\", length(egene_metric))),\n",
    "    Value = c(cgene_metric, egene_metric)\n",
    "  )\n",
    "\n",
    "  # Calculate the y-axis limits (just above the last outlier)\n",
    "stats_cgene <- boxplot.stats(cgene_metric)\n",
    "stats_egene <- boxplot.stats(egene_metric)\n",
    "  \n",
    "  # Max of the outliers for cgene and egene\n",
    "  # Check if there are outliers; if not, use the max values from the metrics\n",
    "if (length(stats_cgene$out) > 0 | length(stats_egene$out) > 0) {\n",
    "    max_outlier <- max(c(stats_cgene$out, stats_egene$out), na.rm = TRUE)\n",
    "} else {\n",
    "    max_outlier <- max(c(cgene_metric, egene_metric), na.rm = TRUE)\n",
    "}\n",
    "  \n",
    "# Set the lower limit as the minimum value and upper limit slightly above the max outlier\n",
    "min_value <- min(combined_data$Value, na.rm = TRUE)\n",
    "max_value <- max_outlier * 1.1  # 10% above the highest outlier\n",
    "\n",
    "#png(\"plots/figure_plots/enhancer_plot.png\", width = 4, height = 6,units = \"in\", res = 300)\n",
    "svg(\"plots/figure_plots/Haploinsuficiency_plot.svg\", width = 4, height = 6)\n",
    "\n",
    "enhancer_plot <- combined_data %>%\n",
    "  ggplot(aes(x = Gene_type, y = Value, fill = Gene_type)) +\n",
    "  geom_violin(trim = FALSE, fill = \"light grey\", color = \"black\", alpha = 0.5) +  # Violin plot\n",
    "  geom_boxplot(width = 0.1, color = \"black\", outlier.shape = NA) +  # Boxplot inside violin plot\n",
    "  scale_fill_manual(values = colors) +\n",
    "  theme_cowplot() +\n",
    "  annotate(\"text\", x = 1.5, y = 1.05, label = paste0(\"P = \", formatC(test$p.value, format = \"e\", digits = 1)), size = 5, hjust = 0.5, family = \"sans\") +\n",
    "  labs(title = \"\", x = \"\", y = \"Haploinsuficiency score\") +\n",
    "  ylim(0,1) +\n",
    "  theme(\n",
    "    axis.title.x = element_text(size = 16),\n",
    "    axis.title.y = element_text(size = 16, margin = margin(r = 10)),\n",
    "    axis.text.x = element_text(size = 16),\n",
    "    axis.text.y = element_text(size = 16),\n",
    "    legend.position = \"none\"\n",
    "  )\n",
    "enhancer_plot\n",
    "\n",
    "dev.off()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "id": "99642020-8941-4219-9488-3227df1096dd",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "18.0953"
      ],
      "text/latex": [
       "18.0953"
      ],
      "text/markdown": [
       "18.0953"
      ],
      "text/plain": [
       "[1] 18.0953"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "2.865715"
      ],
      "text/latex": [
       "2.865715"
      ],
      "text/markdown": [
       "2.865715"
      ],
      "text/plain": [
       "[1] 2.865715"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] 2.165588e-16\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message:\n",
      "“\u001b[1m\u001b[22mRemoved 193 rows containing missing values (`geom_violin()`).”\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<strong>png:</strong> 2"
      ],
      "text/latex": [
       "\\textbf{png:} 2"
      ],
      "text/markdown": [
       "**png:** 2"
      ],
      "text/plain": [
       "png \n",
       "  2 "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "### Expression violin plots ###\n",
    "K562_bulk = fread(\"/gpfs/commons/groups/lappalainen_lab/jmorris/210205_STINGseq-v2/data/K562_scRNAseq_bulkRNAseq.txt\") %>%\n",
    "  mutate(ensembl_gene_id = str_split_fixed(ENSG, \"\\\\.\",2)[,1])\n",
    "\n",
    "# Extract cgene_metric\n",
    "cgene_metric = K562_bulk$TPM[K562_bulk$ensembl_gene_id %in% cgenes & !K562_bulk$ensembl_gene_id %in% egenes]\n",
    "cgene_metric <- unlist(cgene_metric)\n",
    "cgene_metric <- as.numeric(cgene_metric)\n",
    "  \n",
    "# Extract egene_metric\n",
    "egene_df = data.frame(ensembl_gene_id = egenes) %>%\n",
    "  left_join(K562_bulk[,c(\"ensembl_gene_id\",\"TPM\")], \"ensembl_gene_id\")\n",
    "egene_df[is.na(egene_df)] = 0\n",
    "egene_metric <- unlist(egene_df$TPM[!egene_df$ensembl_gene_id %in% cgenes])\n",
    "egene_metric <- as.numeric(egene_metric)\n",
    "\n",
    "# Extract both metric\n",
    "  \n",
    "# Run Wilcoxon test\n",
    "median(cgene_metric)\n",
    "median(egene_metric)\n",
    "test = wilcox.test(cgene_metric, egene_metric)\n",
    "print(test$p.value)\n",
    "\n",
    "# Combine the data into a data frame for plotting\n",
    "combined_data <- data.frame(\n",
    "    Gene_type = c(rep(\"cGenes\", length(cgene_metric)), rep(\"eGenes\", length(egene_metric))),\n",
    "    Value = c(cgene_metric, egene_metric))\n",
    "\n",
    "#png(\"plots/figure_plots/k562_ex_plot.png\", width = 4, height = 6,units = \"in\", res = 300)\n",
    "svg(\"plots/figure_plots/k562_ex_plot.svg\", width = 4, height = 6)\n",
    "\n",
    "# K562 expression violin plot\n",
    "k562_ex_plot <- combined_data %>%\n",
    "  filter(Value<80) %>%\n",
    "  ggplot(aes(x = Gene_type, y = Value, fill = Gene_type)) +\n",
    "  geom_violin(trim = FALSE, fill = \"light grey\", color = \"black\", alpha = 0.5) +  # Violin plot\n",
    "  geom_boxplot(width = 0.1, color = \"black\", outlier.shape = NA) +  # Boxplot inside violin plot\n",
    "  scale_fill_manual(values = colors) +\n",
    "  theme_cowplot() +\n",
    "  ylim(0,80) +\n",
    "  annotate(\"text\", x = 1.5, y = 80, label = paste0(\"P = \", formatC(test$p.value, format = \"e\", digits = 1)), size = 5, hjust = 0.5, family = \"sans\") +\n",
    "  labs(title = \"\", x = \"\", y = \"K562 gene TPM\",  fill = \"Target genes\") +\n",
    "  theme(\n",
    "      plot.title = element_text(size = 16, face = \"plain\", margin = margin(b = 10), hjust = 0),\n",
    "      axis.title.y = element_text(size = 16, margin = margin(r = 10)),\n",
    "      axis.text.x = element_text(size = 16),\n",
    "      axis.text.y = element_text(size = 16),\n",
    "      legend.position = \"none\"\n",
    "  )\n",
    "\n",
    "k562_ex_plot\n",
    "\n",
    "dev.off()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "id": "09a0347b-def9-4de5-84f0-99b87dfba3e7",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "12.7508"
      ],
      "text/latex": [
       "12.7508"
      ],
      "text/markdown": [
       "12.7508"
      ],
      "text/plain": [
       "[1] 12.7508"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "5.03962"
      ],
      "text/latex": [
       "5.03962"
      ],
      "text/markdown": [
       "5.03962"
      ],
      "text/plain": [
       "[1] 5.03962"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "\tWilcoxon rank sum test with continuity correction\n",
      "\n",
      "data:  cgene_metric and egene_metric\n",
      "W = 45415, p-value = 0.001545\n",
      "alternative hypothesis: true location shift is not equal to 0\n",
      "\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message:\n",
      "“\u001b[1m\u001b[22mRemoved 189 rows containing missing values (`geom_violin()`).”\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<strong>png:</strong> 2"
      ],
      "text/latex": [
       "\\textbf{png:} 2"
      ],
      "text/markdown": [
       "**png:** 2"
      ],
      "text/plain": [
       "png \n",
       "  2 "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "### GTEx whole blood gene expression violin plot ###\n",
    "blood_tpm = fread(\"data/GTEx_Analysis_2017-06-05_v8_RNASeQCv1.1.9_gene_median_tpm.gct.gz\")[, c(\"Name\", \"Whole Blood\"), with = FALSE]\n",
    "blood_tpm$Name = str_split_fixed(blood_tpm$Name, \"\\\\.\", 2)[,1]\n",
    "\n",
    "# Extract cgene_metric\n",
    "cgene_metric = blood_tpm$`Whole Blood`[blood_tpm$Name %in% cgenes & !blood_tpm$Name %in% egenes]\n",
    "cgene_metric <- unlist(cgene_metric)\n",
    "cgene_metric <- as.numeric(cgene_metric)\n",
    "  \n",
    "# Extract egene_metric\n",
    "egene_metric = blood_tpm$`Whole Blood`[blood_tpm$Name %in% egenes & !blood_tpm$Name %in% cgenes]\n",
    "egene_metric <- unlist(egene_metric)\n",
    "egene_metric <- as.numeric(egene_metric)\n",
    "\n",
    "# Extract both metric\n",
    "  \n",
    "# Run Wilcoxon test\n",
    "median(cgene_metric)\n",
    "median(egene_metric)\n",
    "test = wilcox.test(cgene_metric, egene_metric)\n",
    "print(test)\n",
    "\n",
    "# Combine the data into a data frame for plotting\n",
    "combined_data <- data.frame(\n",
    "    Gene_type = c(rep(\"cGenes\", length(cgene_metric)), rep(\"eGenes\", length(egene_metric))),\n",
    "    Value = c(cgene_metric, egene_metric))\n",
    "\n",
    "#png(\"plots/figure_plots/blood_ex_plot.png\", width = 4, height = 6,units = \"in\", res = 300)\n",
    "svg(\"plots/figure_plots/blood_ex_plot.svg\", width = 4, height = 6)\n",
    "\n",
    "# blood expression violin plot\n",
    "blood_ex_plot <- combined_data %>%\n",
    "  filter(Value<80) %>%\n",
    "  ggplot(aes(x = Gene_type, y = Value, fill = Gene_type)) +\n",
    "  geom_violin(trim = FALSE, fill = \"light grey\", color = \"black\", alpha = 0.5) +  # Violin plot\n",
    "  geom_boxplot(width = 0.1, color = \"black\", outlier.shape = NA) +  # Boxplot inside violin plot\n",
    "  scale_fill_manual(values = colors) +\n",
    "  theme_cowplot() +\n",
    "  ylim(0,80) +\n",
    "  annotate(\"text\", x = 1.5, y = 80, label = paste0(\"P = \", formatC(test$p.value, format = \"e\", digits = 1)), size = 5, hjust = 0.5, family = \"sans\") +\n",
    "  labs(title = \"\", x = \"\", y = \"Blood gene TPM\",  fill = \"Target genes\") +\n",
    "  theme(\n",
    "      plot.title = element_text(size = 16, face = \"plain\", margin = margin(b = 10), hjust = 0),\n",
    "      axis.title.y = element_text(size = 16, margin = margin(r = 10)),\n",
    "      axis.text.x = element_text(size = 16),\n",
    "      axis.text.y = element_text(size = 16),\n",
    "      legend.position = \"none\"\n",
    "  )\n",
    "\n",
    "blood_ex_plot\n",
    "\n",
    "dev.off()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "74f1884c-43d7-4d6c-a9a7-0ee4bb37f096",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "\tPearson's Chi-squared test\n",
      "\n",
      "data:  data_matrix\n",
      "X-squared = 33.78, df = 2, p-value = 4.622e-08\n",
      "\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A data.frame: 6 × 3</caption>\n",
       "<thead>\n",
       "\t<tr><th scope=col>Rank</th><th scope=col>Gene_type</th><th scope=col>percent</th></tr>\n",
       "\t<tr><th scope=col>&lt;fct&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;dbl&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><td>1st </td><td>cGenes</td><td>0.8344828</td></tr>\n",
       "\t<tr><td>1st </td><td>eGenes</td><td>0.7184466</td></tr>\n",
       "\t<tr><td>2nd </td><td>cGenes</td><td>0.1379310</td></tr>\n",
       "\t<tr><td>2nd </td><td>eGenes</td><td>0.3417476</td></tr>\n",
       "\t<tr><td>&gt;2nd</td><td>cGenes</td><td>0.2344828</td></tr>\n",
       "\t<tr><td>&gt;2nd</td><td>eGenes</td><td>0.6155340</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.frame: 6 × 3\n",
       "\\begin{tabular}{lll}\n",
       " Rank & Gene\\_type & percent\\\\\n",
       " <fct> & <chr> & <dbl>\\\\\n",
       "\\hline\n",
       "\t 1st  & cGenes & 0.8344828\\\\\n",
       "\t 1st  & eGenes & 0.7184466\\\\\n",
       "\t 2nd  & cGenes & 0.1379310\\\\\n",
       "\t 2nd  & eGenes & 0.3417476\\\\\n",
       "\t >2nd & cGenes & 0.2344828\\\\\n",
       "\t >2nd & eGenes & 0.6155340\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.frame: 6 × 3\n",
       "\n",
       "| Rank &lt;fct&gt; | Gene_type &lt;chr&gt; | percent &lt;dbl&gt; |\n",
       "|---|---|---|\n",
       "| 1st  | cGenes | 0.8344828 |\n",
       "| 1st  | eGenes | 0.7184466 |\n",
       "| 2nd  | cGenes | 0.1379310 |\n",
       "| 2nd  | eGenes | 0.3417476 |\n",
       "| &gt;2nd | cGenes | 0.2344828 |\n",
       "| &gt;2nd | eGenes | 0.6155340 |\n",
       "\n"
      ],
      "text/plain": [
       "  Rank Gene_type percent  \n",
       "1 1st  cGenes    0.8344828\n",
       "2 1st  eGenes    0.7184466\n",
       "3 2nd  cGenes    0.1379310\n",
       "4 2nd  eGenes    0.3417476\n",
       "5 >2nd cGenes    0.2344828\n",
       "6 >2nd eGenes    0.6155340"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<strong>png:</strong> 2"
      ],
      "text/latex": [
       "\\textbf{png:} 2"
      ],
      "text/markdown": [
       "**png:** 2"
      ],
      "text/plain": [
       "png \n",
       "  2 "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "### Closest genes barplots ###\n",
    "# Calculate proportion of significant genes in each CRE\n",
    "closest_cgene = cres_w_grnas %>% distinct(target_gene, .keep_all=T) %>% # Remove duplicate target-genes for different eqtl/gwas\n",
    " group_by(grna_target) %>% mutate(dis_rank = rank(abs(tss_distance))) %>% filter(significant ==1)\n",
    "one_cgene = length(unique(closest_cgene$grna_target[closest_cgene$dis_rank ==1]))/length(unique(closest_cgene$grna_target))\n",
    "two_cgene = length(unique(closest_cgene$grna_target[closest_cgene$dis_rank ==2]))/length(unique(closest_cgene$grna_target))\n",
    "three_cgene = length(unique(closest_cgene$grna_target[closest_cgene$dis_rank > 2]))/length(unique(closest_cgene$grna_target))\n",
    "\n",
    "# Same for eqtl\n",
    "closest_egene = cres_w_grnas_egene %>% filter(is.na(dis_rank) ==F) %>% arrange(dis_rank) %>% distinct(target_gene, .keep_all=T)\n",
    "one_egene = length(unique(closest_egene$grna_target[closest_egene$dis_rank ==1]))/length(unique(closest_egene$grna_target))\n",
    "two_egene = length(unique(closest_egene$grna_target[closest_egene$dis_rank ==2]))/length(unique(closest_egene$grna_target))\n",
    "three_egene = length(unique(closest_egene$grna_target[closest_egene$dis_rank >2]))/length(unique(closest_egene$grna_target))\n",
    "\n",
    "a = length(unique(closest_cgene$grna_target[closest_cgene$dis_rank ==1]))\n",
    "b = length(unique(closest_cgene$grna_target[closest_cgene$dis_rank ==2]))\n",
    "c = length(unique(closest_cgene$grna_target[closest_cgene$dis_rank > 2]))\n",
    "e = length(unique(closest_egene$grna_target[closest_egene$dis_rank ==1]))\n",
    "f = length(unique(closest_egene$grna_target[closest_egene$dis_rank ==2]))\n",
    "g = length(unique(closest_egene$grna_target[closest_egene$dis_rank >2]))\n",
    "\n",
    "# construct matrix for chi-square\n",
    "data_matrix <- matrix(c(a, b, c, d, e, f), \n",
    "                      nrow = 2, \n",
    "                      byrow = TRUE)\n",
    "\n",
    "rownames(data_matrix) <- c(\"cGenes\", \"eGenes\")\n",
    "colnames(data_matrix) <- c(\"one\", \"two\", \"more_two\")\n",
    "\n",
    "# Run Chi-Squared Test\n",
    "chi_test <- chisq.test(data_matrix)\n",
    "\n",
    "# Print results\n",
    "print(chi_test)\n",
    "\n",
    "# Barplot of gene distances\n",
    "closest_data <- data.frame(\n",
    "  Rank = factor(c(\"1st\",\"1st\",\"2nd\",\"2nd\",\">2nd\", \">2nd\"), levels = c(\"1st\", \"2nd\", \">2nd\")),\n",
    "  Gene_type = c(rep(c(\"cGenes\", \"eGenes\"),3)),\n",
    "  percent = c(one_cgene,one_egene,two_cgene,two_egene,\n",
    "              three_cgene,three_egene))\n",
    "\n",
    "closest_data\n",
    "\n",
    "#png(\"plots/figure_plots/closest_barplots.png\", width = 4, height = 6,units = \"in\", res = 300)\n",
    "svg(\"plots/figure_plots/closest_barplots.svg\", width = 4, height = 6)\n",
    "\n",
    "# Generate a bar plot for gene distances\n",
    "closest_barplots = ggplot(closest_data, aes(x = Rank, y = percent, fill = Gene_type)) +\n",
    "  geom_bar(stat = \"identity\", color = \"black\", width = 0.5, position = position_dodge(width = 0.7)) +\n",
    "  theme_minimal() +\n",
    "  scale_fill_manual(values = colors) +\n",
    "  labs(title = \"\", x = \"Gene distance rank\", y = \"Proportion of total CREs\") +\n",
    "  annotate(\"text\", x = 2.4, y = 0.80, label = paste0(\"P = \",formatC(chi_test$p.value, format = \"e\", digits = 1)), size = 5, hjust = 0.5) +\n",
    "  theme_cowplot() +\n",
    "  theme(\n",
    "    axis.title.x = element_text(size = 16),\n",
    "    axis.title.y = element_text(size = 16, margin = margin(r=10)),\n",
    "    axis.text.x = element_text(size = 16),\n",
    "    axis.text.y = element_text(size = 16),\n",
    "    legend.position = \"none\"\n",
    "  )\n",
    "closest_barplots\n",
    "\n",
    "dev.off()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "id": "32f4739c-9b97-488b-80dc-cfab771d5aaa",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "\tPearson's Chi-squared test\n",
      "\n",
      "data:  data_matrix\n",
      "X-squared = 93.163, df = 2, p-value < 2.2e-16\n",
      "\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A data.frame: 6 × 3</caption>\n",
       "<thead>\n",
       "\t<tr><th scope=col>n_Genes</th><th scope=col>Gene_type</th><th scope=col>Proportion</th></tr>\n",
       "\t<tr><th scope=col>&lt;fct&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;dbl&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><td>1 </td><td>cGenes</td><td>0.82068966</td></tr>\n",
       "\t<tr><td>1 </td><td>eGenes</td><td>0.37669903</td></tr>\n",
       "\t<tr><td>2 </td><td>cGenes</td><td>0.13103448</td></tr>\n",
       "\t<tr><td>2 </td><td>eGenes</td><td>0.26407767</td></tr>\n",
       "\t<tr><td>2+</td><td>cGenes</td><td>0.04827586</td></tr>\n",
       "\t<tr><td>2+</td><td>eGenes</td><td>0.35922330</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.frame: 6 × 3\n",
       "\\begin{tabular}{lll}\n",
       " n\\_Genes & Gene\\_type & Proportion\\\\\n",
       " <fct> & <chr> & <dbl>\\\\\n",
       "\\hline\n",
       "\t 1  & cGenes & 0.82068966\\\\\n",
       "\t 1  & eGenes & 0.37669903\\\\\n",
       "\t 2  & cGenes & 0.13103448\\\\\n",
       "\t 2  & eGenes & 0.26407767\\\\\n",
       "\t 2+ & cGenes & 0.04827586\\\\\n",
       "\t 2+ & eGenes & 0.35922330\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.frame: 6 × 3\n",
       "\n",
       "| n_Genes &lt;fct&gt; | Gene_type &lt;chr&gt; | Proportion &lt;dbl&gt; |\n",
       "|---|---|---|\n",
       "| 1  | cGenes | 0.82068966 |\n",
       "| 1  | eGenes | 0.37669903 |\n",
       "| 2  | cGenes | 0.13103448 |\n",
       "| 2  | eGenes | 0.26407767 |\n",
       "| 2+ | cGenes | 0.04827586 |\n",
       "| 2+ | eGenes | 0.35922330 |\n",
       "\n"
      ],
      "text/plain": [
       "  n_Genes Gene_type Proportion\n",
       "1 1       cGenes    0.82068966\n",
       "2 1       eGenes    0.37669903\n",
       "3 2       cGenes    0.13103448\n",
       "4 2       eGenes    0.26407767\n",
       "5 2+      cGenes    0.04827586\n",
       "6 2+      eGenes    0.35922330"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<strong>png:</strong> 2"
      ],
      "text/latex": [
       "\\textbf{png:} 2"
      ],
      "text/markdown": [
       "**png:** 2"
      ],
      "text/plain": [
       "png \n",
       "  2 "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "### Number of genes barplots ###\n",
    "cgenes_per_cre = cres_w_grnas %>% filter(significant ==1) %>% distinct(target_gene, .keep_all=T) %>% group_by(grna_target) %>% \n",
    "    summarise(n = n())\n",
    "egenes_per_cre = cres_w_grnas_egene %>% distinct(target_gene, .keep_all=T) %>% group_by(grna_target) %>% \n",
    "    summarise(n = n())\n",
    "\n",
    "# Chi-squared test\n",
    "a = sum(cgenes_per_cre$n ==1)\n",
    "b = sum(cgenes_per_cre$n ==2)\n",
    "c = sum(cgenes_per_cre$n > 2)\n",
    "d = sum(egenes_per_cre$n ==1)\n",
    "e = sum(egenes_per_cre$n ==2)\n",
    "f = sum(egenes_per_cre$n > 2)\n",
    "\n",
    "# construct matrix for chi-square\n",
    "data_matrix <- matrix(c(a, b, c, d, e, f), \n",
    "                      nrow = 2, \n",
    "                      byrow = TRUE)\n",
    "\n",
    "rownames(data_matrix) <- c(\"cgenes\", \"egenes\")\n",
    "colnames(data_matrix) <- c(\"one\", \"two\", \"more_two\")\n",
    "\n",
    "# Run Chi-Squared Test\n",
    "chi_test <- chisq.test(data_matrix)\n",
    "\n",
    "# Print results\n",
    "print(chi_test)\n",
    "\n",
    "# Barplot of gene distances\n",
    "num_genes_data <- data.frame(\n",
    "  n_Genes = factor(c(\"1\",\"1\",\"2\",\"2\",\"2+\", \"2+\"), levels = c(\"1\", \"2\", \"2+\")),\n",
    "  Gene_type = c(rep(c(\"cGenes\", \"eGenes\"),3)),\n",
    "  Proportion = c(sum(cgenes_per_cre$n ==1)/nrow(cgenes_per_cre),\n",
    "              sum(egenes_per_cre$n ==1)/nrow(egenes_per_cre),\n",
    "              sum(cgenes_per_cre$n ==2)/nrow(cgenes_per_cre),\n",
    "              sum(egenes_per_cre$n ==2)/nrow(egenes_per_cre),\n",
    "              sum(cgenes_per_cre$n > 2)/nrow(cgenes_per_cre),\n",
    "              sum(egenes_per_cre$n > 2)/nrow(egenes_per_cre)))\n",
    "\n",
    "num_genes_data\n",
    "\n",
    "# Generate a bar plot for gene distances\n",
    "#png(\"plots/figure_plots/num_genes_barplot.png\", width = 4, height = 6,units = \"in\", res = 300)\n",
    "svg(\"plots/figure_plots/num_genes_barplot.svg\", width = 4, height = 6)\n",
    "\n",
    "num_genes_barplot = ggplot(num_genes_data, aes(x = n_Genes, y = Proportion, fill = Gene_type)) +\n",
    "  geom_bar(stat = \"identity\", color = \"black\", width = 0.5, position = position_dodge(width = 0.7)) +\n",
    "  theme_minimal() +\n",
    "  labs(title = \"\", x = \"Number of genes\", y = \"Proportion of total CREs\") +\n",
    "  annotate(\"text\", x = 2.4, y = 0.80, label = paste0(\"P = \",formatC(chi_test$p.value, format = \"e\", digits = 1)), size = 5, hjust = 0.5) +\n",
    "  scale_fill_manual(values = colors) +\n",
    "  theme_cowplot() +\n",
    "  theme(\n",
    "    axis.title.x = element_text(size = 16),\n",
    "    axis.title.y = element_text(size = 16, margin = margin(r=10)),\n",
    "    axis.text.x = element_text(size = 16),\n",
    "    axis.text.y = element_text(size = 16),\n",
    "    legend.position = \"none\"\n",
    "  )\n",
    "num_genes_barplot\n",
    "dev.off()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "id": "628bc44c-d3f2-4bae-b521-4a2e3c1b3cf4",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<strong>png:</strong> 2"
      ],
      "text/latex": [
       "\\textbf{png:} 2"
      ],
      "text/markdown": [
       "**png:** 2"
      ],
      "text/plain": [
       "png \n",
       "  2 "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "###################\n",
    "### Upset plot ####\n",
    "###################\n",
    "library(UpSetR)\n",
    "\n",
    "# Define the sc_qtls\n",
    "sc_qtls <- c(\"NK_cells\", \"B_cells\", \"CD4_T_cells\", \"CD8_T_cells\", \"DC_mean\", \"Mono_cells\", \"Other_cells\", \"Other_T_cells\")\n",
    "\n",
    "overlapping_eqtls <- cres_w_grnas_egene %>%\n",
    "  filter(target_gene %in% overlapping_cres_cgene)\n",
    "\n",
    "# Step 1: Data Preparation\n",
    "venn_df <- overlapping_eqtls %>%\n",
    "  mutate(eqtl = ifelse(eqtl == \"GTEx\", \"blood\", eqtl)) %>%\n",
    "  mutate(group = ifelse(eqtl %in% c(\"blood\", \"GTEx\"), \"Bulk tissue eQTL\", \n",
    "                        ifelse(eqtl %in% sc_qtls, \"Sc eQTL\", \"Bulk cell type eQTL\")),\n",
    "         group = factor(group, levels = unique(group))) %>%\n",
    "  mutate(uniq_id = paste(group, ensembl_id, grna_target, sep = \"_\")) %>% \n",
    "  distinct(uniq_id, .keep_all = TRUE)  # Remove multiple GWAS per gene and eQTL\n",
    "\n",
    "# Step 2: Create a binary matrix for the UpSet plot\n",
    "upset_data <- venn_df %>%\n",
    "  select(target_gene, group) %>%\n",
    "  distinct() %>%\n",
    "  mutate(value = 1) %>%  # Mark presence with a value of 1\n",
    "  spread(key = group, value = value, fill = 0)  # Create a wide format with binary values\n",
    "\n",
    "# Convert to a matrix format\n",
    "upset_matrix <- as.data.frame(upset_data)\n",
    "upset_matrix <- upset_matrix[, -1]  # Remove target_gene column, as we just need binary columns\n",
    "\n",
    "#png(filename = \"plots/figure_plots/upset_plot.png\", width = 10, height = 8, units = \"in\", res = 300)\n",
    "svg(filename = \"plots/figure_plots/upset_plot.svg\", width = 10, height = 8)\n",
    "\n",
    "# Step 3: Create the UpSet Plot\n",
    "upset_plot = upset(upset_matrix, \n",
    "      sets = c(\"Bulk tissue eQTL\", \"Sc eQTL\", \"Bulk cell type eQTL\"), \n",
    "      order.by = \"freq\", \n",
    "      decreasing = T,\n",
    "      keep.order = FALSE, \n",
    "      main.bar.color = \"black\", \n",
    "      sets.bar.color = rev(c(\"#01a2d9\",\"#6794a7\",\"#014d64\")),\n",
    "      text.scale = c(3, 3, 2, 2, 3.3, 3),\n",
    "      mainbar.y.label = \"Number of intersecting \\n eGenes with cGenes\",\n",
    "      sets.x.label = \"Total \",\n",
    "      query.legend = \"bottom\",\n",
    "    )\n",
    "upset_plot\n",
    "\n",
    "dev.off()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "a54e627f-ced4-4005-8d75-3dc4c9b7ac5a",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "\tWilcoxon rank sum test with continuity correction\n",
      "\n",
      "data:  plot_df$minor_AF[plot_df$unique_CREs == \"cGenes\"] and plot_df$minor_AF[plot_df$unique_CREs == \"eGenes\"]\n",
      "W = 7729, p-value = 0.02902\n",
      "alternative hypothesis: true location shift is not equal to 0\n",
      "\n"
     ]
    }
   ],
   "source": [
    "#################################################\n",
    "### Make boxplot of MAFs of intersecting snps ###\n",
    "#################################################\n",
    "# Load targeted GWAS variants\n",
    "cres_w_grnas = fread(\"cres_with_grnas.txt\")\n",
    "# flip alleles to match gwas snp ids\n",
    "temp = gsub(\"_\", \":\", cres_w_grnas$finemap_snp_intersect_grna)\n",
    "temp = str_split_fixed(temp,\":\",4)[,2]\n",
    "cres_w_grnas$snp_pos = paste0(cres_w_grnas$chr, \":\", temp)\n",
    "# Load sumstats to obtain MAFs\n",
    "# Load GWAS summary stats\n",
    "sumstats = fread(\"/gpfs/commons/groups/lappalainen_lab/sghatan/stingseq_eqtl_overlap/data/UKBB_sumstats/30000_formatted.tsv\")\n",
    "sumstats$snp_pos = paste0(sumstats$Chr,\":\",sumstats$Pos)\n",
    "sumstats_filtered = sumstats %>% filter(snp_pos %in% cres_w_grnas$snp_pos)\n",
    "\n",
    "cres_w_grnas_egene = fread(\"cres_with_grna_eqtls.txt\")\n",
    "\n",
    "grna_cres_w_cgenes = unique(cres_w_grnas$grna_target[cres_w_grnas$significant == 1])\n",
    "grna_cres_w_egenes = unique(cres_w_grnas_egene$grna_target)\n",
    "overlapping_cres = grna_cres_w_cgenes[grna_cres_w_cgenes %in% grna_cres_w_egenes]\n",
    "\n",
    "plot_df = cres_w_grnas %>%\n",
    "  filter(grna_target %in% c(grna_cres_w_cgenes, grna_cres_w_egenes) & !grna_target %in% overlapping_cres) %>%\n",
    "  distinct(grna_target, .keep_all=T) %>%\n",
    "  mutate(unique_CREs = ifelse(grna_target %in% grna_cres_w_cgenes, \"cGenes\",\"eGenes\")) %>%\n",
    "  left_join(sumstats_filtered[,c(\"snp_pos\",\"minor_AF\")], \"snp_pos\")\n",
    "\n",
    "test = wilcox.test(plot_df$minor_AF[plot_df$unique_CREs == \"cGenes\"], plot_df$minor_AF[plot_df$unique_CREs == \"eGenes\"])\n",
    "print(test)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "376743d7-05b8-449a-8df0-9cdc112b0c00",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "48"
      ],
      "text/latex": [
       "48"
      ],
      "text/markdown": [
       "48"
      ],
      "text/plain": [
       "[1] 48"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "418"
      ],
      "text/latex": [
       "418"
      ],
      "text/markdown": [
       "418"
      ],
      "text/plain": [
       "[1] 418"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "length(plot_df$minor_AF[plot_df$unique_CREs == \"cgenes\"])\n",
    "length(plot_df$minor_AF[plot_df$unique_CREs == \"egenes\"])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "9ab1f3e6-882e-4293-8b4e-7363c8bb41d7",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message:\n",
      "“\u001b[1m\u001b[22mRemoved 187 rows containing missing values (`geom_violin()`).”\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<strong>png:</strong> 2"
      ],
      "text/latex": [
       "\\textbf{png:} 2"
      ],
      "text/markdown": [
       "**png:** 2"
      ],
      "text/plain": [
       "png \n",
       "  2 "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "#png(\"plots/figure_plots/boxplot_maf.png\", width = 4, height = 6,units = \"in\", res = 300)\n",
    "svg(\"plots/figure_plots/boxplot_maf.svg\", width = 4, height = 6)\n",
    "\n",
    "boxplot_maf = plot_df %>%\n",
    "  filter(minor_AF < 0.40) %>%\n",
    "  ggplot(aes(x = unique_CREs, y = minor_AF, fill = unique_CREs)) +\n",
    "  geom_violin(trim = FALSE, fill = \"light grey\", color = \"black\", alpha = 0.5) +  # Violin plot\n",
    "  geom_boxplot(width = 0.1, color = \"black\", outlier.shape = NA) +  # Boxplot inside violin plot\n",
    "  scale_fill_manual(values = colors) +\n",
    "  theme_cowplot() +\n",
    "  ylim(0,0.6) +\n",
    "  annotate(\"text\", x = 1.5, y = 0.6, label = paste0(\"P = \",formatC(test$p.value, format = \"e\", digits = 1)), size = 5, hjust = 0.5) +\n",
    "  labs(title = \"\", x = \"\", y = \"Minor allele frequency\", fill = \"Gene type\") +\n",
    "  theme(\n",
    "      plot.title = element_text(size = 16, face = \"plain\", margin = margin(b = 10), hjust = 0),\n",
    "      axis.title.y = element_text(size = 16, margin = margin(r = 10)),\n",
    "      axis.text.x = element_text(size = 16),\n",
    "      axis.text.y = element_text(size = 16),\n",
    "      legend.position = \"none\"\n",
    "  )\n",
    "boxplot_maf\n",
    "\n",
    "dev.off()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9e980382-97cd-4fcb-95c6-128acabc2924",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "id": "e1659778-8483-4188-8157-d561b8ad1776",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A data.frame: 2 × 2</caption>\n",
       "<thead>\n",
       "\t<tr><th scope=col>Gene_type</th><th scope=col>Proportion</th></tr>\n",
       "\t<tr><th scope=col>&lt;fct&gt;</th><th scope=col>&lt;dbl&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><td>cGenes</td><td>0.2149533</td></tr>\n",
       "\t<tr><td>eGenes</td><td>0.1125265</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.frame: 2 × 2\n",
       "\\begin{tabular}{ll}\n",
       " Gene\\_type & Proportion\\\\\n",
       " <fct> & <dbl>\\\\\n",
       "\\hline\n",
       "\t cGenes & 0.2149533\\\\\n",
       "\t eGenes & 0.1125265\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.frame: 2 × 2\n",
       "\n",
       "| Gene_type &lt;fct&gt; | Proportion &lt;dbl&gt; |\n",
       "|---|---|\n",
       "| cGenes | 0.2149533 |\n",
       "| eGenes | 0.1125265 |\n",
       "\n"
      ],
      "text/plain": [
       "  Gene_type Proportion\n",
       "1 cGenes    0.2149533 \n",
       "2 eGenes    0.1125265 "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<strong>png:</strong> 2"
      ],
      "text/latex": [
       "\\textbf{png:} 2"
      ],
      "text/markdown": [
       "**png:** 2"
      ],
      "text/plain": [
       "png \n",
       "  2 "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "#### LOOK AT MENDELIAN ENRICHMENT FILE FOR how numbers for PLOT are generated ####\n",
    "\n",
    "\n",
    "data <- data.frame(\n",
    "  Gene_type = factor(c(\"cGenes\",\"eGenes\"), levels = c(\"cGenes\",\"eGenes\")),\n",
    "  Proportion = c(23/(23+84),53/(53+418)))\n",
    "\n",
    "data\n",
    "\n",
    "#png(\"plots/figure_plots/burden_gene_barplot.png\", width = 4, height = 6,units = \"in\", res = 300)\n",
    "svg(\"plots/figure_plots/burden_gene_barplot.svg\", width = 4, height = 6)\n",
    "\n",
    "burden_gene_barplot = ggplot(data, aes(x = Gene_type, y = Proportion, fill = Gene_type)) +\n",
    "  geom_bar(stat = \"identity\", color = \"black\", width = 0.5, position = position_dodge(width = 0.7)) +\n",
    "  theme_minimal() +\n",
    "  labs(title = \"\", x = \"\", y = \"Proportion of genes classified \\n as gold-standard genes\") +\n",
    "  annotate(\"text\", x = 1.5, y = 0.24, label = paste0(\"P = \", \"6.8e-03\"), size = 5, hjust = 0.5, family = \"sans\") +\n",
    "  scale_fill_manual(values = colors) +\n",
    "  theme_cowplot() +\n",
    "  theme(\n",
    "    axis.title.x = element_text(size = 16),\n",
    "    axis.title.y = element_text(size = 16, margin = margin(r = 10)),\n",
    "    axis.text.x = element_text(size = 16),\n",
    "    axis.text.y = element_text(size = 16),\n",
    "    legend.position = \"none\"\n",
    "  )\n",
    "\n",
    "burden_gene_barplot\n",
    "\n",
    "dev.off()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "id": "e848bd09-341b-4224-9aaa-6b46951e7aec",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "\tWilcoxon rank sum test with continuity correction\n",
      "\n",
      "data:  pp.h4.df$mean_PP.H4[pp.h4.df$gene_intersect == 1] and pp.h4.df$mean_PP.H4[pp.h4.df$gene_intersect == 0]\n",
      "W = 57098, p-value = 1.434e-06\n",
      "alternative hypothesis: true location shift is not equal to 0\n",
      "\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message:\n",
      "“\u001b[1m\u001b[22mRemoved 304 rows containing missing values (`geom_violin()`).”\n",
      "Warning message:\n",
      "“\u001b[1m\u001b[22mRemoved 1 rows containing missing values (`geom_text()`).”\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<strong>png:</strong> 2"
      ],
      "text/latex": [
       "\\textbf{png:} 2"
      ],
      "text/markdown": [
       "**png:** 2"
      ],
      "text/plain": [
       "png \n",
       "  2 "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "###\n",
    "pp.h4.df = cres_w_grnas_egene %>%\n",
    "  mutate(gene_intersect = as.factor(ifelse(target_gene %in% overlapping_target_genes,1,0))) %>%\n",
    "  group_by(target_gene) %>% mutate(mean_PP.H4 = mean(PP.H4.abf,na.rm=T)) %>% # Take the average PP.H4 from multiple GWAS\n",
    "  mutate(scaled_mean_PP.H4 = mean_PP.H4*2) %>% # scale pp.h4 so that 0.5 is one unit change\n",
    "  distinct(target_gene,.keep_all=T) %>%\n",
    "  ungroup()\n",
    "\n",
    "test = wilcox.test(pp.h4.df$mean_PP.H4[pp.h4.df$gene_intersect==1], pp.h4.df$mean_PP.H4[pp.h4.df$gene_intersect == 0])\n",
    "print(test)\n",
    "\n",
    "pp.h4.plot = pp.h4.df %>% \n",
    "  mutate(gene_intersect = ifelse(gene_intersect == 1, \"eQTL with\\ncGene\", \"eQTL without\\ncGene\")) %>%\n",
    "  ggplot(aes(x = gene_intersect, y = mean_PP.H4, fill = gene_intersect)) +\n",
    "  geom_violin(trim = FALSE, fill = \"light grey\", color = \"black\", alpha = 0.5) +  # Violin plot\n",
    "  geom_boxplot(width = 0.1, color = \"black\", outlier.shape = NA) +  # Boxplot inside violin plot\n",
    "  theme_cowplot() +\n",
    "  annotate(\"text\", x = 1.5, y = 0.2, label = paste0(\"P = \", formatC(test$p.value, format = \"e\", digits = 1)), size = 5, hjust = 0.5, family = \"sans\") +\n",
    "  ylim(0.5, 1) +\n",
    "  xlab(\"\") +\n",
    "  ylab(\"Colocalization probability\") +\n",
    "  scale_fill_manual(name = \"\",\n",
    "                    values = c(\"eQTL without\\ncGene\" = \"#76c0c1\", \"eQTL with\\ncGene\" = \"#8e6c8a\"),\n",
    "                    labels = c(\"eQTL without cGene\", \"eQTL with cGene\")\n",
    "                   ) +  # Define custom colors\n",
    "  theme(\n",
    "    plot.title = element_text(size = 16, face = \"plain\", margin = margin(b = 10), hjust = 0),\n",
    "      axis.title.y = element_text(size = 16, margin = margin(r = 10)),\n",
    "      axis.text.x = element_text(size = 16),\n",
    "      axis.text.y = element_text(size = 16),\n",
    "      legend.position = \"none\"\n",
    "  )\n",
    "#png(\"plots/figure_plots/pp.h4_violin_plot.png\", width = 4, height = 6, units = \"in\", res = 300)\n",
    "svg(\"plots/figure_plots/pp.h4_violin_plot.svg\", width = 4, height = 6)\n",
    "pp.h4.plot\n",
    "dev.off()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "id": "025fa007-fef4-4c22-945f-1092e48251f3",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "     [,1] [,2]\n",
      "[1,]   56   12\n",
      "[2,]   97   67\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<strong>png:</strong> 2"
      ],
      "text/latex": [
       "\\textbf{png:} 2"
      ],
      "text/markdown": [
       "**png:** 2"
      ],
      "text/plain": [
       "png \n",
       "  2 "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "     [,1] [,2]\n",
      "[1,]   65    3\n",
      "[2,]  220    7\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<strong>png:</strong> 2"
      ],
      "text/latex": [
       "\\textbf{png:} 2"
      ],
      "text/markdown": [
       "**png:** 2"
      ],
      "text/plain": [
       "png \n",
       "  2 "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "cres_w_grnas = fread(\"cres_with_grnas.txt\") %>%\n",
    "  mutate(target_gene = paste0(grna_target,\"_\",ensembl_id))\n",
    "cres_w_grnas_egene = fread(\"cres_with_grna_eqtls.txt\") %>%\n",
    "  mutate(target_gene = paste0(grna_target,\"_\",ensembl_id))\n",
    "\n",
    "total_cres = unique(cres_w_grnas$grna_target)\n",
    "grna_cres_w_cgenes = unique(cres_w_grnas$grna_target[cres_w_grnas$significant == 1])\n",
    "target_cgenes = unique(cres_w_grnas$target_gene[cres_w_grnas$significant == 1])\n",
    "cgenes = unique(cres_w_grnas$ensembl_id[cres_w_grnas$significant == 1])\n",
    "grna_cres_w_egenes = unique(cres_w_grnas_egene$grna_target)\n",
    "target_egenes = unique(cres_w_grnas_egene$target_gene)\n",
    "egenes = unique(cres_w_grnas_egene$ensembl_id)\n",
    "cres_no_target = unique(cres_w_grnas$grna_target[!cres_w_grnas$grna_target %in% c(grna_cres_w_cgenes,grna_cres_w_egenes)])\n",
    "overlapping_cres = grna_cres_w_cgenes[grna_cres_w_cgenes %in% grna_cres_w_egenes]\n",
    "overlapping_cres_cgene = unique(cres_w_grnas$target_gene[cres_w_grnas$grna_target %in% overlapping_cres & cres_w_grnas$significant == 1])\n",
    "overlapping_cres_egene = unique(cres_w_grnas_egene$target_gene[cres_w_grnas_egene$grna_target %in% overlapping_cres])\n",
    "overlapping_target_genes = overlapping_cres_cgene[overlapping_cres_cgene %in% overlapping_cres_egene]\n",
    "\n",
    "# Load power\n",
    "crispr.power = fread(\"cres_with_grnas_power.txt\") %>%\n",
    "  mutate(target_gene = paste0(grna_target,\"_\",ensembl_id)) %>%\n",
    "  filter(effect_size == 0.85)\n",
    "\n",
    "eqtl.power = fread(\"power_results/eqtl.power.combined.txt\") %>%\n",
    "  mutate(target_gene = paste0(grna_target,\"_\",ensembl_id)) \n",
    "\n",
    "# Filter to overlapping significant CREs\n",
    "crispr.power.overlap = crispr.power %>%\n",
    "  distinct(target_gene, .keep_all=T) %>%\n",
    "  filter(target_gene %in% c(overlapping_cres_egene,overlapping_cres_cgene)) %>%\n",
    "  mutate(intersect = as.factor(ifelse(target_gene %in% overlapping_target_genes, \"Intersection\", \"No intersection\")))\n",
    "\n",
    "#Fisher exact test\n",
    "a = sum(crispr.power.overlap$power > 0.79 & crispr.power.overlap$intersect == \"Intersection\")\n",
    "b = sum(crispr.power.overlap$power > 0.79 & crispr.power.overlap$intersect == \"No intersection\")\n",
    "c = sum(crispr.power.overlap$power < 0.8 & crispr.power.overlap$intersect == \"Intersection\")\n",
    "d = sum(crispr.power.overlap$power < 0.8 & crispr.power.overlap$intersect == \"No intersection\")\n",
    "print(matrix(c(a,b,c,d),nrow=2,ncol=2))\n",
    "test = fisher.test(matrix(c(a,b,c,d),nrow=2,ncol=2))\n",
    "\n",
    "# Proportion of genes with p > 0.8\n",
    "data <- data.frame(\n",
    "  Gene_type = factor(c(\"Intersection\", \"No \\n intersection\"), levels = c(\"Intersection\", \"No \\n intersection\")),\n",
    "  Proportion = c(nrow(crispr.power.overlap[power > 0.79 & intersect == \"Intersection\",])/nrow(crispr.power.overlap[intersect == \"Intersection\",]),\n",
    "                 nrow(crispr.power.overlap[power > 0.79 & intersect == \"No intersection\",])/nrow(crispr.power.overlap[intersect == \"No intersection\",])))\n",
    "\n",
    "svg(\"plots/figure_plots/crispri_power_barplot.svg\", width = 4, height = 6)\n",
    "\n",
    "ggplot(data, aes(x = Gene_type, y = Proportion)) +\n",
    "  geom_bar(stat = \"identity\", color = \"black\", fill = \"#d35e60\", width = 0.5, position = position_dodge(width = 0.7)) +\n",
    "  labs(title = \"\", x = \"\", y = \"Proportion of CRE-genes \\n with CRISPRi power ≥ 0.80\", color = \"Gene Type\") +\n",
    "  annotate(\"text\", x = 1.5, y = 0.9, label = paste0(\"P = \", formatC(test$p.value, format = \"e\", digits = 1)), size = 5, hjust = 0.5, family = \"sans\") +\n",
    "  theme_cowplot() +\n",
    "  theme(\n",
    "    axis.title.x = element_text(size = 16),\n",
    "    axis.title.y = element_text(size = 16, margin = margin(r = 10)),\n",
    "    axis.text.x = element_text(size = 16),\n",
    "    axis.text.y = element_text(size = 16),\n",
    "    legend.position = \"none\"\n",
    "  )\n",
    "\n",
    "dev.off()\n",
    "\n",
    "eqtl.power.overlap = eqtl.power %>%\n",
    "  distinct(target_gene, .keep_all=T) %>%\n",
    "  filter(target_gene %in% c(overlapping_cres_egene,overlapping_cres_cgene)) %>%\n",
    "  mutate(intersect = as.factor(ifelse(target_gene %in% overlapping_target_genes, \"Intersection\", \"No intersection\")))\n",
    "\n",
    "#Fisher exact test\n",
    "a = sum(eqtl.power.overlap$eQTL.power > 0.79 & eqtl.power.overlap$intersect == \"Intersection\")\n",
    "b = sum(eqtl.power.overlap$eQTL.power > 0.79 & eqtl.power.overlap$intersect == \"No intersection\")\n",
    "c = sum(eqtl.power.overlap$eQTL.power < 0.8 & eqtl.power.overlap$intersect == \"Intersection\")\n",
    "d = sum(eqtl.power.overlap$eQTL.power < 0.8 & eqtl.power.overlap$intersect == \"No intersection\")\n",
    "print(matrix(c(a,b,c,d),nrow=2,ncol=2))\n",
    "test = fisher.test(matrix(c(a,b,c,d),nrow=2,ncol=2))\n",
    "\n",
    "# Proportion of genes with p > 0.8\n",
    "data <- data.frame(\n",
    "  Gene_type = factor(c(\"Intersection\", \"No \\n intersection\"), levels = c(\"Intersection\", \"No \\n intersection\")),\n",
    "  Proportion = c(nrow(eqtl.power.overlap[eQTL.power > 0.79 & intersect == \"Intersection\",])/nrow(eqtl.power.overlap[intersect == \"Intersection\",]),\n",
    "                 nrow(eqtl.power.overlap[eQTL.power > 0.79 & intersect == \"No intersection\",])/nrow(eqtl.power.overlap[intersect == \"No intersection\",])))\n",
    "\n",
    "svg(\"plots/figure_plots/eqtl_power_barplot.svg\", width = 4, height = 6)\n",
    "\n",
    "ggplot(data, aes(x = Gene_type, y = Proportion)) +\n",
    "  geom_bar(stat = \"identity\", color = \"black\", fill = \"#76c0c1\", width = 0.5, position = position_dodge(width = 0.7)) +\n",
    "  labs(title = \"\", x = \"\", y = \"Proportion of CRE-genes \\n with eQTL power ≥ 0.80\") +\n",
    "  annotate(\"text\", x = 1.5, y = 1.02, label = paste0(\"P = \", formatC(test$p.value, format = \"e\", digits = 1)), size = 5, hjust = 0.5, family = \"sans\") +\n",
    "  theme_cowplot() +\n",
    "  theme(\n",
    "    axis.title.x = element_text(size = 16),\n",
    "    axis.title.y = element_text(size = 16, margin = margin(r = 10)),\n",
    "    axis.text.x = element_text(size = 16),\n",
    "    axis.text.y = element_text(size = 16),\n",
    "    legend.position = \"none\"\n",
    "  )\n",
    "\n",
    "dev.off()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 57,
   "id": "c121a547-15a1-4c8a-bd8f-4dbcbc7559c1",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAA0gAAANICAIAAAByhViMAAAACXBIWXMAABJ0AAASdAHeZh94\nAAAgAElEQVR4nOzdd2Bb1d3/8XOvhuW9Ha9475Ud9iobyiq0bGiB0tIBLU8ZhcIDpVCgfWih\nv5Yuyh4F2rIpYUMghJDlON7bTrz30rz394fiEBLZsRPb1zp+v/6SJVn6ytb46NxzvkfRdV0A\nAADA/6lGFwAAAICZQbADAACQBMEOAABAEgQ7AAAASRDsAAAAJEGwAwAAkATBDgAAQBIEOwAA\nAEnIHOxGRkb6+vo8Ho/Rhcx3Ho9neHjY6Cr8gPcZpWma0YXMd263e2RkxOgq/MDw8HBfXx9d\n4vfL5XLxjJqKoaEhnlFT4XQ6R0dHja5itsgc7HRdJ9VNha7rhJWp0DTN4/HwprlfPKOmiGfU\nFOm6zl9pKrzPKKOr8A8SP6NkDnYAAAALCsEOAABAEgQ7AAAASRDsAAAAJEGwAwAAkATBDgAA\nQBIEOwAAAEkQ7AAAACRBsAMAAJAEwQ4AAEASBDsAAABJEOwAAAAkQbADAACQBMEOAABAEgQ7\nAAAASRDsAAAAJEGwAwAAkATBDgAAQBIEOwAAAEkQ7AAAACRBsAMAAJAEwQ4AAEASBDsAAABJ\nEOwAAAAkQbADAACQBMEOAABAEgQ7AAAASRDsAAAAJEGwAwAAkATBDgAAQBIEOwAAAEkQ7AAA\nACRBsAMAAJAEwQ4AAEASBDsAAABJEOwAAAAkQbADAACQBMEOAABAEgQ7AAAASRDsAAAAJEGw\nAwAAkATBDgAAQBIEOwAAAEkQ7AAAACRBsAMAAJAEwQ4AAEASBDsAAABJEOwAAAAkQbADAACQ\nBMEOAABAEgQ7AAAASRDsAAAAJEGwAwAAkATBDgAAQBIEOwAAAEkQ7AAAACRBsAMAAJAEwQ4A\nAEASBDsAAABJEOwAAAAkQbADAACQBMEOAABAEgQ7AAAASRDsAAAAJEGwAwAAkATBDgAAQBIE\nOwAAAEkQ7AAAACRBsAMAAJAEwQ4AAEASBDsAAABJEOwAAAAkQbADAACQBMEOAABAEgQ7AAAA\nSRDsAAAAJEGwAwAAkATBDgAAQBIEOwAAAEkQ7AAAACRBsAMAAJAEwQ4AAEASBDsAAABJEOwA\nAAAkQbADAACQBMEOAABAEgQ7AAAASRDsAAAAJEGwAwAAkATBDgAAQBIEOwAAAEkQ7AAAACRB\nsAMAAJAEwQ4AAEASBDsAAABJEOwAAAAkQbADAACQBMEOAABAEgQ7AAAASRDsAAAAJEGwAwAA\nkATBDgAAQBIEOwAAAEkQ7AAAACRBsAMAAJAEwQ4AAEASBDsAAABJEOwAAAAkQbADAACQBMEO\nAABAEgQ7AAAASRDsAAAAJEGwAwAAkATBDgAAQBIEOwAAAEkQ7AAAACRBsAMAAJAEwQ4AAEAS\nBDsAAABJEOwAAAAkQbADAACQBMEOAABAEgQ7AAAASRDsAAAAJEGwAwAAkATBDgAAQBIEOwAA\nAEkQ7AAAACRBsAMAAJAEwQ4AAEASBDsAAABJEOwAAAAkQbADAACQBMEOAABAEgQ7AAAASRDs\nAAAAJEGwAwAAkATBDgAAQBIEOwAAAEkQ7AAAACRBsAMAAJAEwQ4AAEASBDsAAABJEOwAAAAk\nQbADAACQBMEOAABAEgQ7AAAASRDsAAAAJEGwAwAAkATBDgAAQBIEOwAAAEkQ7AAAACRBsAMA\nAJAEwQ4AAEASBDsAAABJEOwAAAAkQbADAACQBMEOAABAEgQ7AAAASRDsAAAAJEGwAwAAkATB\nDgAAQBIEOwAAAEkQ7AAAACRBsAMAAJAEwQ4AAEASBDsAAABJEOwAAAAkQbADAACQBMEOAABA\nEgQ7AAAASRDsAAAAJEGwAwAAkATBDgAAQBIEOwAAAEkQ7AAAACRBsAMAAJAEwQ4AAEASBDsA\nAABJEOwAAAAkQbADAACQBMEOAABAEgQ7AAAASRDsAAAAJEGwAwAAkATBDgAAQBIEOwAAAEkQ\n7AAAACRBsAMAAJAEwQ4AAEASBDsAAABJEOwAAAAkQbADAACQBMEOAABAEgQ7AAAASRDsAAAA\nJEGwAwAAkATBDgAAQBIEOwAAAEkQ7AAAACRBsAMAAJAEwQ4AAEASBDsAAABJEOwAAAAkQbAD\nAACQBMEOAABAEgQ7AAAASRDsAAAAJEGwAwAAkATBDgAAQBIEOwAAAEkQ7AAAACRBsAMAAJAE\nwQ4AAEASBDsAAABJEOwAAAAkQbADAACQBMEOAABAEgQ7AAAASRDsAAAAJEGwAwAAkATBDgAA\nQBIEOwAAAEkQ7AAAACRBsAMAAJAEwQ4AAEASBDsAAABJEOwAAAAkQbADAACQBMEOAABAEgQ7\nAAAASRDsAAAAJEGwAwAAkATBDgAAQBIEOwAAAEkQ7AAAACRBsAMAAJAEwQ4AAEASBDsAAABJ\nEOwAAAAkQbADAACQBMEOAABAEgQ7AAAASRDsAAAAJEGwAwAAkATBDgAAQBIEOwAAAEkQ7AAA\nACRBsAMAAJAEwQ4AAEASBDsAAABJEOwAAAAkQbADAACQBMEOAABAEgQ7AAAASRDsAAAAJGE2\nugBgQbNrWqfTtdPp7HK5+lzuAY9nwO0e9mgjHo9T1x2aNurRvNe0qEqwqiqKEmJSrYoaYTZF\nmM3hZnOU2ZwQYI21mOMsVlUx9tEAAAxGsAPmgi5Es91RPTZWPTpWM2ZvtNub7Y4mh6PX5Z6p\nuzApSoLVkmazpdoC0mwBOYGBuUGBOUGBkWZe5jCGPjKi93SLgT59dFSMjQohhKoKW6ASGqpE\nRCnRMcJqNbpGQDa84wOzYkzTNg+NbB4eLh0Z3To8UjYyOuLxzOo9enS91eFsdTjXDnzl/ASr\ntSQkaElIcElw8IrQkNygQMb1MHv0jjatqkJrqhdOpwgKUWNjRUSUEhoqoqJ3XcM+pg/0a40N\neneHcDiV0DAlI0vNK1BCwwwtHJAEwQ6YMTsczg/7B9YODK4fGt42POLS9an8llVVYi2WeKs1\n0myONJvDzaYIs9mkKOFm075zYO2aPqZpY5qn3+0ZcLsH3J4ul6vT6epyuSa6/Tans63X+VZv\nv/fHCLN5VWjIYeGhR4WHHRYWGmwyHeCjBfagd3V6Nn2uNzUo8QlqboFl9eHCZpvSLw4O6PU1\n7tdfEiMjal6hunSFEhw829UCEiPYAQel2+V6t29gTV//B/0D9WP2Sa5pVpR0my0nyJYbFJga\nYEsPDEgJCEgOCIi2zMDL0KXrnU5Xo93e7HA02R31Y/bqMXvl6GiHc+/A1+92v93X/3ZfvxDC\noigrQ0OOiww/KTLi8PAwi8JYHqZJ17XK7Z71nyohoeqK1epJp4tpPouUsHBl6Up16UrhdmkV\n290vPqMEBZmOPFZJSJqlkgG5EeyAadOF2Dg0/GpP7xs9fZuGhrUJrhZlMa8KDVkaElISHFQc\nEpwXFDh7ycmiKEkB1qQA6xFfPb/f7d42MrpteGTryOjmoeEtXx1HdOn6usGhdYND9zS1hphM\nx0WEfz068uvRUYkBzHzC/mnl2zxrP1QzsyzfukQEBR3szZktavFStXip3t3p+fBdfWzMfPwp\nSkLiTFQKLCCKPrWjRf5oeHjYbrdHRkaaONg0KbfbPTo6GhbGBJf96B8cfK+3b43D9Vpv3w6H\nc98rKEIUBgcdExF+WFjo6rCQ7MDAuS9ycg5N2zw88vng8NqBwY8GBvYdzxNCKEKsDA05Kyb6\nW3HRB/YQXC6X3W4PDQ096HolNzg46HQ6o6KiVNXPOk/pO1rcb72mLE4zH/01ERAwK3fR1+tZ\n84YwqaZTznBZA5xOZ0hIyGzckUwGBgZcLld0dLTC6PuknE6ny+UKlvSgP8EOBLv90IX4dGDw\n6Y6uF7u6u3wtYs0ODDwxKuLEyPCjwsNn5LjqnKkcHfuwf2BNX/97fQP9bh8PbVlI8AVxsRfE\nxaTYpvHhTbCbIr8Mdg6H+61XxdiY6bSz5mC5g97a7P7vq1pugWvpyhCeUftDsJsigp2/IthN\nEcFuIrVj9ic7Op9s72qw7z15LkBVj40IOzM66tToyPSpTRKfzzy6vn5w+PXe3le7e7eNjO51\nqSrEMRHhl8XHnRsbHTqFVxPBbor8Lthp9TWeNW+Yjj9Zzc6bu3vVdedH72k1lQHnXqhERs3d\n/fohgt0UEez8FcFuigh2e3Fo2r+7e/66s+PD/oG9Xh6BqvL16KhvxcWeHBUxlYjjjxrs9pe6\ne5/v7F4/OLTXww8xmS6Ii/luQvzqsMkOihHspsifgp2mud9+QwwOms/4xhSXu84gp9Ppat9p\nXfOGafVhasmyOb53P0KwmyKCnb8i2E0RwW63Brv9TzvaH2vv7P5q9xCrqpweFXVWaPDXAm2J\n0VEL5BnVZHe82NX9VEfXluGRvS5aGhJ8TWL8JYvigny0ZCHYTZW/BDt9eMj9/NPqshWmZasM\nKcDpdDqdzpDAQPebrwjNYz79HLEwXoPTRbCbIoKdvyLYTRHBTgjxfv/AH1rbXunp9Xz1FXFI\nWOhli2IviIuNspiHhoYcDscCfEZtGxl9or3z6Y6uNudXloxEWczfTYj/QWL8XjPwCHZT5BfB\nTt/R4n7tP+azzlPiDVuguivYhYQIIbTSzZ4v1pvPv5R2d/si2E0Rwc5fEeymaCEHO00Xr/f2\n3t3Uun5waM/zw82m8+Nir0mMXxry5St/wQY7L00X7/X3/3Vnx3+6e9x7vG+oQpwWHXV72uJV\nobuOzxLspmj+Bzutoszz2VrL+ZfNQDeTg7BnsBNC6G073K/8y3zuBUpMnIFVzUMEuymSO9j5\n0wo+YAY5NO3R9s7ftuyo+2pX4SUhwdcmJVy4KDZwvn7WGkVVxAmRESdERjTZHQ/vbP9bW7t3\no1tNiNd6el/v6T05KvKW1OSjwhfiNwQpeT7/VK+vs1x6lZhn2w0rCUnmCy53//MJ82lnKckp\nRpcDzC+M2GHBjdg5NO3vbR33Ne9ocTh2n6kKcVZM9LXJCcdGhE/0iwt8xG4vox7tqY7OB3e0\nlX91Fe3XIsNvTUpcFWBhxG6/5vOInefDd/T+fvOZ5053J4nZsNeI3S5jo65nHjcde4KamW1Q\nXfMOI3ZTxIgdIAmXrv+jreOuppY92wsHqOpli2J/tjgpJ2je9ROez4JM6tWJ8d9NjH+tp/f+\n5h1rBwa957/XN/Be38DRoSH3ZmccFka280ue99boDof5rPOMLmRSgUGWS690Pf2o0DQ1O9fo\naoD5gmCHBUHTxfNd3bc1NNXuceA12GT6QWL89YsT463soHWAFCHOiI46Izrq04GhXzW1vNnb\n5z3/o6HhwzeVnhUT9av01KJgI6dnYbo8763RXU7zqWcYXcgUWK2WS77jevoxoShqVo7R1QDz\nwrwb/wdm3Af9A6s2bb2wvGp3qgsxmW5KSW44dMX9mWmkuhlxeHjoGyUF65cvOT06cveZL3f3\nLtmw+cqq2r2W02Le8nz6kT48ZD7560YXMmUWq+Xi73g+ek9rqDO6FGBeINhBZjVjY+eUVR63\npWzT0LD3HJuq/jQ5seHQFfdmpMZaLMaWJ5/VYSGvFResLSk4ZryDsSbEP9o6ctZvuqupZdSj\nGVseJufZtEFvbzOf8Q2jC5kmi8Vy0eWed97UO9qMLgUwHsEOchryeG6sayzasPml7h7vOSZF\nuSJhUdXq5Q9kpccQ6WbT6tCQV7Mz1iwpXDneAGXY47m9oTnv803Pd3YbWxsmotXVaOXbzGd/\ncz6slpg2W6D5ou+4X3pR7+8zuhTAYAQ7yEYX4pmOrrzPN/2mZYdT27Xo+8TIiM0rlz6SmzWt\nzexxME6MjPh8xZIn83MWB+z6m7c4HOeXVx2/tax8n+1oYSy9fafng7ct518q5t/63ClSgoPN\n517gfv4pscdqd2AB8tfXMOBT9ejY8VvKLq6o3jm+7jUvKPD14oI1SwqLmcI/5xQhLlkUW7l6\n+V3pKcHjPWLe6xtY+sWWn9c3jWkcmZ0X9JER9yv/Mp9/qfDzkWwlJtZ00umu554QPLWwgBHs\nIAmXrv+qqWXJF1ve7x/wnhNqMt2fmVa6atlpe0znx9wLMqm/SF1csXrZN2NjvOe4dP3e5tbi\nDZvf6es3tjYIj8f9zyfMZ5yrhMjQm0ZNy1ALit1vvWZ0IYBhCHaQweeDw8u/2HJbQ7N9/Jv6\nhXGxlauX37A4yeKPE4ZktDgg4PnC3HeWFOaP9wusG7OftHX7tytr+txuY2tbyNyv/tu08lAl\nwbB9YGecadWhwunUSjcbXQhgDIId/JtD035e33TE5tKy8WlbGYG2NUsKnynISQygj8m8c3xk\nxJaVy36ZnmJTVSGELsTj7Z1FGza/1tNrdGkLkWfTBmENUEuWGV3IDDN//RzPhs/0zg6jCwEM\nQLCDH9swNLx849Z7m1u9e9KbFeWmlORtK5edGBlhdGmYkFVVbktdvGXl0mPGd2/b6XCesa3i\n8sqafobu5pDe3alt3WQ+xX9a1k2dyWQ+7yL3f54XLhooYsEh2MEveXT9V00tR2wq3b2+sjg4\n6LPlJfdmpAaZeFb7gdygwPeXFv0pJzNkfFHFE+2dS77Y8sH4FEnMLpfT/a9/mr9xgf8ug52c\nEh5uOvYE9yv/MroQYK7J+ZKG3OrH7Edv2XZbQ7NrfKDuF6mLv1ixdEVoyH5/F/OHIsQ1ifGl\nq5Z+LXLX0F2z3XH8lrIb6xodrGqcZe7XXzYde7wSHm50IbNIzc0XgcFMtsNCQ7CDn3m2s2vZ\nxi2fDgx5f8wLCly3vOSu9BSryiIJv5Rus72zpOih7IxAVRVCaEL8pmXH4Zu31YyNGV2atLSq\ncmEyqbkFRhcy68wnn+7ZsE7vYwYnFhCCHfzGqEe7qqr2ovLqQbdHCKEI8YOkhI0rlq5koM7P\nKUL8OCnhixVLloUEe8/ZNDS84outT3d0GVuYlPSRYc+H75pPPcPoQuaEyWQ+8zz3yy8KXTe6\nFGCOEOzgHypGx1Zt2vpI265lbnFWy2vFBX/MzmBGnTQKgoM+W77kppRk7390yOO5pKL6yqpa\n+hjPLPfLL5pPO0uY/bsX8dQpsXFqVo5n3cdGFwLMET4U4Qf+2dm9euPW3eskjo8M37JyKW2H\n5WNVlXszUv+7pHCRdVfs+Edbx+GbSuvG7MYWJg1ty0YldpGSnGJ0IXPKdMQxWlWF3sM+xVgQ\nCHaY11y6fl1t/QXlVcMejxDCrCi/Sk9dU1KUYKVHnbROjIzYsnLpCeM9a7YMj6zcuPVVGt0d\nNH10xLNhnflrJxldyJxTFPPZ33S/+m8OyGIhINhh/up2uU7euv2h1jbvj7EWy5slBbemJrNM\nQnrxVuuaJYX3ZqR636H63e6ztlXcXN+k8bl8EDyv/Nt00ulivL/MgqJERqkZWZ4vPjO6EGDW\nmY0uAPBt49DwOWWVLQ6H98cjwsOeL8hlM4mFQxHippTk4uDgSyure11uXYj7mlurRseeyM8O\nXZDR5CBp1ZUiOFhNTTe6kBnj1vVGu6PRbm+yO5odjnanq93pdGm6dxNBp6aHmEzekX6vsMj4\nG9957UFhHbAF6roIUNUEqyXOakm12VIDAtJsAam2ABM7EML/EewwHz3X2X1FZc3uWfM/Tkr4\nv6x0dn1dgE6LjvxixZJvlFVuGR4RQrzU3XPYprFXi/PTbTajS/Mrbrfn/bctl19ldB0HpdXh\n/GJoaOvwaNnIyJDHY1aUdJstI9CWGhBQHBycEGCJs1gCJu23rEeFPvHJh+ZvXSKEsGtap9O1\nw+lsdTg2Dw+/2NXd5HBougg1mZaEBJeEBK0MDWHKB/wRwQ7ziy7EHY3NdzW2eI+5BajqwzmZ\n34mPM7gsGCfdZvtkWckVVTX/7OwWQmwfGV29sfQ/RXlHhocZXZrfcL+/xnT4UcIWaHQh0+PR\n9S3DIx8PDH7U1z/odqcGBa0MDTkxMuL6xYkHNmqrJC0WATatrkbNzLapaootIMUWIETontfp\nd7tLh0dLR0Ze7OppdzojzeYjwsOOCQ8rCg5mEgj8AsEO84hd066orH22c1f3ssQA678K8w4N\nC538tyC9IJP6bEHu0pDgW+ubNCG6Xa4Ttm7/W27mpYtI/Pun9/boO1rVE08zupCpanE43ujp\ne7dvYMDjXhESclRE2CVRkTbNExIyAx0rzSd/3fXk39W0jInmGkaYzUdHhB0dsetrQ5/bvXZg\n8MmOrq3DjTEWywmR4adFR8Yzkod5jGCH+aLb5TpzW8W6wV1bSiwPDXmlKD+JSXUQQgihCHFz\nSnJhcNBF5dXDHo9D0y6vqKkds9+RlsIwyuTcb7xsPv1so6vYv/KR0Re6ej4ZGEwIsJ4eFflw\nTma0ZdcnlNPpdDo9k//6VNls6tIVnnUfm448dipXjzSbz4iOOiM6SgjR6XS93dd/fW1jj9t1\nbET4N2NjsgKZEoB5h2CHeaF+zH7qtvLq0V27SJ0dE/1UfnYwc+TxVWdER328rPiMbeWtDqcu\nxC8bW1rsjr/kZjH/ciJadaUSGaXEzt+hzSa748mOzg/6B/KCgs6Ljf5FavJsr2AwrTzU9cjD\n6rKVSvD0hgDjrJaLF8VevCjWpesf9A/c39zaYHecFBVxcVws67owfxDsYLwvhoa/vq28w+ny\n/vizxUn3ZaQxnQU+LQ0JXr98yZllFRuHhoUQj7Z3tjldLxTmhvA1YF+a5vngHctlVxpdhw9j\nmvZCZ/fzXd0xFssli2J/njLree5LimI66TTP22+az/7mgd2ARVFOjIw4MTLCretv9fb/rK5x\nVPNcGBd7Tkw0m1bDcPSxg8HW9PYft6XMm+pMivKH7IzfZJLqMJnEAOsHS4tOjdq19ch/e/uO\n3VLW5XIZW9U85Pn8U7Vk6XxbM1E9OnZtTf2Z2yrsmvZsQe5jedknREbMcZ8RNSVNHxvV29sO\n8nbMinJ6dOQzBTmP5WV3ulynlG6/sa6x0e6YkSKBA0Owg5Fe6Oo+o6zc22sqUFVfKMz9UVKC\n0UXBD4SYTK8U51+RsMj748ah4aM2b2vmA3VPDodWutm0+nCj6/jSmt7+s8sq7mluvSw+7u0l\nhVcnxhvYktB88tfdb78+U7cWYTb/OCnhvaVFZ8VE3VLf9K3tVR/1D87UjQPTQrCDYf66s/3C\n8mqnpgshIs3mt5cUnhMTbXRR8BtmRfl7btbtaYu9P1aNjh25eVvl+DRNuD9613TUcWLSvm5z\nw6Prz3V2n7C17OOBwb/kZD2Wl70ydAbWtx4kJSpaiYzW6qpn9maPCA97piDngaz0V3t6Ty7d\n/nJ3L7ulYI4Z/5rHwvTblh3fr67z6LoQIsFq/XBZ8RG0JcM0KULcmZbyYFaG9zBei8Nx9OZt\n3lbGC5w+PKS3Nqv5RcaW4dH1Jzs6T9i6vc3pfKko/670lEVWi7El7cl8/MmeD96ZjQ1kkwOs\nv8lM+2dB7vaR0RO2lr3Y1cNueJgzBDsY4FdNLTfUNXrf6DICbWuXFRcHBxlcE/zWtckJj+dn\nmxVFCNHlcn1tS9mGoWGjizKY5923TF872cACdCGe7+w+Yev2PpfnjZKCnyYnzsfVLYFBanqW\ntr10lm4+wmy+JTX55aL8urGxE0vLXuvpnaU7AvZEsMNcu7Wh6baGZu/pouCgtcuKM+gFhYNz\n6aK4fxfl2VRVCNHndp+wteyTgYU7w0nv79MHBwzcFvbdvv6Ttm5vsNtfLy64NjkhcB4cDp6I\n6chjPZ9+NBuDdruFmEw3pSS/VJT/xdDwqaXlnw4Mzd59AYJghzl2Q13jPU2t3tPLQoLfX1rE\nboyYEWdER71clB9kUoUQg27PKaXlC3b2uued/5qPO9GQu64cHTt3e+Wbvf3PF+belJLs/XfM\na1arml/k2bJxtu8n1GS6Iy3l8bzspzo6L66obrDbZ/sesWDN+1cdJHJjXeNvW3Z4T68OC3l3\naVGMZR5NuIG/Oykq4o3iAu8hv2GP5/Rt5R8vvHE7vadLdzqU5JQ5vt8hj+dndY3/29h8f0ba\nbzPTIs1+0yTVdNiR2uefCk2bg/uKs1r+lJN5a0ryT2ob/rexeWxO7hQLDcEOc+Tm+qbfjKe6\nw8ND3y4p8qO3fviLYyLC3yopDDPvynanlZavXWDZzvP+23M/XPdcZ/eZ2ypOjIz4Z0Fupt/N\nrDBb1KIlczBot1tBcNDLRflLgoNPLS1/lYl3mGkEO8yFXzQ03de86wjsYWGhbxbv+ugFZtzh\n4aFvFBeEmr7Mdp8NLpRZTXpfr+50KglJc3aPDXb72WUV1aNj/y0pODkqYs7ud2aZDjlc27Bu\nbgbtdvtGbPQbxQUf9w9eWF610+Gcy7uG3Ah2mHX3NrfePT6v7pCw0DdLSHWYXUeEh71RsuuY\n7JDHc2pp+QLpgeJ5f4352DkartN08VBr2w+r6+/LSLs9bXHAPF4hsX9mi1pYopVumuO7DTKp\n92em3ZSSfHllzT/aOub43iErf34pwh/8YUfbz+ubvKdXhYa8VVIYTqrD7DsyPOyNkoJgk0kI\n0e92n7R1e4XsvYv1gQF9ZERJnIvhurox+9e3lauKeL2kIDdofm1ZdmBMhxzuWf/prC6PncjS\nkOA3Swo6Xa5zyipbGbrDQSPYYRY92t55XU2993RJSPB/SXWYQ0eFh7003gOly+U6cWtZ/ZjM\nSxE9H79nOuq4Obijv7V1/Kim/v9lZ/woKUGeXZ0tVjUrV6vcbsidmxXl5pTku9NTvl1Z/Vxn\ntyE1QBoEO8yWl7p7vltV6/3+mxsUuKakMMrCagnMqRMiI54vzLUoihBih8N5Uun2dqecIyL6\nyIje1aGmZczqvfS43N/cXtnldL1WnC9f+0nT4Ud7Pv3YwAIKgoPeKC4sHR65vLJm0O0xsBL4\nNYIdZsWH/QMXlld7dwxLswW8s6RoXm0lhIXjjOioJ/NzTIoihKgbs59aWj4g40emtu4j0+FH\nz+pdfNQ/eN72yptSkm9JTfb+PWUTGKjEJ2gNdQaWYFWVezJSr4xfdFZZxcYFv4HlzhgAACAA\nSURBVIEKDgzBDjNv28jo2WWVdk0TQsRaLP8tKUwOoAsxDHN+XMwfs3cNZW0ZHjm1dPuoR67+\nYS6XVl+r5uTP0s3rQvyqqeVvbe0vF+WvDA2ZpXuZD0xHHuf55EOjqxBHR4S9UJh7T3PrQ61t\nRtcC/0OwwwxrtDtO2rq93+0WQoSZTW/KMrcafu17ifG/TN/Vs3fd4NAF5VUeI6bJzxLPxvXq\nikPE7IyiDbo952+vsijKk/k50q9nV8LDlYAAvdP4BaoxFsuLhXm60L+1vUrKMWbMHoIdZlKf\n233a+DSmAFV9qSh/hdTf7+FHbktd/OOkBO/pV3t6b6hrNLScmaPrWulm09IVs3HbW4dHziyr\n+NnipJtSkmfj9uch01HHeda+b3QVQgihCHFdcuI1SfHeToFGlwO/QbDDjHFo2tllFd6mEqoQ\nT+ZnHxcRbnRRwJd+n5XxzdgY7+nfte78fetOY+uZEVrldjUrV5hmfiztha7um+ob/1mQuzps\nAX09U+IT9cFBfWS+ND48LiL8qfycH9bUv9LNHhWYEoIdZoYuxLcra3Zvu/6bzPTdn6DAPKEq\n4on87KPCw7w//k9tw3+6e4wt6eB51n+qHnLEzN6mLsT/NjZ/0D/4anHBAlz2ZDr0CG39J0ZX\n8aWkAOsrxfn/6u55oEWGryKYbQQ7zIw7Gpt3t1+6Ljnx+sWJxtYD+GRT1ZeK8vOCAoUQmhCX\nVFT79dpDvb1NCQ1TgoNn8DYdmvadyppos+WP2RkWKVe/7o+aW6DVVAnPPJrZFqiqj+dlmxRx\ndVWtS6LpoZgNBDvMgKc7uu5qbPGePjsm+oHMdGPrASYRZTG/UVIQZ7UIIUY92pllFTv8tt2/\n59MZ7nLS7XKdVVZxflzMtckJM3izfkZR1KISrWyr0XXs7brkxBMiI87bXkmXO0yChrE4WJ8O\nDF013oh4eWjIU/nZ6kL8ko9pcGp6vd3eaLe3O10tDkePyz3gdo94tBHN49R0IYRVVYJVU5BJ\nDTebYiyWxQEB8VZLqs2WabNZZ+LplW6zvVSU/7UtZXZN2+lwnrGt/ONlxcGzME1tdo2N6QN9\nSsKMjY432O1XVNY+kJW+LGQmhwD9kWnloa5nHlOXLDe6kL19Ky4m1RZwdlnFE/k5tJGCTwQ7\nHJRmu+Oc7RXelnVJAdZXivL979MRs6/H5f58aOjzweHSkZHS4ZFGu8N9QIeTTIqSZgsoCQ5e\nEhK8KjRkdVhIjOUAZ4AdFhb6j7ysi8urdSE2D49cVlnzYmGef30l8Wxcb1pxyEzd2qah4Z/V\nNT6el51iC5ip2/RjAQFKZJTetkNJmIu9d6flkLDQh3MyL62ofjgnM49mUtgHwQ4HbtSjnbO9\nstPpEkIEm0yvFOUn8Q0S44Y9nvf7B97u7X+/f2D7yOiMTAvy6HrdmL1uzO5d9KAIURAcdFxE\n+ImREcdFhodO80vFhXGxVaNjdza2CCH+3dVzV2PL7WmLZ6LMOaKVl1muvGZGbuqj/sFfNjU/\nX5h7wEFZPqbDj/Z88qH5nG8ZXYgPuUGBzxbkXFRefU9G6qFhoUaXg/mFYIcDpAtxVVXtpqFh\nIYQixCO5WctpWQchelzuFzq6/tXZvXZ4xDuUO3t0IbaPjG4fGf1/O9oCVPXYiLBzYqLPjY2e\nejr537SUitGx5zu7hRB3NDbnBwf6y2pura5aSc+YkS4nr/b0/m1nx0tF+SEMt+9BiVuk9/UI\nu13Y5uOuuPFW64uFeRdWVP08JflYGkthDwQ7HKD7m1uf7ezynr4lNfn8OP/4OMQscWjay929\nT3Z0renr886Tm/sC3urtf6u3/9ra+hMjIy5ZFHdOTFSAup/1Yd7vJJWjY6XDI7oQV1bWFgQF\nFQYHzU3NB8Oz/lPzqWce/O280NX9r66eFwpz9/u3WoDUZSs9W74wHXqk0YX4FmUx/7sw78Ly\naoemnxwVYXQ5mC94JeNAvNPXf2tDs/f0GdFRv0xLNbYeGKjBbr+hrjF53Rfnl1e91tNrSKrb\nk1PTX+/pu7C8Kmndhv+pa6gbs09+/RCT6aWiPO8g35DHc64/LDnUh4eEpimRUQd5O091dL3S\n3ftUfg6pzidT8TJt2xajq5hMsMn0fGHuX3a2074Yu/FixrQ12x0Xlld7t9rMDwp8Kj+HZbAL\n06cDQ+eUVWZ9tvG3LTu6XS6jy9lbj8v9QMvOnPUbzyqrWDswOMk102225wtzzYoihKgaHbu8\nsmaeNwrTNq43rVh9kDfyeHvnu339j+Vlmxdks7opMZuV+ES9tdnoOiZjU9XnCnKf6uh6yf+7\nbWNGcCgW0+PQtPO2V3o/xUNMphcL86TfFxz7eq9v4M6m5t0bjRwARYjkgIBUW0BygDXOaok0\nm0NNpj1XPwx5PEMeT5/b3el0tTqcTXbHDofjAKbsaUK80t37SnfvkeFht6ctPjHS9xGr4yLC\n781I/VldoxDipe6ee5tbfz6Pd0fVqiosR33tYG7hn53da/r6H8/LNpHqJmU69AjPxx+Yk1OM\nLmQyVlV5uiDn4vLq0dCgk1gnu+AR7DA919U2bBhfMPF4XnaBP8xGwgxaNzj08/qmD/sHpvuL\nVlVZERKyOix0eUhwSUhwblBg4DQP/9k1rXJ0bNvIyKahkc+HhjYOjTimszhj7cDgSVu3HxUe\n9uuM1CPGdxXb0/WLkz4fGvYupLitofmQ0NCvRc7HOelafa2Smi4O4uDpMx1db/X1P0GqmwIl\ndpHe1yMcDhEwr7vAWBTlqfyc80rLFI92QXS00eXASIou7+Ykw8PDdrs9MjLSxFKvSbnd7tHR\n0bAwHx91e3mmo+viimrv6ZtSku/NWFhT64aGhhwOx4J9RtWN2W+qb/xX1zQO9yhCLAsNOSUq\n4oTIiEPDQqeb5CZn17TPBofe6ev/b2//pqHhab2RnRMTfV9manbg3mMbwx7PIZtKy0dGhRCL\nrJbNK5cmWGexg8/g4KDT6YyKilKn85dxP/+U6YRTlagD/PB+ubv36Y6uZwpy/OgIrNPpdDqd\nISHGrLv3bPhMKIpp5Yy1DJw9XX39lzc0XZuackp0pNG1zGtOp9PlcgXP6F588wdz7DBVlaNj\n36uu854+NiL87vR5fWwCM2jE47mlvqlww+YppjpFiNXBQb/PSm86bOXGFUvuTk89LiJ8ZlOd\nEMKmqsdGhP8qPfWLFUuaD1v1YFbG4eGhU4wq/+nuKdqw+eb6ppGv7gcaYjL9qzDP2/Wjw+m6\naHwu6TzidOijowec6tb09j/W3vlUvj+lOsOZliybh9uL+WRVlT8nJzy4o23d4JDRtcAwBDtM\nyahH+9b2ymGPRwixyGp5piCHgzgLxGs9vcUbtvy6uXUqxz2TAqy3pS4uX7Hknbys65ITF8/V\n0avkAOu1yQmfLCupPWTF7WmLp7LVklPT72tuLdywea/lhHlBgX/OyfSe/qB/4I7xTZDnCc/W\nTWrx0gP73c8Hhx/csfOZgpwZ2ZZtAbEGKMHBek+30XVMSZCqPpefc1tD0/aRUaNrgTEIdpiS\nn9Y1bBsZFUKoQjydnzOrx6cwT3S5XBeUV52xraLBvp+OIUKII8PDXijMazx05S/TUzKN25Mq\nI9B2Z1pK46ErXyzMOzpi/7MLmuyOs8oqvjm+gYrXxYtiv5uwyHv6nqaWd/v6Z6vc6dO2bTUd\nULCrGRu7paHxqfycGR86XQjUFas9Gz83uoqpCjObni3I/XFNfaPdYXQtMACvcOzfi109f93Z\n7j19e1rK8ROsK4RMXuruKdqw+Z+d+xmlUIQ4Mybq0+UlHy8rPi82ep4c4DMpyrmx0R8uLf5s\necnZMdH7renFrp69DjQ/lJ2xJCRYCKEJcVllTdf8aOai9/UqoaFi+l+rdjqcV1fVPZmfE2lm\nwdyBUNOz9IY6Md+Oy08s1mL5e27WFZU1vS630bVgrhHssB8tDsf3qmu9p48KD/tF6vztAYEZ\nMeLxXFVVe07ZVwaxfDo9OvKLFUteLso/bL7uVnlIWOh/ivI2rVx6RvR+evl2u1znba+8orLG\nO9/ApqovFuZ526/sdDgvr5gXne20jevVZaum+1tDHs9lldV/zslkoP3AKYqSlq411htdxzRk\nBNoezM64sKJq1DO7O/thviHYYTJuXb+wvNr7nS/aYn6mIJepdXIrHR5ZuXHrI20dk19tVWjI\nR8uKXysu8IsNgpeGBL9SnL92WfEh+wugj7Z3rti4dfPwiBAiK9D2+6x07/lv9vY91Lpz1gvd\nH62uVs3ImtavuHT9korqO9NScmlvdnBMK1ZrmzcYXcX0FAcH/SQ58aqqWqO3g8GcIthhMvc0\ntX4yMCh2bamZPZU56fBfj7V3HrqptHJ0bJLrLLJaHsvLXr9iyVG+WsHNZ0eEh61bXvJEfnb8\npANX1aNjh28q/XtbhxDiioRFuzdBvrm+aZuhs9G15kZlccp029f9qKb+4rhYn337MC1KTJze\n2yvcfnZk89SoyGMjwm5paDK6EMwdgh0mtGFo+K6mXUsCr0lKOCvmYDemxLzl0vUfVNd9p7Jm\nbOKlr6oQ30+Mr1y9/PL4OD8dtlWEuHRRXNXq5T9KSpjkvc+uad+tqr26qtap6X/JyUqzBXjP\nvLSielotkWeWtuUL0zSPw/62ZUdKQMC3xrMpDpKaV6BVlhtdxbRdnRjv1LTd86QhPYIdfBv1\naJdWVLt1XQiRFxT4m4w0oyvCbOlyuU7YWvbwpO/72YGBHy4rfjgnM8L/Z9+HmU1/yM5Yu7wk\nb9Kjk39r6/ja1jK7pj2Vv6u5z9bhkV80GLRtqKbpbW1KQuLUf+O1nt6ykdFbmBQ7c9Qlyz2l\nm4yu4kD8JjPtrb7+g9kDEH6EYAff/qeuoWp0TIzvVBNk4qkip/KR0UM3lU7yjq8I8eOkhK2r\nlh4p1+G8w8JCN61c+tPkxElGHz8ZGDxk09Zws/nGxUnecx5o2fHB9LdTO3habbWanTv161eN\njv2/HW1/ycn007HV+UkJDRMOh7BPNldhfjIpyuN52bc3NteP7b91Efwdn9bw4a3e/r+Mj9/c\nkZaywh8myOMAfNg/cMTmbZO81y+yWt4sKXwoO0PK5meBqvpAVvqaJYWTLBdtsjuO3Fx6bES4\nd5mIJsR3KmuGvrpfxRzQSjepS5ZP8cp9bvfV1bWP5GYHyPhfM5ZaVOIpLzO6igMRYjI9lpf9\nveq6kTl/9mKO8bLH3gbcnu9W1XoXUR0RHnZTSpLBBWF2vNDVfXJpef/Ek8GPjwzfsnLpyVGS\nty08ITJiy8qlJ038MAfcnjPKyi+Ki7WpqhCi0e74WV3j3NUnhPB49MEBJXpKU+U0XVxRWXt/\nRloSS51mgaloqba91OgqDlCaLeDGlKTvj+8MCVkR7LC362rrWxwOIUSQSX00L4v+JlL66872\nC8snXAqgCHFLavJbJYWTLyCVRpzV8mZx4e1piyd6Q3Rq+o11DaeNh7+/7Wx/s7dvzsrTqivU\nrKkeh72jsfm06Mj9NnbBAQoMFLouRv11t64TIyMKg4MenAe9ezB7CHb4itd6eh9v7/Sevjcj\nLTuQ3lcS+l3rzu9X1020vX2IyfRCYd7d6akLKtOrirgzLeXfRfnepsT70oT4d3dvms0mhNCF\nuLqqdsA9R4e0tNItU9wf9rWe3janc/d+aJgNamGJp3yb0VUcuJtSkj8dHPp4gIUU0iLY4UsD\nbs/uUfpjI8J/mJhgbD2YDfc377i+tmGifqWLAwLWLis+NzZ6TmuaN86KifpkeUnqxHvdNtrt\nFkURQrQ6nD+ra5iLmtxufXhIidx/s6FGu+N3rTv/kJ0xB0UtZKbCYv89GiuEUIT4W07WbQ3N\n+91aBn6KYIcv3VDXsMPhFEKEmEz/yMtSF9B4zUJxb3PrTfWNE126PDRk/YoS7x6pC1ZxcNBn\ny0tWTbxgyDU+0vlIW8d7fbO+QlarqZzKeliXrn+vuvZP2Zk2FkzMNlugUBT/PRorxjv+XFFV\nM9GwPfwabwHY5f3+gb+PbyR1T0Zqus1mbD2Ycf/XsuPn9RM2oD8pKuLDpUVsJyqEiLda319a\ndGpU5ORX04W4smrX3rKzR9u2RS1ast+r3VjXeFXCIvYNmxtqQZGnwi/Xxu5WHBx0ZnTUPc2t\nRheCmUewgxBCjGn6VXushOUgrHz+vLP9honXcp4fF/NqUUHIBNPLFqBgk+nl4vyLFsVOfrVG\nu+O2WW1Z7PHog4NK1H6OjL/W0+vU9W/GssPEHDEVlmh+HuyEEFcnxleOjjHZTj4EOwghxD1t\nHd5mZjZV/XsuB2Fl88/O7h9W10100OWqhEXP5Oda+a9/lUVRnszL+X5i/ITX0IUQ4g872jYM\nDc9SDVpdjZqZPfl1Opyu37Xu/G1m2izVAB8Cg4TbLRwOo+s4WH/Kzry1vqnH5Wcb4GJyBDuI\nrSOjf+rs9p6+LXXx5Psswe+829d/WWX1RFuc/jAp4a9E+QmoivhTTuZ1yRNs5KUIIYRH179X\nVeuenblK2vbSyY/Darq4sqrmoSw5O0jPZ2pegVblf/vG7iXcbLo/M+2HNXS2kwrvBQudposf\n1TZ4P5aKgoNuoB2xXEqHR76xvdKp+Y4dP0xK+EN2BqFuEooQv8tKnzDbCSGE2Dw8MiuNwXRd\n7+lSYuMmucoDrTtOi4osDA6a+XvHpNSCYr9uerLboWGhhcFBj453uYIECHYL3R93tn02NCyE\nUIX4c06mZSG1LpPeDofztG3lgxO0W7syYRGpbiq82W6yY7JC3NbY1GCf4V04tZYmJTllkits\nHxn9eGDwmiRmxBpACQsXI8NCiu25bk1Z/EJnd82Y/+2BC58Idgtam9P5i4ZdyySvTow/Qq5d\n3he4EY/nrLIKb/+afV0QF8MO8VOnCPHH7MxLJl5LMebRf1RTP7N3qm3fphaUTHSpU9Ovra3/\nM/9E46iZ2VpdjdFVzABVEX/Izri2pmGCkX34GYLdgnZDXaN3OCfeYvl1RqrR5WDG6EJ8u7Jm\n4wST+k+JinwiP2dBbSxx8FRFPJqX/fXoCRsFv9HT92p37wzeo97SpC6ecMTujsbm7yfG057G\nQGpBiSbF0VghRGag7eyYqN+17jC6EMwAgt3C9fHA4DMdXd7Tv0qKjzCbja0HM+hXTS0vdvX4\nvGh1WMiLhbkccz8AZkX5Z0HuYRNvw3ppZfWoZ6JlKtOj9/YoMTFign/TpqHhVodzuv1Nel3u\n8pHRjUPD7/T1fzwwuHFouMnumGi/YOyXErdI7+oQsvT4vToxfu3A0PYRP268DC8+yxcot67v\n7n9xZFjoeeO7m0MCr/f03TFBc7XMQNurRQXB9Ks7UEEm9ZXi/CM2b6se9TEhacDtubKq5tmC\n/W8UsV/a9lI1v8jnRXZN+1ld4wuFefu9karRsQ/6B9YPDpWOjFaNjk3US3lxQEBeUOCK0JAj\nwkOPiQifaLdc7EtJSNLbdyoJMqw5U4R4KDv9qqraN4oLGM73awS7Ber3rTu3jYwKISyK8qes\ndEVI8qUT9WP2Syt8NzeJNJtfLy6Is1rmuia5xFgsrxXnH7ap1Gf3r+c6u69KWHR85MF+U9Jq\nqyyHHuHzol82tvwwKSHaMuG794ah4Wc7ul7q7p3ieo4Wh6PF4Xi7r18IYVGUoyPCzo2NOT82\nJmriu4CXml+klZeZpAh2QojFAQHnxcb8vnXn/yyW5BEtTByKXYjanc67mlq8p3+SnJhP4zpZ\nODTtm+VVfW4fgcOsKC8U5rLl1IzIDgz8V2HeRIezz91eaT/I45t2uzCbhcXH/LmtwyONdse5\nsT72ohj1aH/e2V6yYfPqjVt/17rTZ6oLM5vSbbaMQFtGoC0xwBqwT/c7l66/2zfwg+q6xHUb\nLq6oXj84dFAPRHZqWobWOMOLZox1VcKijwYGa8dmeIk35hJfyBaim+ubvGsmFlktv0hdLBiu\nk8X1dY2bJlgw8UBW+sEPI2G3YyLCH8rOuKbaR2fXAbfnG9sr3yguOOAb12oq1WwfR1rduv6z\nusYn8/fei2LE4/nDjrYHWnZ2uVx7nm9WlFWhIUeGh60KCykICsoItO3bx7jd6awds28dHlk/\nOPxef//uZdQOTXumo+uZjq5jI8LvSFt8TET4AT8cmZlMIjBQHx5SQiaceelfFCEeyEy/rrb+\n1eICDsf6KYLdgrNxaPjJ8V6U92ekhZlNbl8DPPA7/+nu+dOONp8XXZGw6Md0O5tp30+M3zw8\n8ted7fte9GZP33OdXRfE7Wer2YloleWmE07Z9/wHW3deGBcTv8dKWE0Xj7V3/KKhuc35ZV8b\nk6KcFBlxQVzM16Oj9ns4Nd5qjbdajwwP+2GSEEJsGR75V1fPUx2djfZd+2V90D9w7JaB06Mj\n/y8znRHffak5+VpNpWnZKqMLmTGZgbZjIsIfb+/8dvxkzbExb3EodmHRhfhxTb33KNEhYaGX\n8rqVRYvDcVVVrc+LVoSG/DE7Y47rWSAeykpfHRbi86IrKmub7Ae0l6iu64P9SuTefVWa7Y53\n+wa+k7Bo9zllI6NHbi69sqp2d6qLNJt/npLccOiKN0oKLouPO4BJcktDgu9KT6k7ZOWbJQUn\n77Gm6vWevpIvNt/e0Mwq2r2oeQVaZYXRVcywnyYnPtXR2el07f+qmH8IdgvLsx1d6waHhHcB\nVBa7DkhC08W3K2t6fc3ljzSbXyzMs7GR6OwIUNUXCvJ8rmMY07QLy6s80++Foe9sVRKS9z3/\np3UND2Sle1+zmi7ua25duXHruvE5cGFm0y/TUxoPXXlPRurigIDp3uleVEWcEhX535LCDSuW\nnBIV6T3Tqel3NbWs3Lh1y/DIQd6+TJSQUDE6LOTKu2ZFuSc97ab6RqMLwYHg7X4BGdO0W8b3\nmbhkUdxEIw3wO79v3fle38C+5ytCPJaXnWY72I95TCLFFvB4Xo7P70jrBod+2zLtjq9aZbma\nu/cEu5e6ewqCgvKCAoUQHU7XSaVlN9c3eQfPFCGuTFhUs3rFbamLw8wz3KlkZWjImyUF/y0p\nzBk/CFs2MnroptL/N8FB/4VJWZyqtTQZXcUMWx0WEqiq7/f7eGPBPEewW0B+17LTe2woUFV/\nlT7ZHpTwI5WjY7v3hdvLdcmJZ8ZMuFMCZsrp0ZHXT9Ae4rbG5tJpjm9pjfVq6lcOnY9p2u9a\nd96ckiSE+HxweMXGLe+O5/isQNsHS4v/nps1q11sTo6K2Lpy6S2pyWZFEUI4NO3HNfWXVlQf\n7OJfWag5+Vq1bEdjhRB3Z6T+srHFJUsH5oWDYLdQdDpd97W0ek/fkJKUwiiOFNy6fllF9Ziv\nz9elIcH3sk3cXLknPXVlqI8hcJemf7uyZhofjXa7sFrFV7eB+VVTy/XJScEm0wtd3cdu2bZ7\n4eoVCYs2r1x6dMRcbPFsU9W701M/Wla8ewD4qY6u47aUMQ1LCKGmpOnNso3YCSEizebL4uMe\nat1pdCGYHoLdQnFHY7O3xUmC1XoDzSdl8duWHRt89TexqeoT+Tn7dinDLLGqyrMFuSG+9mzY\nPDxyf/NUD8hqNZVqVs6e59SO2ctHxs6KiXqote2C7VXeEB+oqo/lZT+Sm+XzHmfPYWGhm1Yu\nPT1616y7zwaHjthcOsU2yDJTVWGz6SMSTj38dnzcmr7+9j3WXGP+431/QageHft7W4f39F3p\nKXP8YYBZUj069svGFp8X3ZuRWhwcNMf1LHBZgbb7JhgivauppcLXFmT70moq1Zz8Pc+5qb7x\n3ozUu5tar6vdtZ49wWr9cFnx5QYtaY80m18pKrgxZdeXw9ox+1Gbt1VN7dFJTM3O1WurjK5i\n5ilC/Do99dYJtijE/ESwWxBuaWjyHgzKCwo06vMAM0sX4gc1dT4Pwn4tMvza5MS5LwnXJCXs\nXkO6J4emXV1VO5XDsXp3lxIds/vHd/r6M2y2F7q6d0+jzAsKXLe8ZJWvw75zRlXEfRlpf8rJ\n9H5+7HA4j9tS5nPz3IVDzc7TaiQMdkKI5aEhihDsQeJHCHby+3xw+N9dPd7Tv8lMM7O7sxSe\nbO9819dK2FCT6ZHcbP7HhlCE+Gtups+lqWsHBv8xPmo+Eb2nW4n58nuXW9fvaWqNsphvGx8v\nWRoS/NGy4tT5MUH2msT4pwtyve8nbU7n8VvLDrBvnxSUyCi9t9voKmbL3emp/9vYzBoKf0Gw\nk99N9Y3eF+RR4WFfj2aNpAx6Xe6f1TX6vOjXGan0NzHQ4oCA+zPSfF50U31jt2uypQZaTaWa\nnbv7x7/ubE+3Bdxav2usbllI8LtLimIts7j6dbouiIt5djzbtTqcp5Ru7/HVTHGBUOLi9a5O\no6uYFYuslmMiwp/vlDa5SoZgJ7k1vf0f9A8IIRQh7stMM7oczIxbGpq6fEWEI8PDrklk6zCD\nXZ0Yf6yvnVV7XO6b6ydbO6nV1aiZu1ZODHk8f2/reKqzy/utrCA4aM2SwgPYSWK2nRcb/Vhe\ntveDpHJ07OyyigW7NYWanavJOM3O66fJiQ/vbKfBjV8g2Enu9sZdB3HOjIk6LEySbaoXuE1D\nw3/ztT+pVVX+kpOpchTWaIoQf8nJ9Lkk+dG2jgnnKum6sNtF0K4lLzfWNdbZ7U5NF0IkB1jf\nKimMmU9jdXu6eFHsA1np3tNrBwavqa4zth6jqBnZWr3vbf0kYFPVqxMW/YHG1P6AYCezf3f1\neD9FVCHuTKMjsQx0Ia6tbfD5rfnGxckFrISdH3KCAn+e4mNbME2I62obfM5V0ne2KgnjS01H\n7c90dnn7E4WaTK8XFyQHWGex3IN2XXLitcm7hoofbe/848L8+A8MFGNjku0ttqcLF8Wu6e1f\nyEfb/QXBTlqaLu5s2tUL48JFsUtCgo2tBzPi2Y6uTwYG9z0/3Wa7JdVHkoBRbk5Jyg4M3Pf8\n9YNDT7b7mIml1Vbv7mB3+rZyb6pThXi6IKfEH168D2SmnxwV4T19fV3DjLlO4AAAIABJREFU\n54M+2itKT0lK1tumvYmcv1CE+EVq8r3NrUYXgv0g2Enruc4u715GFkW5g+E6KYxp2s8n2D3s\nwez0QNoRzycBqvpQdrrPi25taBr17D2uozXWq2kZQoj7mndUj+1qHXJb2uIz/GTBk0lRnsnP\nTbfZhBBOTb+wosqbTRcUNStH4ml2QohjIsJrxsaaF/DyZ7/AJ4GcPLp+1/hw3bfj47ICbcbW\ngxnx+9adPt9ST4uO9JeP/wXllKhIn3v1tjqcv2356riOpglNE1Zro91x23h2PyEy4vZUf/pK\nFmUxv1CY651cWD9m/1FNvdEVzTU1LUNrajS6itl1e+ri3R8umJ8IdnJ6trO7cnRMCGFRlJ9z\nhE4K3S7Xfb4OglhV5XeZvkeGYLj/y0z3uYrity079txlVWtpUpIXa7r45vZKby/xOKvlyfxs\nv1sKsyI05Nfj22882dH5UnePsfXMNYtVuF3CI/NQ5fLQkCGPp3Jh96Oe5wh2EvLo+t3j36iu\nTFjkPTgCf3d3U+uAr2NbP0hMyAnyMZcL80FWoO3aJB8NaIY8nj2HPfT6WjUj6887274Y3/n3\nbzlZ8dZ5vWBiIj9JTjwhctdku2uq63sX2Fx7JTlF2yH5LLRfpqXczaDdPEawk9Du4Tqrqvhc\nmge/02x3POyrxUmUxXxb6uK5rwdTd2vqYp9dhf/a1t40vre61tTQsSjpxvpG74+Xx8f5PIbr\nFxQhHsnN8m6/0e503jD+oBYINSNbr68xuorZlRMUaFaUspFRowuBbwQ72Wi62P1d6or4RSls\nQiCFXza1+Oz7emvK4nnYtBZ7CjebbkvzEb6dmn5vW6cQQmia0LXrmlpHPJoQIsFq/V2Wfx9b\nT7EF3De+/cajbR1rfa3jlpWamq41NRhdxaz7Rerie5okH5j0XzJ/JDQ1NV166aUmk0nZZ3fU\nU0899c477zSkqtlz+OGHu93u3pIldZd8WwiheDzrfvSD1f19WVlZzzzzjNHVzbDLL7+8oqLC\n50Xr16/f9z/u1376f//3j2WrxT5Ttax9vSeNxvj8Ff/13HPPPfDAAz4vuuOOO0477bQ5rmdG\nfC8h/sHWnXVj9r3Of66nb/MN/7Pa7VyenfFCerH3zKBH/37yTzZfdNFFP/nJT+a80hlzdUL8\nkx2dnw4M6UL8qKZ+44olI0NDJ5xwgs8rr1q16o9//OMcVzhbLBbh8QhNO+Osszo6fOwOHBkZ\n+dZbb819XTMrM9BmU5WtwyPeRlrXX3/92rVrfV7zjTfeiImR6m3q4YcffvTRR31e9OCDDx52\n2GFzXM++ZA52o6Ojmzdv9nlRTk7OHBczB9atWycURXz3Gu+P+n/f2Pr2GiHE2JiEs1zLyso2\nbdrk8yJd1yULdq9Fxui+JuA7//Kw69d3z309s6q9vX3Dhg0+L+rq6prjYmaKVVXuSk+5qLx6\nr/M1ISpWHXLq26+9Urhk11mbvqj7+1/rhDjyyCPnusoZpSriT9mZKzZu9ej61uGRR9o6zrNZ\nJ/rPhoZKtSmOkpikt+3YvHnzjh0+etolJSXNfUmz4dbUxb9oaHq2IFcIUVVVNdE/1zk+30Aa\nLS0tEz3Y/v7+OS7GJw7FyuWwI0R2jhBCaJp47mmjq8EMqBkbq0vP9HFBU6N42++/9y8c58fG\n+u4zfNzxhxy6+vO8YiGE8HjEQ7+f48Jmz5KQ4O8lxntP397YPCLvlgx7UdMzJd5bbLfMQFuA\nqjLTbh4i2Mnl4kt3nXjvHdHKBAgZ3Nu8Q/c5APnoIxJvXiQfVRF3+ppppypKYGLikNkshBCv\nvSLk6oJ2R9ricLNJCNHhdP2xq9focuaImpqxEKbZCSFuTkm+v1nanTb8F8FOIkuXiYJCIYTQ\nNPH0k0ZXgxnQZHf43H5K1NeJjz6Y62pwcM6KiV62z6Bd4fBgeVikEEKMjorH/mFAWbMp1mK5\neXxh/p+6e0RYuLH1zBGbTdjtUk0HmUBeUKBb16vpaTfPEOwkctElu058+oloXBDfF6X3QOsO\nb7vavT32D+HzfMxjihD7bu53RF/Xp5GxQgjx4vOiv8+AsmbZtUmJCVarEGLIo4lvnW90OXNE\niYvLDA8zuoq5cFNK0t7bqMBoBDtJlA6PiJWrd/3wnGxrYBembpfrkTZfw3WNDWLtx3NeDmbA\nGTFRS746aHdob/e6yGihC7FGzhmTQSb1y81vzjlXyLVOYiJqasZyv+1EOC1LQoK7XG57cIjR\nheBLBqyK1TRNn5PBhknuRdd1j1y7vvy6qVV4Z2Jt3SK2b9vzov0+WO9/xL/+ILv+uWaLcLv2\nusjj8czSE8x7s3P2hxpwuo6PCHutp2/vmXRPPSH0XedpmjYP/3EH84zSJp44OD8f7HTdnJx4\nYeWXDWzjXPYuq020NAvtKw/N716Sk7giLuaeptZ2p1MEBYuzvyGefHyvK8j3HqWnpK2Mi57o\n0tl7ILvfo+ayM8C1iYuuWrlKvPC8z0s9Hs88/MdpmnbA7yeTfL7M2XuUoiiqr1YJXgYEu6Gh\nIZdr7w/j2WB3OMUxx4nmJtGw917UTqezr0+qox6nBlqfq6wQefni2b0Xw2qaNpUH619/kLGo\naPHj68SKVeLKy/fambGvr2+SZ/z/Z+8+49uosj4A31EZ9W7LvcuWLdckJGGXzi6hQwhLQi9h\nqUsvoQZIKEtoSyi7lAABAil0CAll6e9S0t1tWS6yZcmWrd5H0sz7weAoiZM4saWRRuf58UUF\n+8iRRv+5c++5U+d2J6jbqgyh13LUvSrFxR9+pK+oQmMbTFnM6PtvY4tJ2n+4w2tz4AqFUE4u\nspj3fcjn8yXti528E9msCh6uDxEIIVmE+FWegQgC/eNa5NnjfRUMBhnwYsddo5QtHxqhfvsV\nTdSlKBKJTObFhkKhOJQWLy7exHs5TvKAPBUJbrpRi1CIiyMMm3B+iMvlEgiSdM/Dw3tHDXE4\nCMfRRMc3r9ebmI8tm81WKBT7e5SGYMflcuP6vTtuM8ZCDz2Mvv8WLXtgr4dYLBaPx6gtGc7J\n5F123VWopha1tuz1EIZhB36xFEVFIhHuRLseJSFrOHJTv6njvgd/H6E89nj03TexT+DxeHF6\ng0UikWg0iuN4Is+GK3m8Ob/9on9gKTp7Pjp3IVr3bmyQ5XK5SfhOHjsb5nAO5/BC8XjoxlvQ\nvUv2fSg5X+xhuCFbfXv/YJiiXGx8uaYGGbpQYO+eEWw2mxkvdszV2Vl/4rBPOeGOCR896AF5\nKu8oWrT4A3eU6iZ86KAH5KkIh8MkSSb+nVP9xeeD+xnHwnE8Cd/JU3lHfVc/C7HZEz6UsGPU\ngb+DaPicCIXCxPyi03Du46NWdOzxKD9/r94fXC6XYS0xf9fSvO99LBbrwC82Eon4/f5U+YPw\nKapF343G39YXXYK+/zb2TFEikcQp2Hk8nmg0KhKJ2Pv5VMcJl8tFDjta/Tr64D205ymmSCRK\nwn+4cDgcDAYPrzAli4UsA0hXjdpa93qIz+cn4Ys9DBfx+csGh22RMBp7FyuVKBLZ6zk4jjPj\nxY6RIMQn93uJis1mH/jFEgRBEIRYnDITuf5tski+3DxhhzcMw+L3L+tyuUiSFIvFCW7SfoCE\nJBaLk/CdTBBEOBwWiSZqLXlAX9qdBW6ncT9t/wUCQTK8WCYvnuBjCH38IWKx0MIL6K4FTBsu\nht2Sn7v7dplm96oRxvN4Jhz/Z5p176LzL6K7iDh6c3jEFjs3VChE7JQZiwIH1RMIhimKMwid\nRBno+UHzrP4+uqs4CCYHO4QQ+vB9FAigU05Dqv3OYwUp56qcLE7spauLL6WvFhAHI1Y0OoKq\na+iuIy5CJLlHewhDF2rcia65jr6KwDR7YmDwjgKG7BsGYm31eEv4fGHSn10zPdh53GjTRsTl\nogXn0V0KmDZiNjvz/37cfbuuHtU30FcOiIN33mJqXn9jyGoOxXwxvLsGPbUCnXwqwph+NE4P\nphBhJcJHSFLmqjGYvKcHBm8ryD348+iWBoeSDetQJILOno+g0Q6DZP7wPfLHDtpdRl8tIA5s\nNmToQkf+me46plmYolb0775Cl0+R6IfvkM2GurvQ4sU0Fgamy2PGgfHNNgCTtPsDQharhD/x\nYuekkgbBzjqMvvkaicRowbl0lwKmDSfgR59+tPv2EbOZeuUufa19By26AKXOQsjJeGvI2hfc\nvfzlPCL4+4a//3wUnXkODNqluv5gaDgcniOFQQQGesw4cHuKXGFPj+PI228ikkR/W4iStZsO\nOBwb1u+xRPRyGPBgFr8fffUFOoc552NhinrUuHu4roTPPy7yxzXZESvq60VXwHs4tS03DtxV\ncJDhuiTs1gsOqjsQJBGqFiWop8cUMepseC/5+fmrVq0SiUQsFmtlOPSzVHbJhvfP8HsRQkVF\nRXRXN/3ee++9Cbv2y+XyxBcTbytWrLDb7W9Fw5+jP5oGHTFn2UcfVxKhxHRJTKTrrrvulFNO\nmfChkpKSBBcTb6effnpu7u+zWCiEPSZXXr/wPAUZRQjNnTuX1tKmavWQtTcYHL95b1F+rVqx\natWqseYUI2z2PbWXvnzKPDZFIYS0Wi19lcaFWCxev379hA9lZWUluJh40PsDvmh0fLjupZde\n8vsnaHjCT4VreYfqnnvuueyyiefDKJVM2Frtn/2m2//oxnDBBRc0NEw8q3vGjBkJLGq/sMTs\n7kULr9cbDAYVCgWbzW71+eu27lTjeM+RswSM++KforE+dlJp6m1ZPUQQZb9t90d/j7N/Ucj+\nWx/HC7IejycUCo29o+L3WxhgKn3s9tLo9T3eb1qrS/mUQ5CUdsv28euwJXx+59yZAY+HIAil\nUjl2NvKXxpYjJOIVpcV0FpqUUqKP3YVt+geKCyqFdF4Xcrlc4XBYpVIluI9dyjmkPnZ9wdCS\n7r4N1SlzFEqXiFMtEp6RoRwiiNctw3TXAqZNNo5fm5s9fvMbh+tbh4vGesC0qxeLivn8D0ds\ndBcyVa9ahmJn191XlM/d56v3zcry1y1WD1yqS0FbPV4Zh01vqgNx8phxYElhasyuG5MuwQ4h\ndH9RAULoiYHB0P53GQcpZ0lBvihm/Oz+XiONxYB4eLC44LlBy2hCNpiOE3+UjJ1dpxHwL81W\n7/u0fB5vrkRys2Hvva1B8nuor/+B4gK6qwDTzxgMuaPR1Opfk0bBbrZEfKpS0R8MvT5kpbsW\nMG2ycO5NeTnjN39xez6z2WmsB0w7Pov1dFnx9foUjjvPD5otMU1NHyou3He4bsyrlWWbbc6B\nlNrtHnw4YpsrkeTgON2FgOn3SAr2r0mjYIcQerikEEPon0YTDNoxyZ2FefKYphj39hijzJ05\nmp5mScQzJaIXBy10F3I47OFI7HBdjUh4gTpzf0/OwfHj5dJrOrsTUhqYBiGSfG7QAltNMFJP\nIOiORhvEh7ylLL3SK9jNkohPVSkGQqHXYKYdgyg4nNijaovP//bwCI31gHhYUpD/tcO53eOl\nu5BDtrSvP3ba3MMlhawDzmt/RlPS7PP/n8sd98rAdHjGZL46J0vITq8v0zSxzDhwb6oN16F0\nC3YIoeXFhRhCjxhNARi0Y5DbCnLzebuvg9zfaxxfKguYgYWh17Tlt3X3jqTUZDtjMPSyefdA\n41yp5OyMg+xbnYPj8zOUN3f1kjDunPQGQ8QPTtcFWfsdggWpq9MfQAjVp9pwHUrDYDdLIj5d\npbQQBAzaMYmAxVpatHvm8mCIeG7QTGM9IB5UXM6zmtLL2ruI1Ik8Jze1RmOKfbKseDJdKJaX\nFPqi5KuWobjVBabHkp6+x0sn9W8KUs7S3v5UHK5DaRjsEEIPFRdgCD1mNMGgDpNcmZOli2kL\n/pjRNBQzXR0wwwyx6OrcrGv1hpRIdnd29+n9gfGbZ2coj5FNqlukgsNZpFb925zaa4EZ7xuH\nM4PLSbkJWGAytnu8Si5Hm5r9a9Ix2M2SiOdnqCwE8aI5JedigwmxMeyfJbs3FPFEow/1DdBY\nD4iT+RmqGRLxvT3J3tdmvXV0zfDIeABlY9g/D6Xz8B2FeVyMdWd33/RXBqZDiCQfNZqWFRfS\nXQiIi+XGgfuLUrV/TToGOzQ2fxmhFf0mdwR6gTLHWRnKE+Sy8ZurLMPNvgm29AGp7sa8HJyF\nPRaz1DTZbLTZV5rMsWPG1+ZmVx3K2b+Ezb4sWz0QIr5xOONQIJiqR4ymf+TlxK7HB4zxld1Z\nJRTGzttOLWka7KpFwguyMm3hyDOmQbprAdPpaU3J+Hs6SlE3d6Vw8zNwAMuKC52RyD/7kzHb\nbbTZXzEPm4ndvegkbPaDh9699trc7BBJPtQ34IW9KJJMk9fX5vefm3mQdTAgFVEIPTFgWpLK\n/WvSNNghhB4uKcRZ2NMDZisBs1iYY4ZYdFlMT//vnK4PUn83KjChJ8qKfdHofb3GpJpvt846\n+uaQdYZEZAzuHq57oLggk8s91B/FxbB/5OWUCwRLe/untUYwJRGKur2773lNKd2FgLh4d3jk\nVKVCyU3hsdj0DXYlfP4V2VneaPSJARi0Y5R/lhZLObs3GbujuxdWyTDVIyVFGVzuVZ2GcHK0\npP6Xyfy1w7mitPipmKNKqYB/U17u4f3AReqMgVBoJBz+CdraJY0V/YMXZ2Xmpux1OnAAIZJc\nZRn+R8xuRqkofYMdQmhpUYGAxXpx0NIfhA18mCML58a2PukLhh7rh1UUjHVrfu7JSvk5Le30\nLiAlSOpafbc7El2l1dzZ0xd7LvG8phQ/cEvi/cMQuq8oX8pmL+3thwuyyaDR69vl9V420Va/\ngAGeG7RcmZPFZ6V2NErt6qcoj4fflJ8bJMmlfXClg1FuysutiJmo/tSAObbrBGCY8zIzlhcX\nntva8ZvbQ0sB/cHQmS1t8xTyB4sLvrQ7Poy5+j9PKT9NpZjKDz9eLhsJhy/LzrwdVsjSLUiS\ntxp6ny+Hi7DMZA9HvrQ7Ltz/jn+pIq2DHULo7sI8FZezZsi60+ujuxYwbXAWFjsDJkSSNxpg\nFQWTzZSIP66pesZkftRoiiT2suw66+jizq4XyksXZKqCJHljzHodDob9u7xs6r/i8dLiL+xO\nAYv16ah96j8NHLZ7eow35edk43ARlpke7Ot/sPggO/6lhHQPdnIO557CfBKh+5K+LRY4JPOU\n8r/FrFn7yu5ca4UNZJlMweGs12lzcO6pTW27EnKeZg4Ri9o6W3y+TbXV5QIBQugR44AhEBx/\nwpLCvDIBf+q/qEzAz8HxkxWyZ01mmDdCly/tziBJzj/YjnAgRXX6A6Ph8CRbiCe5dA92CKEb\n83JLBfzNdgf0i2KYZzQlYvbuVRS3GnodkQiN9YAEWJyT9VZV+VMDg//o6onfrrIBkny833Rl\np+HewvxHSorGptC1+fxPxqyZUO8513OKHigqeMpkXqkpub6rO5ocK0XSioUgHu83PV1WQnch\nIF7u7jE+HNPiPqVBsEM4C3uouAAhdFePMXW2oAQHV8DjxfaFHybCS2CWUhrIwfE1VRUXqDMu\nadff32uc3ngXIMkXBi2nNrWV8Pmb6nTjG4STFLpG3x27ie0bWs00TsFWcjkLMlT/c3vOyVAt\nN8JioIQiKXR1Z/ezmhIhG74xmWmjzV4pFGimY3w9GcDbFCGELlKrj5CIt3u8a4atdNcCptNN\n+TkzJeLxm69Zhr9zumisByTM0TLpF3XVx8pkizsMV3catnq8U/yB3YHgvT3GM5rbpGz2f+ur\nF6kzYqfivGS2/F9MR5Lj5NLTVMop/sa9XJubvd46Oj9D1R8MfWmHywuJs7TPeG6mqh72hGUo\ngqSeHBi8pzCf7kKmDQQ7hBBiYeipsmKE0L29Rh/0FGAQDoa9XFHGxn7/CqYQWtzRBW0j0sc8\npfyz2qqb83PXDo/8pbFlaW//L27PIa2uaPX5nxoYPLWp7WHjwMlK+Tf1NZdmqznYHvOr+4Oh\nu2Mm6XIw9G6Vdtpewx/YGPZAccGyvv7/VJQ9NTDYEzOZD8TPZza7PRy5HPqbMNezJvPfc7Ji\nu5+muhTurTy9jpPL5meoPh61PWMyT+PMGEC7IyTim/Nznhkwj93sC4Ye7OuHuTJppVokfEZT\nQlLoV7dnk82xvG+AhaEqoVArFBTyeFk4V/rHXEwfSdrC4f5gqDMQaPMF/GS0Uig4Ram4IS/n\nANdVr9F3e2LOFu4vKoxT99oT5LJXzEOGQPClirKr9YZPa3RwcTCuOvyB502Wz2qr6C4ExIuF\nIL5xOr+oq6a7kOmEUcydh+v1eoPBoEKhYLMnlcS7A0Hd1h04xtLPnZmTTgvaI5GI3++XSpmw\nGmhC/ihZt21n9x8jHCyEfpxRe9Shr37yeDyhUGjy76i0FQ6Hg8GgRCKhu5D9ClNUpz+g9wf6\nQyErEXZGImO5TM5hZ3C5+Ty8XCDQCYWTiU2vW4av7DSM38zj8Qb+dMQkGya43W6CIJRKJWvS\ns/EGQqGrO7s31em+cThXWYbX6rSp35zh4AiCIAhCLBYf/KnTxx2Jzm9pf6uqIoU2g3e5XOFw\nWKVSYVg6vC8OH0EQ4XBYJBJd3K6/KS93jjShb614gxG73coE/Otyc1aazPf39r+m1dBdDpg2\nQjbrNa3mhF0tYycxJEKXdnQ1HtEghnyWrrgYViMS1oiEU/w5gyEitm8whqEPqivj+o1awOMd\nL5euGbZekqVu9vkfMQ7AFYZ4iFDUxe36x0qLUijVgUP1k8ut5HAYluoQzLHbywNFBUouZ7Vl\nePuUp1qDpHKcXHZtbvb4zZ5A8F7oXAimhkLo8o4uZ0wPnTNUyrnx/5K4NT/vdYvVGYncmp87\nGCLWW0fj/RvT0M2G3vPVGUdKk3fUGUxRhKKW9w0sLyk8+FNTDQS7PSi5nOXFhSRCN3b1MPYS\ndbp6oqy4NGY1+wuDlq+hcyGYgudM5v/GvIUELNZ63fSvmdgXzsKWFhU80NuPEHq+vHStdSR2\nQS6YuqcGBrNx7oVZKb+1FDiA54dGLsnKlHMYeN0Sgt3ers3NrhOLfnF73h2GjQoYRcxmv1VZ\nPv6OHxtusYWhZTE4HG0+/z17Dvo+qykRJGrv8BMVMlc0+ovbw8Wwt6sqHurrb4fdkKfJWutI\nVyAIF7iZrT9E/J/XewlDFztDsNsbG8Oe1ZQghO7o7nNHoC8Goxwlk95SkDt+0xwirtUbDvB8\nACYUIsmL2vUBkhy/p0oouDrmWn8CPFNWcn+vMUxREjZ7rU77D323EXYbm7JvHM4PR2wvlpce\n/KkglS3p63+sII+pC0wg2E3gBLns3EzVEEGsGDDRXQuYZo+VFNXGTJl/f8S2yjJMYz0gFd3d\nY4zdjpaFsPerKxNcg4rLuVCd+ZzJjBDK5HJf1Wou6+gaJuK1i1o6+J/L/YzJ/HZVBQeWlDLa\nOutorVCo5fPoLiReINhN7KmyEgGL9fSAWQ8XOJiFx2Kt1Wlje5LdbOiBy1hg8r6wO1aazLH3\n/D0nSzflBbaHYXFO1rdOV18whBAqE/BfKC+9sL3TDrMLDss2j3dZ38DaKu00bgQHkpAjEnnF\nPHRHXkLH1xMM3sETK+bz7irMD5HkDV09dNcCplm1SPhY6e7Nnv1R8vy2ztjLagDsjzlEXNbR\nFbuySsHhrCijZ+9wDKFnykpuNvy+0qtGJHyytHhhWwdku0O1w+O9q6dvnU7LpO0HwITu7O5b\nXlLIY3R8Z/Jrm6K7C/MqhIKvHc4PRmx01wKm2S35uaepFOM3m7y+myHBg4OJUNSF7Xrrnpc7\nny8vpXFhnVYo+JNUsnro902uZ0rEK0qLF7Z1wKqgydvu8S7p6dugq1RyGbhAEsT61uESslhH\nH3p3+tQCwW6/eCzW85pShNAthh7YXZRhMITe0JbH7i/yqmV4DayDBgf0YF//D05X7D1HSaW0\nN8W4oyDvnWHrYIgYuzlLIn6qrGRhWwfMt5uMX9yee3qM71VXqiDVMZ0vGn3EOPBoKT3j64kE\nwe5A5inl8zNUphDxiBFWUTCNGueuqapgx8ySvk7f3erz01gSSGaf2xyPxxwHMITYGPbvilLa\np9lzMOzpspLbunvH72kQi57TlC5q6+yDdbIH9KXd+XDfwPvVlQomNjMDe7m313hXYb4kDTYc\ngmB3ECs1JSI2+5mBQfjKZ54TFbIHi3d3q/JGowtaO1zQ4wbsozsQvKRdHzsNk0Lo1vzcOrGI\ntppi1ItFlULBWuvuIedqkfCNSs3lHV1NMat3Qay11pFXLUMf1FTCvLp08IPT5Y+SJyvldBeS\nCBDsDqKQz1talB+mqGv03SRsRsE49xUWzIv5qOv9gcv3nBoPgD9Kntva4YjsMWutgMeLPSug\n3dKigtctVtMfF2QRQiV8/nvV2jt7+r6BHVb28eTA4A9O9zqdNmE9pQGNfNHog30DT5YV011I\ngsB7+uBuL8hrEIv+53KvsgzRXQuYZiwMra3SlvB3bzX28ajtEeMAjSWBpEIh9PdOQ2PMuNfY\nerrny0vFyXRNh4NhKzUltxr2WAOUyeV+VF31imUYmjWOGztLj1DUSxVl0K8uTdzZ3Xd3YR4j\ndw+bEAS7g+Ng2MsVGjaGLenpM8ecEANmUHI566u1savfH+rt/3gUlkIDhBBa0W+KvcQpZLNC\nJLkgU3V2hpLGqiakEwlniMXjK2THCNmstVVaQyB4i6E3QqX7YPRIODy/pf0khfyewny6awEJ\nMrah8ylKxUGfyRgQ7CZljlR8TW62KxK9PWaGMmCM2RLxCzGbCJEIXdre1QyzKtPeRpv9vpgN\nYTkYFiRJCZu9UpOkW04tKcxbax3pDQZj72Rh6PHSojkS8fyWdmsaL5Xd5vGe19r5aEnR3zJV\ndNcCEsQejjxiND2RNhdhx0Cwm6zHSopyefg66+gmm4PuWsD0+3tO1j/ycsZveqLRM5vboGFE\nOmvy+i5s22PBRA6OkxR6tLQon4fv93+jFQfD/lNedr2+J7rP4Nwkg4JaAAAgAElEQVSFWZmP\nlhQtbOv8n8tNS230esk89KjR9EF1ZUNyrHcBiXGToeefpUVJNWsiASDYTZaMw35OU4oQulbf\n7YaFk0z0rKbkRIVs/KYxGDqjuc0fhR0p0tEQQZzZ0u6JaWA5SyIaCIXmSiXX5yb1ZkSlAv6C\nTNUTA4P7PlQvFn1SU/Xi4NBjRtO+yY+pnJHIhW16ezjyATSrSzNvD1vLBPw/SSV0F5JoEOwO\nwbmZqgWZqoFQ6O6ePrprAdOPg2HrddpSwe6FFNs83ss7umA1dLrxRaNnNbf3xzSBmy0Vt3oD\nOAtbpdWwk37G/VU5WTs83u0e774PyTjsd3UVmTj3zOb2dOhy953TNb+l44a8nHuL8lnJ/u8G\nplNfMPT20Mj9RUm0dD1hINgdmhfLS5VczkvmIeggwEgZXO7GWl3s4qn3RkaXQI5PJ1GKOr9N\nvzUmFWkEfDbCghR5X2FBjUhIY22T9++Kstu7+/a3Zc5VOVkry0uu0RteMQ8x9bTFHyVvNvS8\nMzzyaU3Vn2VpN2aT5iIUda3e8GJFKTfpT8PiAYLdocnG8afKSiiErurs9sE+Y0xUJRRsqNbG\n9kF4emDwOZOFxpJAIv2jq2ejzT5+U8HhLFJn/ur21IqEd6fOUspMLveh4oKbDftd7FUuEGyq\n1bmi0TOb27oCgUTWlgBf2Z2nNbedrFCs0mqg/3AaWtY3cIE6s1wgoLsQekCwO2SXZ6vnKeW9\nweBDfdDtjJlOUshfqiiLvecWQ89be3aRAIz0QG//y+bd7SpxFvZ8eelKk5mNYa9VluMpdTHv\neLksk8t9Z/87ILMx7M6CvJWa0ju6+x7q6w+QTJhOOhgiLmnXf263f1ZbdZoqjTpcgHE/Ot3d\ngeBl2Wq6C6ENBLtDhiH0coVGwmb/y2ROz/Vl6eDKnKwHYvYVoBC6Sm/41j3BpCXAGC8OWh6O\n6U2NIfRGZflrlmFvNHpbfu5siZjG2g7PwyWFbw5ZuwPBAzynTMD/pKaqTiQ6taltvXU0da/M\nBkjyUaPpqk7D7QV5KzWl6bAlKNjXaDi8tM+415l5uoFgdziK+bynyoqjFHVpR9f+ZrGAVPdQ\nceEVMed8BEld1N231X+g70iQut4ZHrmpa49tG1aUFdvCke+crkqhYFlJIV2FTQUXw1ZpNVd1\nGg46GrcgU/VFna43GDylqfU7pysx5U2XCEW9Zhk+ralNKxRsqtNBQ5O0RSF0vb7nmbKSNL/+\nDsHuMF2Vm32qUtETCN7Xazz4s0EKwhB6Ras5U7V7gwEfSV5gNO2EXdUZ56NR2+UdXbHZ5/aC\nvHMzVPf2GDkY9mZlReruKFrI591WkHtHd99Bn8lnse4uzH+3SrvJ5jijue2HVIh3EYp6c8j6\n18bWEEl+VV8NnYfT3D+NpuPl0lkpOLg+vVL1aEU7DKFVWo2Cw3nBZPk+FY6A4DBwMGx9tfZo\nmXT8Hlc0empLRwtsSsEgm+2O89s6Y7fbujRbvaK0+IrOLm80eldh3hxpan9PnKFS8lnYmv1P\ntoul4nKeLCt+TVu+0eY4uan1wxFbcna880Sjz5ksf21s9USjX9Tprs/LSc/1j2Dcj053i89/\nfUyf+bQFwe7w5fLwZzQlJEKLOwzQspipBCzWZ7VVM2Iu7oyGw39pbGmDbMcIX9qdC1o6iJhe\nhfMzVK9pNc+aBn90umtFwgeKUvIi7F4eLy1eM2xt90929WsWzn2yrHi9TmsIBP/S2PpE/2Dy\n7EXW5vPfbOg5t6UjE+f8t776hrwcfsqOp4LpMkQQD/T1p/nUunEYlZRnY9PC6/UGg0GFQsGO\n5yzas1vaPx21X56tfqOyPH6/Ja4ikYjf75dKpQd/aroaCYeP37VHmMvCud/W1+hSpKtZgoXD\n4WAwKJEke/Owr+zOs1vagzHzz05Wyj+pqer0B+bsaKIQ9evM+hnxnLDldrsJglAqlaz4R5Mh\ngrigTf9JTdWhTj+KUtTnNsfqISuXhV2kzjxZKeclPEgRBDHsD2z2Bz4etefx8Ktzs1NxLUsC\nuFyucDisUqmwdBq/jFDU2S3tK0qLJ99mkiCIcDgsEjFzOiYEu6kaCYfrtu4aIoj1Ou1CdUb8\nflH8QLCbDHOIOHZHU3dod7N+Nc79uq66DmZq7yMlgt0mm+Pc1o7YVHecXLapVsfG0JwdTU1e\n3+OlRXfFuXFdIoMdQuhnl+dZk3l9tfbwvvOHCGKtdfRLuyMbx89UKecp5QlYeWoOERtt9s9H\n7YgiF2Vnzc9QCdkwPrdf6Rnsbjb0HCERX5J1CP1NINilqsQEO4TQF3bHaU1tcg6ncXZDAY8X\n198VDxDsJqnDZj+t09Abc01KxeV8WVcNc3X3kvzB7uNR26K2ztgrsMfIpJvqdGI2+zZD779M\n5qNl0u8bauK9e1iCgx1C6JkBcxRRdxbkTeWH9AdDG232rxxOgqSOlEqOlUtnS8Si6TvMjoTD\n/3O5f3S6m33+HBw/XaU4USwSkFGxGD5oB5GGwW6tdWSr2/uMpuSQ/i8IdqkqYcEOIXStvvtl\n89BJCvmX9dUp93mCYDdJHo+n1+tb0D8Y2xhMxmF/WqM7Vg5/vd2SPNi9PWxd3GGIXS1xtEy6\nuU4nZrN/dLpP2NUs4bB3HTGjmB/3k7TEBzsKoUvb9Zdkqecp5VP/aSGS3OLx/uh0b/N4/WQ0\nG8drRcJKoUAjEBTzeZOc+uaMRPqCIb0/0OEPNPt87mg0k8s9Uio5Ti6rEQnHDqcEQRAEAcHu\noNIt2O30+pb2Gj+uqeIc4uuFYJeqEhnsfNHojG2NXYHASk3pTfkptioHgt0keTyeUCjkF4rm\ntbR3xsxDF7BYG6q1Z8Q0RklzyRzsnjNZbjH0xB71TpDLPq2tErPZ9nCkfttOU4h4s7L80oS0\nrU98sEMIBUhyfkv7C+Wl077h0hBBNPv8en/AEAgag6EQRUYpio1hGEJC1h4HYU80ihBiYShK\nUSout5DHKxfwK4XCGpFwwimAEOwmKa2CnS0cOa+1Y321NpPLPdT/l9nBjnPwp4BJELHZa6oq\njt7ZtKSn7zi5tB7mXTFXHg//vqHmr42trX+spQiQ5IKWjle0msvTeBOb5EchtLTX+KjRFHvn\nyUr5RzVVY23qruw0mELEInVGYlIdXQQs1hva8ks69B9WV8mmtY9rNo5n4/hJignGAj3R6PgQ\nKZ/FSt2+gCBJhCnqknb9k2XFh5HqGA8+XdNmjlS8vKQwRJLntXZ6YDsKRsvG8e8bao6ImVoX\npqjFHV2PGGH74CQVpqgrOrr2SnXnZKg++SPVvTho+XjUVsDj/buc+R0Tcnn4suLCxZ1dCetR\nJ2GzFRzO2H+Q6sDU3djVszgnC+Y3Twg+YNNpSUH+SQp5VyBwm6GX7lpAfGVwud821Jwgl43f\nQyG0tLf/qk5DmLnTG1KUKxI9vantzSFr7J1XZKvfq9aOde5o9fnv7O7jYNg6nVbJTYvrGEfL\npGeqlHf1wMY5IPU8M2DOxXHYaGR/INhNJxaG3q6qyMbxVZbhtdZJ9XkHqUvCZm+q0y3Y8+Cy\nyjJ8alOrIxKhqyqwl95g8M87m752OGPvXFKY91pl+dii1yBJXtiuD5DkQ8WFf5Yl47zAOLk8\nW81C6D/mIboLAeAQbLY7tnm8S4sL6C4keUGwm2ZZOPc1rQZD6Dp9d+zaScBIfBbrPV3ljXtu\nYvONw/XnHU36SXf5B/Hzk8s9d3tTbGdpFkLPlZeuKC0en1t+Y1dPk9d3glx2T5y71iWhx0uL\nv3O4vtkz9QKQtBq9vpUm8+uVGuavDZkCCHbT7zSV4taCXFckuqitM0SSB/8fQCpjYei58tIn\ny4pjP0sd/sDcHU1f2B20lQUQetk89NfGlpHw7r6DQjbrveo9gvia4ZFVluFMLvftqgpW+n1X\nsDD0RqXm8f7BVtgiDyS9wRBxs6H3rcoK2ETuwOCvExcrSouPkkm3e7y3dffRXQtIhDsK8t6v\nqYzt0eqMRM5obn+83wQT7hIvRJLX6Luv1XfHtiDOwfHvG2pjL53r/YHr9d0shN6uqsjj4XRU\nSj8Rm71WV3FDV89AzK4qACQbTzR6WYf+5YoyNQ7LYA8Cgl1ccDBsra4ig8v996DlnWGYbJcW\nzslQ/dBQE5sPohR1T49xfku7E6bcJVBfMHT0zuZX9pw61iAW/TarLnaD0QBJLmzr9ESj9xUV\nnDwd3XpTVwaXu0qrubyjC96oIDmFKeridv3y4iKtcJqbLzISBLt4KeDxVleWj022g+lWaWKW\nRLx1Vv2fpHtMwP901D5re+NWj5euqtLKJ6P2Wdt3bdvzr31eZsb/zajda7u/6/TdjV7fCXLZ\ngzALG6EyAX95ceGl7V2xY5wAJAMKoas6DZdlqdNqbdNUQLCLo9NViiWF+Z5o9G+tHT7obJce\ncnD8u4aaK3OyYu/sCQSP3tn0zIAZvjPjJ0SSN3X1nNPSbg/vHnZiY9ijJUXrq7V77WT6inno\nzSFrNo6/q6uI94awqeIomfSq3KyL2jsT1twOgMm4o7t3pli8AJqbTBoEu/h6tKRwnlLe7PP/\nvdNAdy0gQXgs1iqt5s3KciF79+eLIKnbu3vnNbYOhggaa2OqNp//Tzuanh+0xEYSFZfzea3u\n3qL8vYLbFrf3JkMvB8PW6Sqy8TSdWjehM1XKszNUN3b10F0IAL97zGiSczgpt1EnvSDYxRcb\nw96t0hbzeeusoytNZrrLAYlzabb6p4baMgE/9s7/Opz123Z+MGKjqyrmISn0L5N51vbGnV5f\n7P1/kkp2HtGw7+Q5WziyqK0jRJJPlhUfF9NfGoy5OCuzmM9/sK+f7kIAQK9ahocIYmkRTJY4\nNBDs4k7F5azXVfJYrDu6+35wuuguByTOTIl45xENF2Zlxt5pC0f+1tqxsLUztg0HODw9geBf\nGltuM/QGY/oKYQjdlJ/zfcPek+oQQlGKurC9sy8YOl+dcUt+bmKLTRlLCvNCJPX0wCDdhYC0\nttY68j+X+1lNKd2FpB4IdokwRyp+VlMSoajz2/QmuBKXTiRs9jtVFau0GvGec7zeGxmt3bpz\nvXWUrsJSXZSinhkw127b+f2eJ0s5OP5FXfVKTSk+UVe6u3uMX9md1SLhq1pNoipNSY+XFlmI\nMGxKAejy6aj9oxH7Kq0mDbtLTh0EuwS5Njf7imz1EEGc09IegK7FaebKnKydRzQcuedq2WEi\nfH5b5+nNbX1B6B92aHZ4vHN2NN3e3euP7vFRWpCpaprdMG8/vUvWDI88NTAo47A/rK7cK2eD\nfT1ZVrzN43172HrwpwIwrb6yO98Ysq6pquDAwqbDAsEucf5TUXakVLLN470KFlKkH42A/9OM\n2kdKivYaSdpkc1Rv3fGo0QSblEyGIxK5oatnzo6mHXs2NJFzOKsryz+orszgTty8dIfHe43e\nwEJoTVVFBbTCmgQMoVcqyr60O2Hba5BI3ziczw9a3tVVTDjoDiYDgl3i8Fisj2oq83n4O8Mj\nT8L8lfTDwbD7ivK3z2qYFdMmFyHkj5L39xprtu78ZNROV23JL0pRr5iHtL/teHHQslc/jtNV\nipbZMy7LVu/v/x0iiLNb2v1R8rHSojNUyvgXyxBsDHuzsnyzzQld1kFi/ORyPzVgXqurEMCm\nYVMAf7uEysbx96sr+SzWPT1G2Eg0PdWIhL/OrHuyrHivzmqGQHB+S/uJu1r2Wt0JEEJf2Z0z\ntu26Rt+914qTLJz7rq5iY63uABuChUjy3NYOU4i4QJ25pDA//sUyChvDXq/UfG5zbID5oCDO\nvne6Hu83vV+thZkSUwTBLtHmSiUvVZRFKer8ts422Hg7LXEw7I6CvJbZM05XKfZ66Dun64ht\nuy5s0xsCQVpqSzZb3N6/Nrac3NTavOeHhYXQVTlZ7XNmXqDO3N//i37vWd/9s8szUyJepdXA\npZ3DwMGwN6vKP7HZ4ZosiJ9vHM6nBgY36PbuJQ4OAwQ7GlyWrb6zIM8ViZ7e3GYloOdFmirm\n8zbW6j6trSrds9cdidBa64huy46rOw3pvK5ip9c3v6X9yB2N3zj2bhI0WyL+ZWb9K1qNgsM5\n8A951Djw9rA1B8c/rqmM7RcNDgkXw96qLP+vw/W6ZZjuWgADbbY7nh4wr4NUN03gSEePx0uL\nz8lQ9QVD57S2w6z5dHamStk6e8ZjpUWSPY9oYYp61TJc8dv2KzsNnWm21/Cvbs/ZLe2ztu36\nZNS+1+ZWOTj+mlbz68z6OVLxxP9zjA9GbA/29gtYrI9rqvbtaQcOCRvDXq3Q/ObxvgQ9UMC0\n+njU9ppl+KMaWKs+bSDY0YOFoXd0FXOk4p9dnks7umBrxnTGZ7HuKczvmjvr6tzsvZb3hynq\ndcuwbsuOc1s7fnZ56KowMUgKfWazH7er+U87mj7dJ9KJ2eylRQX6uTMX52RNZrXcdo/30g49\nhdAbleWTSYHgoFgYeqmirN3vX9FvorsWwBBvDVk3WG1rdVoerJaYPvCnpI2AxfqwuiqPh2+w\njj4EG/ikvSyc+3JFWevsGYvUGXvlFhKhD0dsR+1smr298e1hK/OGeF2R6EqTWbtl+1nN7T86\n3Xs9irOwG/NyDHNnLS8pnOQ5fW8weHpzmz9K/rO0aJE6Iw4lpykMoZWaUoTQzYYeOB0FU/Ti\noOU7p+utqnIu9KubVhDs6JTHwz+tqRKz2cv7Bl6DySsAoQqhYJ1Ou+uIhnMzVft+OLd5vJe2\nd+X/su327t52Rlyf/cXtWdzRlfvLllsMvfuuF8FZ2LW52V1zZj1XXpqFT9ygbl+2cOTUprZh\nIrw4J+suWAYbB3cV5pfxBdd0GvbqOwPA5D3Q298fCr1eWQ5diKcdRjH3k+n1eoPBoEKhYCf3\nlfvNdsdZze0UQh9WV56VQUOTrUgk4vf7pVJp4n91avF4PKFQKGHvqGaf//F+0wbraGQ/H9I5\nUvElWepF6ozM/XTlpUs4HA4GgxKJZH9P6A0G3xkeeXt4RL+feCpis6/MUd9RkHeoc+OCJHlS\nY+v/udwnKmSba6uTvMep2+0mCEKpVLJS8DrUmuGRT0ftqyvLE7AqhSAIgiDEYrikfhAulysc\nDqtUKiyJ01KEom7o6tEI+HcU5NFVA0EQ4XBYJBLRVUBcQbBLCq9Zhv/eaRCyWd/W18yV7vfr\nME4g2E1SgoPdmL5g6F+mwTcsVk80OuET2Bh2vFy6MDPjzAxlDr7fdm6JtL9gZwgEPx61bbCO\nbt1z34hY2Th+XW72P/JyVNyDrHjdF0mhRW0d74/YakXCn2bUyTjJ/sFP6WCHEPrW4Xq837Sm\nqkI96fHUwwPBbpKSP9j5otHLOroWZKguzDpQo6J4g2CXqlIo2CGEHujtf9g4kIPjP8+sK+Yn\ndAUfBLtJoiXYjXFHom8MDf/bPLS/IS6EEAuhI6TiM1XKkxTyIyRiNn1H9thgFyLJ/7k8Xzoc\nG22OAzdunCuV3JCXszAz47CH2W419D5rMhfxeT/PqMvdf8vi5JHqwQ4h1Oj13WLoXaXVlO3Z\ntWd6QbCbpCQPdlYifHG7/p6i/BPkMnorgWCXqlIr2FEILe7oWj1kLRcI/m9GbbzPgGNBsJsk\nGoPdGAqhH5yuV8zDH43aggdcQqHkco6RSY+Ty46WSRvEogTPTXYFQ7/YHdvC4R+d7v+53f7o\ngUqVczgXZmVenZNVL57SQfbxftM9PUYll/NTQ61OJJzKj0oYBgQ7hFB/MHRFZ9ey4sKjZfE6\nhkCwm6RkDnatPv8/unqeLy+tTYKPJ7OD3SFf7ABxgiG0SqtxR6MfjthOamr9oaFGfrDmqyDd\nYAgdL5cdL5c5I5EN1tG3h0f+53JPeGZmD0c+GbWPbT7LZ7FmiEWzJOI6saheJKoUCqTTfY1y\nNBxu9fmbff5Gr2+rx9vq8+9vXuA4LobNU8ovyVKfnaHkTznWvDVkvbfHKGCxPqmpSpVUxxiF\nfN4nNVVXdHT1BIKX7n/HXpDOvnY4nxoYXKeryE6O6SLMBiN2ySVAkic3tv7kch8vl22u0039\nC28yYMRukmgfsdvXQCj0/ojt/ZHRX12eyTdBycHxciG/iMcv4vPyeXg2jqtxroLDkXHYMjZn\nwrnwnmjUHYk6IxFHJDJMhC0EYQoRxmCwLxjSBwL2cGSSv5rHYh0vl56XmXFOhkp56LPoJvTp\nqP3c1g4MoY9rqk7bZ5e2ZMaMEbsxUYq61dAr43CWFRdO+5IVGLGbpOQcsXtx0PI/l+e1So0g\nad7nzB6xg2CXdJyRyLE7m5t9/gWZqvU6bQKWgkOwm6QkDHbjrER4o82+2e74xuFyRCYbs/aH\njWHSP16jMxKZ+jEil4fPU8hPUylOUSok0/rX+87pOq2pLUSSb1VVXEzrdOzDwKRgN+YV89Bm\nu3N1Zfn0rlyBYDdJyRbsCJK6oatbyeU+VlKUVCvUmR3s4GJf0pFzOF/UVR+9s/nDEdviDsPq\nyvKk+jyA5KTGuYtzshbnZEUpapvH+53TNTa/zR2ZeC3tgUUpaurpUI1zj5FJj5XJTlTIauJz\nefQXt+es5vYgSf5LU5JyqY6Rrs7NrhYJz2ppe7lCUykU0F0OoNNIOLy4w3BFtnpBporuWtIL\nBLtklMvDv2+oOWZX89vDVi4LW6XVQLQDk8TGsLlSyVyp5O5CFKWodn9gi9uz1eNt8vmavf79\n9UyZFioup14sqheJZggFdTi3PiO+R/Mmr++M5jZvNPpQceEt+blx/V1g8o6SSddUVSzuMNyU\nn3OmiobGnCAZ/M/lvq+3/8Xy0mqY85pwEOySVCGf93Vd9XG7Wl63DEvY7Gc1JXRXBFIPG8Nq\nRMIakXBxThZCiEKoPxjSBwJ6f6A7EDSGQgPBkJkgRsJhgjyEy60CFkuNc/N4eCGPV8jnaQSC\nCgG/Uigc3xxirN1JXF7SH/T+wLymVns4cnN+7oPFBXH9XeBQFfB4G2ur7uoxfuNwPVVWDFsL\npJtXzENfOpwf1VQqYAkgHeCPnrwqhILNdboTd7WsNJmlbPbykkK6KwKpDUOoiM8r4vNOUsj3\nesgZidjDEWck4opGoxQVIMnxdipCFpvHwnCMJeOwZRyOisuZ3klyh8EQCJ7Q2DJMhK/Lzf4X\nnPMkJR6L9aymZPWQ9ZyWjlVazeR3hAMpzRONXqvvrheJ3q+uhDhPFwh2Sa1BLNpUp5vX2Pqw\ncYCDYQ/AyASIDzmHkyrtdboDwRN2NZtDxKXZ6hfKy+DLI5ldnq2eKRad39Z5X1H+X/c5nQAM\ns93jvaO776HiguPo7j+c5hiyFIvBjpRKvqirlrDZD/b1P2wcoLscAOhkDIb+2thiChEL1Rmv\naTWwrij51YlFm+p0n9sctxh6w8xtwpDmKIRWmswP9Q2s02kh1dEOgl0K+LNMsrlOJ2GzH+jt\nfwSyHUhX/cHQCbta+oKh8zIz3qmqgJlbqULAYv1LUzJbIj6juc0QiO/kS5B4QwSxoKWDQujT\n2iq45p4MINilhqNk0s9qq0Rs9tLe/keNJrrLASDRegLBY3c19waDi9QZ7+og1aWei7IyX6nQ\n3NjVs9JkprsWMG022x0XtesfKC64JT8XPpNJAoJdyjhOLhvLdvf3Gpf29tNdDgCJ0+kPHLer\n2RgMLVRnrIGxupRVxOdtrK3yRcnzWjuGCILucsCUuCLRv3cavrY7P6/VzZjaRs9gekGwSyUn\nyGVf1OmkHPYjxoE7u/tgugpIB+3+wImNLaYQcb4arsCmPDaG3VuU/0Bx4cXt+lfMQ3SXAw7T\nl3bnWS1tF6ozn9GUJGbrSzB58O+RYo6WSb+tr1FxOU8NDF6v7z6U7mMApJ5dXt9xO5vNIeLK\nnKx3qhKxwx5IgFqRcHNdtZkgFrZ2WmDoLqU4I5GrOg0bbfZNtboTFbBOIhlBsEs9syTir+qq\nVVzOS+ahq/SGKCw0Awz1s8tz4q6WkXD4+rycV2ENLLNwMeyh4sL7ivIvbe96yTwE56gpYYN1\ndH5Lx8VZmc+Xl4ro7mcJ9geCXUqaKRF/11Cbg+OvW4YXtnWG/uglCwBjfGF3zGtqdUQidxTk\nvVBeCqGOkerFoi/qdCGSPK25tdXnp7scsF/GYGhBS0er3/9lnQ4amiS51GhJCvZVKxL+PLPu\npMaWD0dsp4TbPqmpknLg/AkwxEejtgva9CGSvKsw//HSIrrLAXHExrCb83MXZKru7O7L4+EP\nFhXCoSypBEnyqYHBX92eJ8tKqoQCussBBwcjdimsmM/7cUZtrUj4vdP118aW0XCY7ooAmAb/\nMQ/9raUjQlEvV5RBqksTBTzeOp32ZIXi7Jb2NcMjcGE2SWy02U9paisXCDbW6iDVpQoIdqkt\nB8d/mFH7J6lkq8d71M7mHmj+CVIZhdDS3v7r9d0cFvZuVcXVudl0VwQSap5S/mVdtTlEnNLU\n+rPLQ3c5aa3F5z+7pf1nl2djbdUidQbd5YBDAMEu5Sk4nK/rq09XKfT+wJ93Nm33eOmuCIDD\nEaaoxR1djxgHJGz2xlrdQvguSUs4C1tSmLemqmLNsPWidj2crCbeEEFcq+9+vN/0nKb0sdIi\nMSySSDUwx44JRGz2JzVVN3T1vGQeOm5Xy3qd9nSVgu6iADgEvmh0YVvnJpsjG8c/r62aKRHT\nXRGgUyaX+++Ksjaf/66evmwcX5KbBUe0BHBGIk8NDO70+pYWFRwpldBdDjhMMGLHEGwM+09F\n2UPFhf5odH5L+yrLMN0VATBZ5hBx7K6WTTZHlVDwy8w6SHVgjE4kfK+68gJ15tWG3mWDFns4\nQndFjOWNRlf0m/7W2nG0TPp5rQ5SXUqDYMcoDxYXrNJqMISu6jQs6e6D1lAg+e3y+ubuaNzh\n8R4tk/7fjLpiPo/uikBy+bNM8klVxVFi8QXtnff1Gm0Q76bVWKQ7q7m9VMD/ur7mFCWMjaY8\nuBTLNItzsor4vL+1dj45MNji86+v1kpghgRIVl/YHYvaOlrCiOAAACAASURBVN2R6EJ1xurK\ncgHsTQT2468yyfy8nK/szovb9TUi4a35ubk8nO6iUps9HHnRbPne6bouN+e/9fnQAJwx4DDK\nQH9RyH+eWVcq4G+2O07Y1TIYgh17QDJ6emDwjOZ2TyS6rLhwnU4LqQ4c1DylfHOd7kyV8oau\nnr93Gtr9AborSknGYOgWQ+/F7fqZYvF/62v+lqmCVMckMGLHTFVCwW8z685p6fg/l3v29sb3\nqyv/LIM5EyBZBEnyGn33W0NWPov1RlXF+bAAFhyKY+XSY+XSFp//yX6TPRK5Jjf7FKUCkslk\n/ORyvzhoQQjdkp8LE+mYCoIdY2Vwuf+tr75W3716yHpiY8sL5aV/z8miuygA0GCIWNDavsXt\nzcHxD2sq4dsFHJ4akfD1ynIrEX7ZMrSif3B+hvLSLLWSC19qE/BGo+uso+usI7Mk4hWlxUUw\nk5XR4DPAZDwW643K8j9JJTd09VzVadjq8b5QXsrF4MwW0OYXt+fclg4LQcyUiD+qriyELxgw\nNWqcu7So4K6C/E9stss6upRczt9zso6WSeEwN2aHx/uKZbgvGDxfnflpjU7IhgkPzAfBjvmu\nzs3WCAQL2zpeMQ+1+/zrdFqYdAxo8e9By63dvQRJXZ6t/k9FGR8m1YFpgrOw8zIzzsvM6A4E\nXx8afrC3/zi57JKszFIBn+7S6GEOEe9aR76wO2pEon/k5dSKhHRXBBIHgl1aOFEh2zqrfkFL\nx08u96ztjet0FcfJZXQXBdKILxq9Rt/9zvAIF8NWakpvys+huyLATGUC/qMlRSSFfnC5HjEO\nWIjwKUr53zIz8tLjbNYWiXxsd35nGZGw2RdmZdySn8uBSzTpB4Jduijh83+ZWXdTV8+rluG/\nNLbeX5T/QFEhrIQCCdAVCJzb0tHs8+fx8PU67VEyKd0VAYZjYegEuewEuSxIkl/anXf39NnC\nkRMVsrNUygom7mTfHwxttNm/sDtZ0chpEtG7VeUSDny5py/4t08jfBbrFa1mtlRyU1fPsr6B\nXV7f69pymGsM4mqddfQavcEdiZ4gl63VabNwLt0VgTTCZ7HOzlCenaEMkuR3Tte/TOauQKBB\nLJ6nkB8jl6Z0h50wRf3i8nzlcGxxe3N5+Jkq5bu6iqjXGw6HYXfXNAdf6mnnqpysmWLRea2d\nn4zaZ3h3vVtVMRemX4A48EfJmw09qyzDGEJ3F+Y/UlLIhqtCgCZ8FutUpeJUpYJCaJfX95Xd\n8dygmYWwo2SSY2Sy2VJxSqwqi1JUo9f3o8v9k8vti0b/LJWeoVIuK979yXLRWx9IDhhFMXbb\nKa/XGwwGFQoFG05f9uGORK/Vd6+1jrAx7N6C3FtVCoUMZt0dhMfjCYVC8I46qHA43Oh0Le4b\naPb5M7nc1ZXlp6lgn6IJuN1ugiCUSiUrlYeOEoAgCIIgxOJp3kTYHyV/dXt+dLm2ebwUQnUi\n0RypeI5EklQT8kbC4S1u7xaPZ6fHF6aoBrHoGJn0aJlUypngKORyucLhsEqlwlIhp9KIIIhw\nOCwSieguJC4g2KW1l81Dtxp6AyR5rET0To0uP5kOZ0kIgt1kUAi90G+6q28gQJInKmRrqipy\ncHhfTQyC3STFKdjFilBUs8//m9uzxe0ZJAgOhlUJhbUiYZVQWCHkyxM1Zc0bjer9gc5AoMnr\nb/P7gySZyeXOkUjmSMUzxCLewd4nEOwmCYJdqoJgNxnNPv+i1o52f0DB4fynomwR7AGwfxDs\nDmqIIBZ3GDbbHRwMe7C44N7CAligcwAQ7CYpAcFuL2GK6vAHWny+dl9AHwg4IxEMYRlcTiGf\nV8jj5fLwTC43F8fVOPcwuvYQJGUNh4cIYpgIDxFEfyjUFwxZiTCGkJDNqhAItEJBrUioEwkP\n9YdDsJskZgc7mGOX7mpFwi0NNUu7+54bHjm/rfODEdtLFWWwogIchs12x+IOwxBBFPHwV4oL\n5uVk010RAIeJi2G1IuFe7d/s4Uh/KNQfDPUHQ9s93mEiPEyEgyQ59miEongsVoSiMITGly/4\nyWiYovgslj9Kjk/jw1lYJpebg+OZODcPx+vFokIeTw3risA0ge9vgPgs1rK87JOzMhd3GN4b\nGf3N41ml1ZykkNNdF0gZzkjkVkPv6iErQuiqnKwVRfmccJjuogCYZkouR8nlNIiZOcwDGAPG\n/8HvTlEqmmY3nJup6g+GTm5svbrT4I5E6S4KpIBNNkfN1p2rh6xqnPtxTdUrWg10WwAAALpA\nsAO7ZXC571dXrtVpVVzuq5bh2m07v7I76S4KJC9nJHJFR9fpzW2DIWKROqN19oyzM5R0FwUA\nAGkNgh3Y2/nqjJbZMxZkqvqDoVOaWi/r6BqFy2pgHxuso1Vbdq4esmbh3PerK9fptBlcmCQE\nAAA0g2AHJpCFcz+orlyr02bi3LeGrFVbdr45ZGXs8mlwiIzB0BnNbYvaOocI4sKszJbZM87N\nVNFdFAAAAIQg2IEDOF+d0T575hXZals4fHlH10mNLXp/gO6iAJ0iFPXMgLlm687PbY4SPn9z\nne6dqgoYqAMAgOQBwQ4ciJLLeb2y/NuGmgqh4BuHq3bbznt6jL4oLKpIR987XTO27bq9uzdI\nkncU5LXMnnGKEvaTAACA5ALBDhzc8XJZ4xENy4oL2Qh7vN9UuWXHBuso3UWBxBkMERe0dZ6w\nq6XF5z9KJt06q/7JsmIhG44eAACQdODQDCaFz2I9UFzQNmfm/AyVKUQsaus8cVfLTq+P7rpA\nfAVI8lGjqXLLjnXW0Wwcf7Oy/KcZtdDHCwAAkhYEO3AIivm8j2oqN9fpKoSC75yuI7btWtzR\nZQ4RdNcFph+F0DvDI9rfdtzfawyR5C35uR1zZl6arYaNigAAIJlBsAOH7BSlomX2jGc1JXIu\n540ha8WWHcv6Bjww8Y5BfnC6jtzReHG7fiAUOlOlbJ4941+aEhkH2g4DAECyg2AHDgcXw27O\nzzXMnXVrfm6YIh/q6y/7dftKkzn0x7aJIEXt9PpObWo7flfLFre3QSz6pr7m09oqrVBAd10A\nAAAmBYIdOHwKDucZTUnr7JnnqzNGw+FbDL3aLTtWD1mjFPS8Sz16f+D8ts5Z23Z9YXcU8Xmr\nK8u3z2o4USGjuy4AAACHAIIdmCqNgL9Wp91+RMOpSoUxGLqio2usoXEE4l2K6PQHLm3v0m3d\nud46molzn9WUdM6ZeVm2mgXz6QAAINVAsAPTY4ZYtKlO90ND7bFyaVcgcHlHV9WWHash3iW3\nDn/g4nZ99dadbw9bpWz28pLC7rmzbs7P5bHgyAAAACmJQ3cBgFGOlUt/aKj93ula1jfwvdN1\nRUfXQ339t+XnXZmjFrFh6n0S2eL2PjFg+mjERiKk5HJuzc+9KS9XCssjAAAgxUGwA9PveLns\n+AbZD07Xw8aBbxyumw09y439N+Tl3JCXA9tP0YtC6Au744n+we+dLoRQJpd7c37uTfk5Eojd\nAADACBDsQLwcJ5cdJ5dt93hX9A9+OGpb1jfwRP/gxVmZN+bn1oqEdFeXdvxRcs2w9blBS6vP\njxAq4fNvL8i9IjsLNpAAAAAmgWAH4muWRLyhWmsIBJ8eGHxr2PqqZfhVy/CJCtmNeTlnqpRs\nDObnx11fMPTvQctrQ8P2cAQhNEsivr0g97zMDA788QEAgHEg2IFE0Aj4/6koe6y0aJVl+MVB\ny7cO17cOVz4PvzIn68qcrAIej+4CGShCUZ/Z7K+ah7+0O0iEOBi2UJ1xU17OUTIp3aUBAACI\nFwh2IHEUHM6dBXm35ed+Mmp/YdAytsbiEaPpFKX87zlZpymVODTYmA6GQHD10PDrFquFIBBC\nWTh3cXbW9Xk5+Tyc7tIAAADEFwQ7kGhsDFuQqVqQqdL7A6ssw6uHrJ/bHJ/bHBlc7vnqjEuy\n1HOkYrprTEmOSGSDdfStYevPLg9CiIXQSQr5VblZZ6tUkJgBACBNQLADtKkQCp4oK36kpOjj\nUdubw9av7M4XBi0vDFoqhYKF6oyFmRnVsMZiErzR6EabY4N1dJPdMbalWxGfd0mW+opsdamA\nT3d1AAAAEgqCHaAZzsIWqjMWqjOGCGKtdfStIesur29538DyvoFqkXBhZsaCTFUNJLx9eKLR\nzTbHhpHRTTZHgCQRQlIO+0K1+rJs9TEyGYzQAQBAeoJgB5JFNo7fmp97a35uq8+/YWR0g3W0\n1ed/0Nf/YF9/qYB/lkp5pkp5rFya5ms5+4Ohz2z2T232750ugqQQQkI2a2yA8zSVQgA7RgAA\nQHqDYAeSTrVIuExUuKy4sNnnf886+qnN3uj1PWsyP2syyzmcExWyeQr5PKW8hJ8u1xkDJPmT\n0/2Vw/mV3dHs84/dKeOwz8lQLMhQnaFSQi86AAAAYyDYgeRVKxLWlhQuLynsC4Y+s9k/G7X/\n4HJ9OGL7cMSGENII+H9RyI+RSY+Ty5i33jNEkls83h+d7h+crp9c7iBJjt1fwuefmaE4U6U8\nTi7jpvfgJQAAgH1BsAMpoJjPuzEv58a8HF80+r3T/ZXD8ZXd2eEPGAJDL5uHEEIlfP4xcukc\niXiuVFIvFqVo4hkMEVs8ni1u789u9xa3dzzMidnsszKUJynk8xTyCqGA3iIBAAAkMwh2IJWI\n2OzTVYrTVQqEkClE/OB0/ehy/+R0tfsDvUPBt4asCCE+i9UgFs2UiOtFwnqxqEYkFCXlRqgU\nQr2BYKPP1+T1N3p9WzyewRAx/qiUwz5RoThGJj1WJpstFadoVAUAAJBgEOxAqsrn4RdlZV6U\nlYkQshLhsVGu3zyebR7vr27Pr27P2NNYCJUK+FqhQCsUVAgEFUJBGZ+fx8MTvJuZMxIxBkP6\nQEDvD3b6A50Bf4c/4I5Ex5/AQqhKKJgjlcyRiI+USurFIthvDQAAwKGCYAeYQI1z52eo5meo\nEEIkhToDgUavr9Hra/L5mrw+QyBoCAQ/tznGn8/BsDweXsTnFfH4mTgnF8czuVw1zs3kcmUc\njozDlrE5h9TU1xWJuiIRVzTqikSsRNhCECPh8DARHgwRfcFgfyjkislwYwQs1hEScb1YVCcS\n1YmFM8RiGScZRxYBAACkEAh2gGlYGKoSCqqEgvPVGWP3OCMRvT+oDwQ6/H69PziWtIzBkDEY\nQsi9v58jYLH4LBaPxRKyWWyESTnsaDRKURSbbQpTlI8kSYoai2uOSOTAJfFZLK1QUMTnlfF3\njx0W8XkwJgcAAGB6QbADzCfncOZIxXvtVBYgSWMwZAqFhojwSDhsCRHD4bA9HBkfeHNHogGS\nPGhoU3A4WThXxuHI2GwZh6PkctRcbiaXm41zs3E8G8cL+Xg2zrRFuwAAAJITBDuQpgQsVqVQ\nUDmJRaZhivJGowghr9cbCoXkcjmbzeZgmCQp12QAAABIZxDsADgILoYpOByEEIfNDrHZCg6H\nDZEOAABAUoKG9QAAAAAADAHBDgAAAACAISDYAQAAAAAwBAQ7AAAAAACGgGAHAAAAAMAQEOwA\nAAAAABgCgh0AAAAAAENAsAMAAAAAYAgIdgAAAAAADAHBDgAAAACAISDYAQAAAAAwBAQ7AAAA\nAACGgGAHAAAAAMAQEOwAAAAAABgCgh0AAAAAAENAsAMAAAAAYAgIdgAAAAAADAHBDgAAAACA\nISDYAQAAAAAwBAQ7AAAAAACGgGAHAAAAAMAQEOwAAAAAABgCgh0AAAAAAENAsAMAAAAAYAgI\ndgAAAAAADAHBDgAAAACAISDYAQAAAAAwBAQ7AAAAAACGgGAHAAAAAMAQEOwAAAAAABgCgh0A\nAAAAAENAsAMAAAAAYAgIdgAAAAAADAHBDgAAAACAISDYAQAAAAAwBAQ7AAAAAACGgGAHAAAA\nAMAQEOwAAAAAABgCgh0AAAAAAENAsAMAAAAAYAgIdgAAAAAADAHBDgAAAACAISDYAQAAAAAw\nBAQ7AAAAAACGgGAHAAAAAMAQEOwAAAAAABgCgh0AAAAAAENAsAMAAAAAYAgIdgAAAAAADAHB\nDgAAAACAISDYAQAAAAAwBAQ7AAAAAACGgGAHAAAAAMAQEOwAAAAAABgCgh0A4P/bu9PYqMq3\ngcOne0Fb6gIYGq2VxaAWEyqIGjeCC1FxSdAIYjRxI0ajQFyICgUlgaCyRFzQ1AjigiDGpRKr\nUo36AaLGJWipVoNKIlEECwJW+v8wybx9WQpCa9ub6/p0tj7nzORJ8+vMmSkAQQg7AIAghB0A\nQBDCDgAgCGEHABCEsAMACELYAQAEIewAAIIQdgAAQQg7AIAghB0AQBDCDgAgCGEHABCEsAMA\nCELYAQAEIewAAIIQdgAAQQg7AIAghB0AQBDCDgAgCGEHABCEsAMACELYAQAEIewAAIIQdgAA\nQQg7AIAghB0AQBDCDgAgCGEHABCEsAMACELYAQAEIewAAIIQdgAAQQg7AIAghB0AQBDCDgAg\nCGEHABCEsAMACELYAQAEIewAAIIQdgAAQQg7AIAghB0AQBDCDgAgCGEHABCEsAMACELYAQAE\nIewAAIIQdgAAQQg7AIAghB0AQBDCDgAgCGEHABCEsAMACELYAQAEIewAAIIQdgAAQQg7AIAg\nhB0AQBDCDgAgCGEHABCEsAMACELYAQAEIewAAIIQdgAAQQg7AIAghB0AQBDCDgAgCGEHABCE\nsAMACELYAQAEIewAAIIQdgAAQQg7AIAghB0AQBDCDgAgCGEHABCEsAMACELYAQAEIewAAIIQ\ndgAAQQg7AIAghB0AQBDCDgAgCGEHABCEsAMACELYAQAEIewAAIIQdgAAQQg7AIAghB0AQBDC\nDgAgCGEHABCEsAMACELYAQAEIewAAIIQdgAAQQg7AIAghB0AQBDCDgAgCGEHABCEsAMACELY\nAQAEIewAAIIQdgAAQQg7AIAghB0AQBDCDgAgCGEHABCEsAMACELYAQAEIewAAIIQdgAAQQg7\nAIAghB0AQBDCDgAgCGEHABCEsAMACELYAQAEIewAAIIQdgAAQQg7AIAghB0AQBDCDgAgCGEH\nABCEsAMACELYAQAEIewAAIIQdgAAQQg7AIAghB0AQBDCDgAgCGEHABCEsAMACELYAQAEIewA\nAIIQdgAAQQg7AIAghB0AQBDCDgAgCGEHABCEsAMACELYAQAEIewAAIIQdgAAQQg7AIAghB0A\nQBDCDgAgCGEHABCEsAMACELYAQAEIewAAIIQdgAAQQg7AIAghB0AQBDCDgAgCGEHABCEsAMA\nCELYAQAEIewAAIIQdgAAQQg7AIAghB0AQBDCDgAgCGEHABCEsAMACELYAQAEIewAAIIQdgAA\nQQg7AIAgstv7AtpQXl5ednZ2ZqZ43YusrKz8/Pz2vopOID8/Pycnx4zaKzNqH+Xn5+fm5mZk\nZLT3hXR02dnZnqV90aVLl7y8PM/VXmVlZQV+ljKampra+xoAAGgFXnsAAAhC2AEABCHsAACC\nEHYAAEEIOwCAIIQdAEAQwg4AIIiwX1C8atWqBQsW/PTTT926dRs2bNjVV18d+NsIaUF1dXVN\nTc0PP/ywbdu2Xr16XXTRReedd15q15tvvvnkk082P3jq1Kknn3xyarmFKdTqu+hEOs60MaNi\nGDduXF1dXfMtGRkZL774YpcuXTrOZKMTiRl233777YMPPjh8+PBx48Z999138+bN27FjxzXX\nXNPe10U7eO+990488cRLL720a9euH3/88dy5cxsbG4cPH57aW1BQMHXq1PTBvXr1Si20MIVa\nfRedTkeYNmZUGOPHj9+2bVt6dfr06cXFxV26dEmtdoTJRucSM+yWLl1aXFx88803J0lSUlKy\nbt261157beTIkXl5ee19afzXpk2bll4+4YQT6uvrP/roo3TYZWVlHXfccbv+VAtTqNV3/QdP\nAq2rI0wbMyqM4uLi9HJdXd26detuvPHG9JaOMNna+hmgdcW8x2716tUDBw5Mrw4cOHDr1q3f\nf/99O14SHcT27du7deuWXv3zzz+vvfbaUaNG3XXXXR999FF6ewtTqNV30el0hGljRoX01ltv\n9ezZs7y8PL2lI0w2OpeAr9g1NTX98ccfhx12WHpLavn3339vv4uiQ6iurq6rq7vppptSq0cf\nffTYsWNLSkq2b99eU1Mzffr0G264YcSIES1MoVbf1bYPmDbQEaaNGRVSQ0PDBx980PzOto4w\n2dr2MdMGAoYd7NaHH374xBNP3HnnnX379k1tGTBgwIABA1LLZWVlmzdvXrJkyYgRI9rvGukE\nTBvaSHV1dVNT07Bhw9JbTDb2Q8C3YjMyMoqKijZs2JDeklo+/PDD2++iaGdVVVWzZ8+eMGHC\nWWedtadj+vfvv2HDhsbGxhamUKvvas0HSXtol2ljRsXT1NRUVVV1xhlnNL9XZCd+R7EvAoZd\nkiT9+/f/9NNP06uffvppfn7+bu8/5WDw4osvVlZW3n///UOGDGnhsNWrVxcVFWVnZyctTqFW\n30Wn1l7TxowK5rPPPlu3bl36c1275XcU+yJr8uTJ7X0Nra9Hjx5Lly7duHFj9+7dP/vss+ee\ne+7SSy9tflsoB4/58+cvW7bshhtu6NWr14YNGzZs2NDQ0JD6m/ixxx5raGjYunXrL7/8snjx\n4hUrVlx99dX9+/dPWpxCrb6LzqWDTBszKphnnnkmJyfn2muvbb6xg0w2OpeMpqam9r6GNrFy\n5cqFCxeuXbs29UWLo0aN8kWLB6fRo0f/+eefzbccddRRTz31VJIk8+fPX7Vq1W+//Zabm1tc\nXDxixIgzzzwzfVgLU6jVd9GJdJxpY0aFsX79+htvvPHmm2/e6RW7jjPZ6ETChh0AwMEm5j12\nAAAHIWEHABCEsAMACELYAQAEIewAAIIQdgAAQQg7oAOprq7OyMh49tlnD/CYTuqVV17JyMhY\ntmxZe18I0FkJO+DfWbVqVUYzXbt2PemkkyZPnrxly5ZWP1dtbe3kyZO/+OKLVh/5IDFjxowr\nr7yyT58+mZmZGRkZjY2N7X1FQNvKbu8LADqlwYMHjxw5MkmS9evXv/baaxUVFTU1Ne+9994B\nflX90KFD//rrr5ycnNRqbW1tRUVFnz59BgwYsKdjaMHEiRMLCgoGDhy4adOm9evXt/flAG1O\n2AH7o6ysbMKECanlqVOnDho0aMWKFTU1Neecc86BDJuZmZmfn3/gxxxUtmzZ0rVr193uqq2t\nTf0f9yFDhgg7OBh4KxY4ULm5uRdeeGGSJGvWrEmS5I8//hg/fnxpaWleXl7Pnj1Hjx5dV1eX\nPrixsXHGjBllZWUFBQUFBQV9+/a97rrr0v/Pt/n9c5MnT77kkkuSJBkzZkzqbd9UNe56j13L\nZ0zduLZkyZLp06f369cvLy/vmGOOeeihh/b6DxXbaNjGxsZZs2aVl5cfcsghBQUFAwYMmDRp\n0k7HzJ8//4QTTtjtmKnzvvzyyxUVFX379s3NzZ0yZcqezpWqOuDg4RU7oBXU1tYmSXLEEUds\n3rz5rLPO+vLLL0ePHn366aevWbPm8ccfr6qq+uSTT44//vgkSe69996ZM2eOGjXq9ttvz8zM\n/PHHH994441NmzYVFBTsNOZ1112Xl5c3ceLEiRMnnnfeeUmSFBUV7XrqvZ4x5a677urXr9+c\nOXOKioqefvrp++6774gjjrjlllv29IjaaNjGxsaLL754+fLlZ5999gMPPFBYWPjNN98sXry4\noqIifczDDz+8du3aMWPGFBUVLVq0aLdj3n333cXFxdOmTTvqqKO8Kw38nyaAf2PlypVJklx1\n1VX19fX19fUrV65MvSfbvXv3jRs3pgIl9SJTyvLly5MkueCCC1KrpaWl55577p4Gf+edd5Ik\nqaysTK2+/vrrSZIsWLCghWP2esbFixcnSXLKKafs2LEjteWff/7p27dv//79W3iYbTTso48+\nmiTJbbfdlv6p1A82H7OkpGTjxo2pLZs3b+7Ro0fzMVPH9OvX7++//27hRDs59dRTkyT5Vz8C\ndEbeigX2x0svvVRaWlpaWjpo0KCZM2eWlZVVVVUVFhYuWbLk0EMPHTduXPrI888//7TTTnvn\nnXc2bdqUJElRUdHq1atTddgq9nrGlNT7uanlzMzMU0455bvvvtuxY8d/POzChQu7dOkybdq0\n5p8yycz8f7+Kx44dW1hYmFru2rXrmWeeueuY119/fXa2t1yAnQk7YH8MGzbs1VdfXbZs2dtv\nv11bW/vFF1+Ul5cnSfL999/37t17pw83lJWV7dix44cffkiSZObMmX///ffgwYNLSkpGjx5d\nWVl5gN+Tstczphx99NHNDygsLNy+fXv63r7/bNja2to+ffoceuihLTyi3r17N1898sgjdx2z\ntLS0hRGAg5Y/+ID9UVJSctlll+26vampqeVvPBk6dGh9ff3bb7/9/vvv19TULFq0aNKkSZ98\n8klxcfH+Xclez5iy22Oa9vxBh3Ycdrcvxe00Zl5e3l6vDTgIecUOaE29e/euq6vbunVr841f\nffVVZmbmsccem1otKCgYOXLkvHnzvv766xdeeGHt2rVz5szZ7Wj7klb7csb90EbDHn/88WvW\nrGloaNjvEQBaIOyA1nTFFVc0NDTMmjUrvaW6uvrjjz8eNmxY6r6x33//vfnxQ4YM2XVjWuqj\nsnvau49nbKMHsn+uueaav/766/7772++sYVX+AD+FW/FAq1pwoQJr7zyyr333vv111+nvyXk\nsMMOmz17duqAXr16XXzxxeXl5cXFxb/++uvTTz+dlZU1ZsyY3Y528skn5+fnz507Nzc3t6io\nqEePHkOHDv23Z2yjB7J/br311jfeeGPWrFmff/758OHDCwsL16xZs3z58q+++upAht2T559/\nvr6+PkmSn3/+OUmSadOmZWZmZmdn33PPPW1xOqDdCTugNR1yyCEffvjhlClTli5d+tJLLxUV\nFV1++eVTpkzp06dP6oDx48evWLHikUce2bhxY48eBZSmkwAAAJ1JREFUPQYNGlRZWXnaaaft\ndrRu3botWrSooqLijjvu2LZt29lnn71r2O31jG30QPZPTk5OVVXVrFmzFixYMGnSpJycnNLS\n0tQ/Z2sLlZWV7777bno19U3IeXl5wg6iyvAWAABADO6xAwAIQtgBAAQh7AAAghB2AABBCDsA\ngCCEHQBAEMIOACAIYQcAEISwAwAIQtgBAAQh7AAAgvgfwQF1a7AfJyEAAAAASUVORK5CYII=",
      "text/plain": [
       "plot without title"
      ]
     },
     "metadata": {
      "image/png": {
       "height": 420,
       "width": 420
      }
     },
     "output_type": "display_data"
    }
   ],
   "source": [
    "library(ggtranscript)\n",
    "\n",
    "df = data.frame(seqnames = rep(1,5),\n",
    "                start = seq(1,1000000,100000),\n",
    "                end = seq(1,1000000,100000) + 10000,\n",
    "                strand = \"+\",\n",
    "                type = \"exons\",\n",
    "                gene_name = \"Gene\",\n",
    "                transcript_name = \"transcript_name\",\n",
    "                transcript_biotype = \"transcript_biotype\")\n",
    "\n",
    "links_df = data.frame(seqnames = rep(1,10),\n",
    "                      start = rep(500000,10),\n",
    "                      end = df$start,\n",
    "                      strand = \"+\",\n",
    "                      n_studies = c(2,5,10, 30, 1,4, 5,1 ,1 ,1))\n",
    "\n",
    "links_df_crispr = data.frame(seqnames = 1,\n",
    "                      start = 500000,\n",
    "                      end = 700000,\n",
    "                      strand = \"+\",\n",
    "                      n_studies = c(1))\n",
    "\n",
    "df %>% ggplot(aes(\n",
    "        xstart = start,\n",
    "        xend = end,\n",
    "        y = \"\",\n",
    "    )) + geom_range(fill = \"black\", height = 0.025) +\n",
    "  geom_junction(\n",
    "    data = links_df,\n",
    "    aes(size = n_studies),\n",
    "    junction.orientation = \"alternating\",\n",
    "    junction.y.max = 0.2,\n",
    "    angle = 90,\n",
    "    ncp = 100, \n",
    "    colour = \"#00BFC4\") +\n",
    "  geom_junction(\n",
    "    data = links_df_crispr,\n",
    "    aes(size = n_studies),\n",
    "    junction.orientation = \"alternating\",\n",
    "    junction.y.max = 0.2,\n",
    "    angle = 90,\n",
    "    ncp = 100, \n",
    "    colour = \"#F8766D\") +\n",
    "  ylab(\"\") + xlab(\"Position on chr 1\") +\n",
    "  scale_size_continuous(range = c(0.2, 2), guide = \"none\") +\n",
    "  theme_minimal()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "id": "fb17c188-6544-4fc2-85c8-bfcb45eac168",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A tibble: 5 × 5</caption>\n",
       "<thead>\n",
       "\t<tr><th scope=col>seqnames</th><th scope=col>start</th><th scope=col>end</th><th scope=col>strand</th><th scope=col>mean_count</th></tr>\n",
       "\t<tr><th scope=col>&lt;fct&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;fct&gt;</th><th scope=col>&lt;dbl&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><td>chr21</td><td>31659787</td><td>31666448</td><td>+</td><td>0.4632353</td></tr>\n",
       "\t<tr><td>chr21</td><td>31659842</td><td>31660554</td><td>+</td><td>0.8308824</td></tr>\n",
       "\t<tr><td>chr21</td><td>31659842</td><td>31663794</td><td>+</td><td>0.3161765</td></tr>\n",
       "\t<tr><td>chr21</td><td>31659842</td><td>31667257</td><td>+</td><td>4.3529412</td></tr>\n",
       "\t<tr><td>chr21</td><td>31660351</td><td>31663789</td><td>+</td><td>0.3235294</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A tibble: 5 × 5\n",
       "\\begin{tabular}{lllll}\n",
       " seqnames & start & end & strand & mean\\_count\\\\\n",
       " <fct> & <int> & <int> & <fct> & <dbl>\\\\\n",
       "\\hline\n",
       "\t chr21 & 31659787 & 31666448 & + & 0.4632353\\\\\n",
       "\t chr21 & 31659842 & 31660554 & + & 0.8308824\\\\\n",
       "\t chr21 & 31659842 & 31663794 & + & 0.3161765\\\\\n",
       "\t chr21 & 31659842 & 31667257 & + & 4.3529412\\\\\n",
       "\t chr21 & 31660351 & 31663789 & + & 0.3235294\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A tibble: 5 × 5\n",
       "\n",
       "| seqnames &lt;fct&gt; | start &lt;int&gt; | end &lt;int&gt; | strand &lt;fct&gt; | mean_count &lt;dbl&gt; |\n",
       "|---|---|---|---|---|\n",
       "| chr21 | 31659787 | 31666448 | + | 0.4632353 |\n",
       "| chr21 | 31659842 | 31660554 | + | 0.8308824 |\n",
       "| chr21 | 31659842 | 31663794 | + | 0.3161765 |\n",
       "| chr21 | 31659842 | 31667257 | + | 4.3529412 |\n",
       "| chr21 | 31660351 | 31663789 | + | 0.3235294 |\n",
       "\n"
      ],
      "text/plain": [
       "  seqnames start    end      strand mean_count\n",
       "1 chr21    31659787 31666448 +      0.4632353 \n",
       "2 chr21    31659842 31660554 +      0.8308824 \n",
       "3 chr21    31659842 31663794 +      0.3161765 \n",
       "4 chr21    31659842 31667257 +      4.3529412 \n",
       "5 chr21    31660351 31663789 +      0.3235294 "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "sod1_junctions"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "R",
   "language": "R",
   "name": "ir"
  },
  "language_info": {
   "codemirror_mode": "r",
   "file_extension": ".r",
   "mimetype": "text/x-r-source",
   "name": "R",
   "pygments_lexer": "r",
   "version": "4.3.1"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
